// @douyinfe/semi-ui/lib/es/anchor/link.js
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import _bindInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/bind";
import React from 'react';
import cls from 'classnames';
import PropTypes from 'prop-types';
import BaseComponent from '../_base/baseComponent';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/anchor/constants';
import LinkFoundation from '@douyinfe/semi-foundation/lib/es/anchor/linkFoundation';
import AnchorContext from './anchor-context';
import Typography from '../typography/index';
const prefixCls = cssClasses.PREFIX; // eslint-disable-next-line @typescript-eslint/ban-types

export default class Link extends BaseComponent {
  constructor(props) {
    var _context;

    super(props);

    this.renderTitle = () => {
      const {
        href,
        title,
        disabled = false
      } = this.props;
      const {
        activeLink,
        showTooltip,
        position,
        size
      } = this.context;
      const active = activeLink === href;
      const linkTitleCls = cls("".concat(prefixCls, "-link-tooltip"), {
        ["".concat(prefixCls, "-link-tooltip-small")]: size === 'small',
        ["".concat(prefixCls, "-link-tooltip-active")]: active,
        ["".concat(prefixCls, "-link-tooltip-disabled")]: disabled
      });
      const toolTipOpt = position ? {
        position
      } : {};

      if (showTooltip) {
        return /*#__PURE__*/React.createElement(Typography.Text, {
          size: size === 'default' ? 'normal' : 'small',
          ellipsis: {
            showTooltip: {
              opts: _Object$assign({}, toolTipOpt)
            }
          },
          type: 'tertiary',
          className: linkTitleCls
        }, title);
      } else {
        return title;
      }
    };

    this.renderChildren = () => {
      const {
        activeLink,
        childMap
      } = this.context;
      const {
        href,
        children
      } = this.props;

      if (!this.context.autoCollapse) {
        return this.props.children;
      }

      return activeLink === href || childMap[href] && childMap[href].has(activeLink) ? children : null;
    };

    this.foundation = new LinkFoundation(this.adapter);
    this.handleClick = _bindInstanceProperty(_context = this.handleClick).call(_context, this);
  }

  get adapter() {
    return _Object$assign(_Object$assign({}, super.adapter), {
      addLink: href => {
        this.context.addLink(href);
      },
      removeLink: href => {
        this.context.removeLink(href);
      }
    });
  }

  handleAddLink() {
    this.foundation.handleAddLink();
  }

  handleRemoveLink() {
    this.foundation.handleRemoveLink();
  }

  handleUpdateLink(href, prevHref) {
    this.foundation.handleUpdateLink(href, prevHref);
  }

  handleClick(e) {
    const {
      disabled,
      href
    } = this.props;
    const {
      onClick
    } = this.context;
    !disabled && onClick(e, href);
  }

  componentDidMount() {
    this.handleAddLink();
  }

  componentDidUpdate(prevProps) {
    const prevHref = prevProps.href;
    const {
      href
    } = this.props;
    this.handleUpdateLink(href, prevHref);
  }

  componentWillUnmount() {
    this.handleRemoveLink();
  }

  render() {
    const {
      href,
      className,
      style,
      disabled = false,
      title
    } = this.props;
    const {
      activeLink,
      showTooltip
    } = this.context;
    const active = activeLink === href;
    const linkCls = cls("".concat(prefixCls, "-link"), className);
    const linkTitleCls = cls("".concat(prefixCls, "-link-title"), {
      ["".concat(prefixCls, "-link-title-active")]: active,
      ["".concat(prefixCls, "-link-title-disabled")]: disabled
    });
    const ariaProps = {
      'aria-disabled': disabled,
      'aria-label': href
    };

    if (active) {
      ariaProps['aria-details'] = 'active';
    }

    if (!showTooltip && typeof title === 'string') {
      ariaProps['title'] = title;
    }

    return /*#__PURE__*/React.createElement("div", {
      className: linkCls,
      style: style
    }, /*#__PURE__*/React.createElement("div", _Object$assign({
      role: "link",
      tabIndex: 0
    }, ariaProps, {
      className: linkTitleCls,
      onClick: e => this.handleClick(e),
      onKeyPress: e => this.handleClick(e)
    }), this.renderTitle()), this.renderChildren());
  }

}
Link.propTypes = {
  href: PropTypes.string,
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  className: PropTypes.string,
  style: PropTypes.object
};
Link.defaultProps = {
  href: '#',
  title: '',
  className: ''
};
Link.contextType = AnchorContext;
