// @douyinfe/semi-animation-react/lib/cjs/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

_Object$defineProperty(exports, "Animation", {
  enumerable: true,
  get: function () {
    return _Animation.default;
  }
});

_Object$defineProperty(exports, "KeyFrames", {
  enumerable: true,
  get: function () {
    return _KeyFrames.default;
  }
});

_Object$defineProperty(exports, "StyledAnimation", {
  enumerable: true,
  get: function () {
    return _StyledAnimation.default;
  }
});

_Object$defineProperty(exports, "StyledTransition", {
  enumerable: true,
  get: function () {
    return _StyledTransition.default;
  }
});

_Object$defineProperty(exports, "Transition", {
  enumerable: true,
  get: function () {
    return _Transition.default;
  }
});

_Object$defineProperty(exports, "interpolate", {
  enumerable: true,
  get: function () {
    return _semiAnimation.interpolate;
  }
});

_Object$defineProperty(exports, "presets", {
  enumerable: true,
  get: function () {
    return _semiAnimation.presets;
  }
});

var _StyledAnimation = _interopRequireDefault(require("./src/StyledAnimation"));

var _StyledTransition = _interopRequireDefault(require("./src/StyledTransition"));

var _Animation = _interopRequireDefault(require("./src/Animation"));

var _KeyFrames = _interopRequireDefault(require("./src/KeyFrames"));

var _Transition = _interopRequireDefault(require("./src/Transition"));

var _semiAnimation = require("@douyinfe/semi-animation");
