// @douyinfe/semi-ui/lib/cjs/checkbox/checkboxGroup.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _bind = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/bind"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _slice = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/slice"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/checkbox/constants");

var _checkboxGroupFoundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/checkbox/checkboxGroupFoundation"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _context7 = require("./context");

var _checkbox = _interopRequireDefault(require("./checkbox"));

/* eslint-disable max-len */
class CheckboxGroup extends _baseComponent.default {
  constructor(props) {
    var _context;

    super(props);
    this.state = {
      value: props.value || props.defaultValue
    };
    this.foundation = new _checkboxGroupFoundation.default(this.adapter);
    this.onChange = (0, _bind.default)(_context = this.onChange).call(_context, this);
  }

  get adapter() {
    return (0, _assign.default)((0, _assign.default)({}, super.adapter), {
      updateGroupValue: value => {
        this.setState({
          value
        });
      },
      notifyChange: evt => {
        this.props.onChange && this.props.onChange(evt);
      }
    });
  }

  componentDidMount() {
    this.foundation.init();
  }

  componentDidUpdate(prevProps) {
    if (!(0, _isEqual2.default)(prevProps.value, this.props.value)) {
      this.foundation.handlePropValueChange(this.props.value);
    }
  }

  componentWillUnmount() {
    this.foundation.destroy();
  }

  onChange(evt) {
    this.foundation.handleChange(evt);
  }

  render() {
    var _context2, _context3, _context4;

    const {
      children,
      options,
      prefixCls,
      direction,
      className,
      id,
      style,
      type,
      disabled
    } = this.props;
    const isPureCardType = type === _constants.strings.TYPE_PURECARD;
    const isCardType = type === _constants.strings.TYPE_CARD || isPureCardType;
    const prefix = prefixCls || _constants.checkboxGroupClasses.PREFIX;
    const prefixClsDisplay = (0, _classnames.default)({
      [prefix]: true,
      ["".concat(prefix, "-wrapper")]: true,
      [(0, _concat.default)(_context2 = "".concat(prefix, "-")).call(_context2, direction)]: direction,
      [(0, _concat.default)(_context3 = "".concat(prefix, "-")).call(_context3, direction, "-cardType")]: direction && isCardType
    }, className);
    const realValue = (0, _slice.default)(_context4 = this.state.value).call(_context4);
    let inner;

    if (options) {
      var _context5;

      inner = (0, _map.default)(_context5 = options || []).call(_context5, (option, index) => {
        if (typeof option === 'string') {
          return /*#__PURE__*/_react.default.createElement(_checkbox.default, {
            role: "listitem",
            key: index,
            disabled: this.props.disabled,
            value: option,
            prefixCls: prefixCls
          }, option);
        } else {
          return /*#__PURE__*/_react.default.createElement(_checkbox.default, {
            role: "listitem",
            key: index,
            disabled: option.disabled || this.props.disabled,
            value: option.value,
            prefixCls: prefixCls,
            extra: option.extra,
            className: option.className,
            style: option.style,
            onChange: option.onChange
          }, option.label);
        }
      });
    } else if (children) {
      var _context6;

      inner = (0, _map.default)(_context6 = _react.default.Children.toArray(children)).call(_context6, (itm, index) => /*#__PURE__*/_react.default.cloneElement(itm, {
        key: index,
        role: 'listitem'
      }));
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      id: id,
      role: "list",
      "aria-label": this.props['aria-label'],
      className: prefixClsDisplay,
      style: style,
      "aria-labelledby": this.props['aria-labelledby'],
      "aria-describedby": this.props['aria-describedby']
    }, /*#__PURE__*/_react.default.createElement(_context7.Context.Provider, {
      value: {
        checkboxGroup: {
          onChange: this.onChange,
          value: realValue,
          disabled: this.props.disabled,
          name: this.foundation.getFormatName(),
          isCardType,
          isPureCardType
        }
      }
    }, inner));
  }

}

CheckboxGroup.propTypes = {
  'aria-describedby': _propTypes.default.string,
  'aria-errormessage': _propTypes.default.string,
  'aria-invalid': _propTypes.default.bool,
  'aria-labelledby': _propTypes.default.string,
  'aria-required': _propTypes.default.bool,
  defaultValue: _propTypes.default.array,
  disabled: _propTypes.default.bool,
  name: _propTypes.default.string,
  options: _propTypes.default.array,
  value: _propTypes.default.array,
  onChange: _propTypes.default.func,
  children: _propTypes.default.node,
  prefixCls: _propTypes.default.string,
  direction: _propTypes.default.oneOf(_constants.strings.DIRECTION_SET),
  className: _propTypes.default.string,
  type: _propTypes.default.oneOf([_constants.strings.TYPE_DEFAULT, _constants.strings.TYPE_CARD, _constants.strings.TYPE_PURECARD]),
  style: _propTypes.default.object
};
CheckboxGroup.defaultProps = {
  disabled: false,
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  onChange: () => {},
  type: _constants.strings.TYPE_DEFAULT,
  defaultValue: [],
  direction: _constants.strings.DEFAULT_DIRECTION
};
var _default = CheckboxGroup;
exports.default = _default;
