// @douyinfe/semi-ui/lib/cjs/checkbox/checkbox.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/includes"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _isBoolean2 = _interopRequireDefault(require("lodash/isBoolean"));

var _isUndefined2 = _interopRequireDefault(require("lodash/isUndefined"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/checkbox/constants");

var _checkboxFoundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/checkbox/checkboxFoundation"));

var _checkboxInner = _interopRequireDefault(require("./checkboxInner"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

require("@douyinfe/semi-foundation/lib/cjs/checkbox/checkbox.css");

var _context2 = require("./context");

var _uuid = require("@douyinfe/semi-foundation/lib/cjs/utils/uuid");

/* eslint-disable max-len */
class Checkbox extends _baseComponent.default {
  constructor(props) {
    super(props);

    this.handleChange = e => this.foundation.handleChange(e);

    this.handleEnterPress = e => this.foundation.handleEnterPress(e);

    this.handleFocusVisible = event => {
      this.foundation.handleFocusVisible(event);
    };

    this.handleBlur = event => {
      this.foundation.handleBlur();
    };

    const checked = false;
    this.state = {
      checked: props.checked || props.defaultChecked || checked,
      addonId: props.addonId,
      extraId: props.extraId,
      focusVisible: false
    };
    this.checkboxEntity = null;
    this.foundation = new _checkboxFoundation.default(this.adapter);
  }

  get adapter() {
    return (0, _assign.default)((0, _assign.default)({}, super.adapter), {
      setNativeControlChecked: checked => {
        this.setState({
          checked
        });
      },
      notifyChange: cbContent => {
        const {
          onChange
        } = this.props;
        onChange && onChange(cbContent);
      },
      getIsInGroup: () => this.isInGroup(),
      getGroupValue: () => this.context && this.context.checkboxGroup.value || [],
      notifyGroupChange: cbContent => {
        this.context.checkboxGroup.onChange(cbContent);
      },
      getGroupDisabled: () => this.context && this.context.checkboxGroup.disabled,
      setAddonId: () => {
        this.setState({
          addonId: (0, _uuid.getUuidShort)({
            prefix: 'addon'
          })
        });
      },
      setExtraId: () => {
        this.setState({
          extraId: (0, _uuid.getUuidShort)({
            prefix: 'extra'
          })
        });
      },
      setFocusVisible: focusVisible => {
        this.setState({
          focusVisible
        });
      },
      focusCheckboxEntity: () => {
        this.focus();
      }
    });
  }

  componentDidUpdate(prevProps) {
    if (this.props.checked !== prevProps.checked) {
      if ((0, _isUndefined2.default)(this.props.checked)) {
        this.foundation.setChecked(false);
      } else if ((0, _isBoolean2.default)(this.props.checked)) {
        this.foundation.setChecked(this.props.checked);
      }
    }
  }

  isInGroup() {
    // Why do we need to determine whether there is a value in props?
    // If there is no value in the props of the checkbox in the context of the checkboxGroup, 
    // it will be considered not to belong to the checkboxGroup。
    return Boolean(this.context && this.context.checkboxGroup && 'value' in this.props);
  }

  focus() {
    this.checkboxEntity && this.checkboxEntity.focus();
  }

  blur() {
    this.checkboxEntity && this.checkboxEntity.blur();
  }

  render() {
    const {
      disabled,
      style,
      prefixCls,
      className,
      indeterminate,
      children,
      onMouseEnter,
      onMouseLeave,
      extra,
      value,
      role,
      tabIndex,
      id
    } = this.props;
    const {
      checked,
      addonId,
      extraId,
      focusVisible
    } = this.state;
    const props = {
      checked,
      disabled
    };
    const inGroup = this.isInGroup();

    if (inGroup) {
      if (this.context.checkboxGroup.value) {
        var _context;

        const realChecked = (0, _includes.default)(_context = this.context.checkboxGroup.value || []).call(_context, value);
        props.checked = realChecked;
      }

      if (this.context.checkboxGroup.disabled) {
        props.disabled = this.context.checkboxGroup.disabled || this.props.disabled;
      }

      const {
        isCardType,
        isPureCardType
      } = this.context.checkboxGroup;
      props.isCardType = isCardType;
      props.isPureCardType = isPureCardType;
      props['name'] = this.context.checkboxGroup.name;
    }

    const prefix = prefixCls || _constants.checkboxClasses.PREFIX;
    const focusOuter = props.isCardType || props.isPureCardType;
    const wrapper = (0, _classnames.default)(prefix, {
      ["".concat(prefix, "-disabled")]: props.disabled,
      ["".concat(prefix, "-indeterminate")]: indeterminate,
      ["".concat(prefix, "-checked")]: props.checked,
      ["".concat(prefix, "-unChecked")]: !props.checked,
      ["".concat(prefix, "-cardType")]: props.isCardType,
      ["".concat(prefix, "-cardType_disabled")]: props.disabled && props.isCardType,
      ["".concat(prefix, "-cardType_unDisabled")]: !(props.disabled && props.isCardType),
      ["".concat(prefix, "-cardType_checked")]: props.isCardType && props.checked && !props.disabled,
      ["".concat(prefix, "-cardType_checked_disabled")]: props.isCardType && props.checked && props.disabled,
      [className]: Boolean(className),
      ["".concat(prefix, "-focus")]: focusVisible && focusOuter
    });
    const extraCls = (0, _classnames.default)("".concat(prefix, "-extra"), {
      ["".concat(prefix, "-cardType_extra_noChildren")]: props.isCardType && !children
    });
    const name = inGroup && this.context.checkboxGroup.name;
    const xSemiPropChildren = this.props['x-semi-children-alias'] || 'children';

    const renderContent = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children ? /*#__PURE__*/_react.default.createElement("span", {
      id: addonId,
      className: "".concat(prefix, "-addon"),
      "x-semi-prop": xSemiPropChildren
    }, children) : null, extra ? /*#__PURE__*/_react.default.createElement("div", {
      id: extraId,
      className: extraCls,
      "x-semi-prop": "extra"
    }, extra) : null);

    return (
      /*#__PURE__*/
      // label is better than span, however span is here which is to solve gitlab issue #364
      _react.default.createElement("span", {
        role: role,
        tabIndex: tabIndex,
        style: style,
        className: wrapper,
        id: id,
        onMouseEnter: onMouseEnter,
        onMouseLeave: onMouseLeave,
        onClick: this.handleChange,
        onKeyPress: this.handleEnterPress,
        "aria-labelledby": this.props['aria-labelledby']
      }, /*#__PURE__*/_react.default.createElement(_checkboxInner.default, (0, _assign.default)({}, this.props, props, {
        addonId: children && addonId,
        extraId: extra && extraId,
        isPureCardType: props.isPureCardType,
        ref: ref => {
          this.checkboxEntity = ref;
        },
        focusInner: focusVisible && !focusOuter,
        onInputFocus: this.handleFocusVisible,
        onInputBlur: this.handleBlur
      })), props.isCardType ? /*#__PURE__*/_react.default.createElement("div", null, renderContent()) : renderContent())
    );
  }

}

Checkbox.contextType = _context2.Context;
Checkbox.propTypes = {
  'aria-describedby': _propTypes.default.string,
  'aria-errormessage': _propTypes.default.string,
  'aria-invalid': _propTypes.default.bool,
  'aria-labelledby': _propTypes.default.string,
  'aria-required': _propTypes.default.bool,
  // Specifies whether it is currently selected
  checked: _propTypes.default.bool,
  // Initial check
  defaultChecked: _propTypes.default.bool,
  // Failure state
  disabled: _propTypes.default.bool,
  // Set indeterminate state, only responsible for style control
  indeterminate: _propTypes.default.bool,
  // Callback function when changing
  onChange: _propTypes.default.func,
  value: _propTypes.default.any,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  prefixCls: _propTypes.default.string,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func,
  extra: _propTypes.default.node,
  index: _propTypes.default.number,
  'aria-label': _propTypes.default.string,
  tabIndex: _propTypes.default.number,
  preventScroll: _propTypes.default.bool
};
Checkbox.defaultProps = {
  defaultChecked: false,
  indeterminate: false,
  onChange: _noop2.default,
  onMouseEnter: _noop2.default,
  onMouseLeave: _noop2.default
};
var _default = Checkbox;
exports.default = _default;
