// @douyinfe/semi-ui/lib/es/notification/NoticeTransition.js
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
// @ts-ignore  Currently there is no types for semi-animation-react
import { Transition } from '@douyinfe/semi-animation-react';
import React from 'react';
export default function NoticeTransition() {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let {
    motion = {}
  } = props;
  const {
    position = 'topRight'
  } = props;

  const formatStyle = function formatStyle(_ref) {
    let {
      translate,
      opacity
    } = _ref;
    let transform = "translateX(".concat(translate, "%)");

    if (position && typeof position === 'string') {
      if (/left/i.test(position)) {
        transform = "translateX(".concat(-translate, "%)");
      } else if (/right/i.test(position)) {
        transform = "translateX(".concat(translate, "%)");
      } else if (/top/i.test(position)) {
        transform = "translateY(".concat(-translate, "%)");
      } else {
        transform = "translateY(".concat(translate, "%)");
      }
    }

    return {
      transform,
      opacity
    };
  };

  if (typeof motion === 'function') {
    motion = motion(props);
  } else if (!motion || typeof motion !== 'object') {
    motion = {};
  }

  return /*#__PURE__*/React.createElement(Transition // onFrame={style => console.log(formatStyle(style))}
  , _Object$assign({
    // onFrame={style => console.log(formatStyle(style))}
    from: {
      translate: 100,
      opacity: 0
    },
    enter: {
      translate: {
        val: 0,
        tension: 560,
        friction: 32
      },
      opacity: {
        val: 1,
        duration: 200
      }
    },
    leave: {
      translate: {
        val: 100,
        easing: 'easeOutCubic',
        duration: 300
      },
      opacity: {
        val: 0,
        duration: 200
      }
    }
  }, motion), typeof props.children === 'function' ? transitionStyle => props.children(formatStyle(transitionStyle)) : props.children);
}
