// @douyinfe/semi-ui/lib/cjs/trigger/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

/* eslint-disable @typescript-eslint/ban-types */
var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

/**
 * `Trigger` is a HOC that will cover the inner of components which have popups
 */
class Trigger extends _react.default.PureComponent {
  render() {
    // eslint-disable-next-line no-unused-vars
    const _a = this.props,
          {
      triggerRender,
      componentName
    } = _a,
          rest = __rest(_a, ["triggerRender", "componentName"]);

    return triggerRender((0, _assign.default)({}, rest));
  }

}

Trigger.propTypes = {
  /**
   * ({ value?: any, className?: string, style?: React.CSSProperties, ... }) => React.ReactNode
   */
  triggerRender: _propTypes.default.func.isRequired,

  /**
   * e.g. "AutoComplete", "DatePicker", ...
   */
  componentName: _propTypes.default.string,
  componentProps: _propTypes.default.object,
  value: _propTypes.default.any,
  inputValue: _propTypes.default.string,
  placeholder: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  className: _propTypes.default.string,
  style: _propTypes.default.object
};
var _default = Trigger;
exports.default = _default;
