// @douyinfe/semi-foundation/lib/cjs/overflowList/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.numbers = exports.cssClasses = void 0;

var _values = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/values"));

var _constants = require("../base/constants");

const cssClasses = {
  PREFIX: _constants.BASE_CLASS_PREFIX + '-overflow-list'
};
exports.cssClasses = cssClasses;
const MODE_MAP = {
  COLLAPSE: 'collapse',
  SCROLL: 'scroll'
};
const BOUNDARY_MAP = {
  START: 'start',
  END: 'end'
};
const OVERFLOW_DIR = {
  NONE: 0,
  GROW: 1,
  SHRINK: 2
};
const strings = {
  BOUNDARY_SET: (0, _values.default)(BOUNDARY_MAP),
  POSITION_SET: ['vertical', 'horizontal'],
  MODE_SET: (0, _values.default)(MODE_MAP),
  MODE_MAP,
  BOUNDARY_MAP,
  OVERFLOW_DIR
};
exports.strings = strings;
const numbers = {};
exports.numbers = numbers;
