// @douyinfe/semi-ui/lib/cjs/tabs/TabPaneTransition.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _toInteger2 = _interopRequireDefault(require("lodash/toInteger"));

var _react = _interopRequireDefault(require("react"));

var _semiAnimationReact = require("@douyinfe/semi-animation-react");

/* eslint-disable react/destructuring-assignment */
const TabPaneTransition = function () {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const direction = props.direction ? 1 : -1;
  const {
    mode
  } = props;
  let {
    motion
  } = props;
  const ratio = 60;

  if (typeof motion === 'function') {
    motion = motion(props);
  } else if (!motion || typeof motion !== 'object') {
    motion = {};
  }

  if (mode === 'vertical') {
    return /*#__PURE__*/_react.default.createElement(_semiAnimationReact.Transition, (0, _assign.default)({}, props, {
      config: {
        tension: 612,
        friction: 32
      },
      from: {
        translateY: direction * ratio,
        opacity: 0
      },
      enter: {
        translateY: 0,
        opacity: {
          val: 1,
          duration: 200
        }
      },
      leave: {
        translateY: -1 * direction * ratio,
        opacity: {
          val: 0,
          duration: 200
        }
      }
    }, motion), typeof props.children === 'function' ? _ref => {
      let {
        translateY,
        opacity
      } = _ref;
      // delete translateX in 0 in case of zIndex problems.
      const finalState = (0, _toInteger2.default)(translateY) === 0 ? {
        opacity
      } : {
        transform: "translateY(".concat((0, _toInteger2.default)(translateY), "px)"),
        opacity
      };
      return props.children(finalState);
    } : props.children);
  }

  return /*#__PURE__*/_react.default.createElement(_semiAnimationReact.Transition, (0, _assign.default)({}, props, {
    config: {
      tension: 612,
      friction: 32
    },
    from: {
      translateX: direction * ratio,
      opacity: 0
    },
    enter: {
      translateX: 0,
      opacity: {
        val: 1,
        duration: 200
      }
    },
    leave: {
      translateX: -1 * direction * ratio,
      opacity: {
        val: 0,
        duration: 200
      }
    }
  }, motion), typeof props.children === 'function' ? _ref2 => {
    let {
      translateX,
      opacity
    } = _ref2;
    // delete translateX in 0 in case of zIndex problems.
    const finalState = (0, _toInteger2.default)(translateX) === 0 ? {
      opacity
    } : {
      transform: "translateX(".concat((0, _toInteger2.default)(translateX), "px)"),
      opacity
    };
    return props.children(finalState);
  } : props.children);
};

var _default = TabPaneTransition;
exports.default = _default;
