// @douyinfe/semi-foundation/lib/cjs/timePicker/inputFoundation.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _foundation = _interopRequireDefault(require("../base/foundation"));

var _isNullOrUndefined = _interopRequireDefault(require("../utils/isNullOrUndefined"));

class TimePickerFoundation extends _foundation.default {
  constructor(adapter) {
    super((0, _assign.default)({}, adapter));
  } // eslint-disable-next-line @typescript-eslint/no-empty-function


  init() {} // eslint-disable-next-line @typescript-eslint/no-empty-function


  destroy() {}

  handleFocus(e) {
    this.storeCursor();

    this._adapter.notifyFocus(e);
  }

  handleChange(v) {
    this.storeCursor();

    this._adapter.notifyChange(v);
  }

  handleBlur(e) {
    this.clearCursor();

    this._adapter.notifyBlur(e);
  }

  storeCursor() {
    const inputNode = this.getCache('inputNode');

    if (inputNode) {
      const {
        selectionStart: start
      } = inputNode; // const beforeStr = typeof value === 'string' ? value.substr(0, start) : null;
      // const afterStr = typeof value === 'string' ? value.substr(start, value.length - start + 1) : null;
      // console.log(start, beforeStr, afterStr);

      this.setCache('cursorIndex', start);
    }
  }

  restoreCursor() {
    const inputNode = this.getCache('inputNode');
    const cursorIndex = this.getCache('cursorIndex');

    if (inputNode && !(0, _isNullOrUndefined.default)(cursorIndex)) {
      inputNode.selectionStart = cursorIndex;
      inputNode.selectionEnd = cursorIndex;
    }
  }

  clearCursor() {
    this.setCache('cursorIndex', null);
    this.setCache('beforeStr', null);
    this.setCache('afterStr', null);
  }

}

var _default = TimePickerFoundation;
exports.default = _default;
