// @douyinfe/semi-foundation/lib/cjs/switch/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.cssClasses = void 0;

var _constants = require("../base/constants");

const cssClasses = {
  PREFIX: "".concat(_constants.BASE_CLASS_PREFIX, "-switch"),
  FOCUS: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-focus"),
  LARGE: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-large"),
  SMALL: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-small"),
  CHECKED: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-checked"),
  DISABLED: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-disabled"),
  ACTIVE: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-active"),
  KNOB: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-knob"),
  NATIVE_CONTROL: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-native-control"),
  CHECKED_TEXT: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-checked-text"),
  UNCHECKED_TEXT: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-unchecked-text"),
  LOADING_SPIN: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-loading-spin"),
  LOADING: "".concat(_constants.BASE_CLASS_PREFIX, "-switch-loading")
};
exports.cssClasses = cssClasses;
const strings = {
  SIZE_MAP: ['default', 'small', 'large']
};
exports.strings = strings;
