// @douyinfe/semi-ui/lib/cjs/carousel/CarouselArrow.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/carousel/constants");

var _semiIcons = require("@douyinfe/semi-icons");

/* eslint-disable jsx-a11y/click-events-have-key-events */

/* eslint-disable jsx-a11y/no-static-element-interactions */
class CarouselArrow extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    this.renderLeftIcon = () => {
      return (0, _get2.default)(this.props, 'arrowProps.leftArrow.children', /*#__PURE__*/_react.default.createElement(_semiIcons.IconChevronLeft, {
        "aria-label": "Previous index",
        size: "inherit"
      }));
    };

    this.renderRightIcon = () => {
      return (0, _get2.default)(this.props, 'arrowProps.rightArrow.children', /*#__PURE__*/_react.default.createElement(_semiIcons.IconChevronRight, {
        "aria-label": "Next index",
        size: "inherit"
      }));
    };
  }

  render() {
    var _context, _context2, _context3;

    const {
      type,
      theme,
      prev,
      next
    } = this.props;
    const classNames = (0, _classnames.default)({
      [_constants.cssClasses.CAROUSEL_ARROW]: true,
      [(0, _concat.default)(_context = "".concat(_constants.cssClasses.CAROUSEL_ARROW, "-")).call(_context, theme)]: theme,
      ["".concat(_constants.cssClasses.CAROUSEL_ARROW, "-hover")]: type === 'hover'
    });
    const leftClassNames = (0, _classnames.default)({
      ["".concat(_constants.cssClasses.CAROUSEL_ARROW, "-prev")]: true,
      [(0, _concat.default)(_context2 = "".concat(_constants.cssClasses.CAROUSEL_ARROW, "-")).call(_context2, theme)]: theme
    });
    const rightClassNames = (0, _classnames.default)({
      ["".concat(_constants.cssClasses.CAROUSEL_ARROW, "-next")]: true,
      [(0, _concat.default)(_context3 = "".concat(_constants.cssClasses.CAROUSEL_ARROW, "-")).call(_context3, theme)]: theme
    });
    return /*#__PURE__*/_react.default.createElement("div", {
      className: classNames
    }, /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({
      // role='button'
      className: leftClassNames,
      onClick: prev
    }, (0, _get2.default)(this.props, 'arrowProps.leftArrow.props'), {
      "x-semi-prop": "arrowProps.leftArrow.children"
    }), this.renderLeftIcon()), /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({
      // role='button'
      // tabIndex={0} 
      className: rightClassNames,
      onClick: next
    }, (0, _get2.default)(this.props, 'arrowProps.rightArrow.props'), {
      "x-semi-prop": "arrowProps.rightArrow.children"
    }), this.renderRightIcon()));
  }

}

var _default = CarouselArrow;
exports.default = _default;
