// @douyinfe/semi-ui/lib/cjs/carousel/CarouselIndicator.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/carousel/constants");

var _getDataAttr = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/utils/getDataAttr"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/* eslint-disable jsx-a11y/no-static-element-interactions */

/* eslint-disable jsx-a11y/click-events-have-key-events */


class CarouselIndicator extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    this.onIndicatorChange = activeIndex => {
      this.props.onIndicatorChange(activeIndex);
    };

    this.handleIndicatorClick = activeIndex => {
      const {
        trigger
      } = this.props;

      if (trigger === 'click') {
        this.onIndicatorChange(activeIndex);
      }
    };

    this.handleIndicatorHover = activeIndex => {
      const {
        trigger
      } = this.props;

      if (trigger === 'hover') {
        this.onIndicatorChange(activeIndex);
      }
    };
  }

  renderIndicatorContent() {
    const {
      total,
      theme,
      size,
      activeIndex
    } = this.props;
    const indicatorContent = [];

    for (let i = 0; i < total; i++) {
      var _context, _context2;

      indicatorContent.push( /*#__PURE__*/_react.default.createElement("span", {
        // role='none' 
        key: i,
        "data-index": i,
        className: (0, _classnames.default)(["".concat(_constants.cssClasses.CAROUSEL_INDICATOR, "-item")], {
          ["".concat(_constants.cssClasses.CAROUSEL_INDICATOR, "-item-active")]: i === activeIndex,
          [(0, _concat.default)(_context = "".concat(_constants.cssClasses.CAROUSEL_INDICATOR, "-item-")).call(_context, theme)]: theme,
          [(0, _concat.default)(_context2 = "".concat(_constants.cssClasses.CAROUSEL_INDICATOR, "-item-")).call(_context2, size)]: size
        }),
        onClick: () => this.handleIndicatorClick(i),
        onMouseEnter: () => this.handleIndicatorHover(i)
      }));
    }

    return indicatorContent;
  }

  render() {
    var _context3, _context4;

    const _a = this.props,
          {
      type,
      size,
      theme,
      style,
      className,
      position
    } = _a,
          restProps = __rest(_a, ["type", "size", "theme", "style", "className", "position"]);

    const classNames = (0, _classnames.default)(className, {
      [_constants.cssClasses.CAROUSEL_INDICATOR]: true,
      [(0, _concat.default)(_context3 = "".concat(_constants.cssClasses.CAROUSEL_INDICATOR, "-")).call(_context3, type)]: type,
      [(0, _concat.default)(_context4 = "".concat(_constants.cssClasses.CAROUSEL_INDICATOR, "-")).call(_context4, position)]: position
    });
    const indicatorContent = this.renderIndicatorContent();
    return /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({
      className: classNames,
      style: style
    }, (0, _getDataAttr.default)(restProps)), indicatorContent);
  }

}

CarouselIndicator.propTypes = {
  activeKey: _propTypes.default.number,
  className: _propTypes.default.string,
  position: _propTypes.default.oneOf(_constants.strings.POSITION_MAP),
  size: _propTypes.default.oneOf(_constants.strings.SIZE),
  style: _propTypes.default.object,
  theme: _propTypes.default.oneOf(_constants.strings.THEME_MAP),
  total: _propTypes.default.number,
  onIndicatorChange: _propTypes.default.func,
  type: _propTypes.default.oneOf(_constants.strings.TYPE_MAP),
  trigger: _propTypes.default.oneOf(_constants.strings.TRIGGER)
};
var _default = CarouselIndicator;
exports.default = _default;
