// @douyinfe/semi-foundation/lib/cjs/tooltip/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.numbers = exports.cssClasses = void 0;

var _constants = require("../base/constants");

const cssClasses = {
  PREFIX: "".concat(_constants.BASE_CLASS_PREFIX, "-tooltip")
};
exports.cssClasses = cssClasses;
const strings = {
  POSITION_SET: ['top', 'topLeft', 'topRight', 'left', 'leftTop', 'leftBottom', 'right', 'rightTop', 'rightBottom', 'bottom', 'bottomLeft', 'bottomRight', 'leftTopOver', 'rightTopOver', 'leftBottomOver', 'rightBottomOver'],
  TRIGGER_SET: ['hover', 'focus', 'click', 'custom'],
  STATUS_DISABLED: 'disabled',
  STATUS_LOADING: 'loading'
};
exports.strings = strings;
const numbers = {
  ARROW_BOUNDING: {
    offsetX: 0,
    offsetY: 2,
    width: 24,
    height: 7
  },
  DEFAULT_Z_INDEX: 1060,
  MOUSE_ENTER_DELAY: 50,
  MOUSE_LEAVE_DELAY: 50,
  SPACING: 8 // Values are consistent with spacing-tight in scss

};
exports.numbers = numbers;
