// @douyinfe/semi-ui/lib/cjs/collapse/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _set = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set"));

var _bind = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/bind"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _isArray = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/array/is-array"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/collapse/constants");

var _foundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/collapse/foundation"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _item = _interopRequireDefault(require("./item"));

require("@douyinfe/semi-foundation/lib/cjs/collapse/collapse.css");

var _function = require("@douyinfe/semi-foundation/lib/cjs/utils/function");

var _collapseContext = _interopRequireDefault(require("./collapse-context"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

class Collapse extends _baseComponent.default {
  constructor(props) {
    var _context;

    super(props);

    this.onChange = (activeKey, e) => {
      this.foundation.handleChange(activeKey, e);
    };

    this.foundation = new _foundation.default(this.adapter);
    const initKeys = this.foundation.initActiveKey();
    this.state = {
      activeSet: new _set.default(initKeys)
    };
    this.onChange = (0, _bind.default)(_context = this.onChange).call(_context, this);
  }

  get adapter() {
    return (0, _assign.default)((0, _assign.default)({}, super.adapter), {
      handleChange: (activeKey, e) => this.props.onChange(activeKey, e),
      addActiveKey: activeSet => this.setState({
        activeSet
      })
    });
  }

  static getDerivedStateFromProps(props, state) {
    if (props.activeKey) {
      const keys = (0, _isArray.default)(props.activeKey) ? props.activeKey : [props.activeKey];
      const newSet = new _set.default(keys);

      if (!(0, _isEqual2.default)(newSet, state.activeSet)) {
        return (0, _assign.default)((0, _assign.default)({}, state), {
          activeSet: newSet
        });
      }

      return state;
    }

    return state;
  }

  componentWillUnmount() {
    this.foundation.destroy();
  }

  render() {
    // eslint-disable-next-line max-len
    const _a = this.props,
          {
      defaultActiveKey,
      accordion,
      style,
      motion,
      className,
      keepDOM,
      expandIconPosition,
      expandIcon,
      collapseIcon,
      children
    } = _a,
          rest = __rest(_a, ["defaultActiveKey", "accordion", "style", "motion", "className", "keepDOM", "expandIconPosition", "expandIcon", "collapseIcon", "children"]);

    const clsPrefix = (0, _classnames.default)(_constants.cssClasses.PREFIX, className);
    const {
      activeSet
    } = this.state;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: clsPrefix,
      style: style
    }, /*#__PURE__*/_react.default.createElement(_collapseContext.default.Provider, {
      value: {
        activeSet,
        expandIcon,
        collapseIcon,
        keepDOM,
        expandIconPosition,
        onClick: this.onChange,
        motion
      }
    }, children));
  }

}

Collapse.Panel = _item.default;
Collapse.propTypes = {
  activeKey: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  defaultActiveKey: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.array]),
  accordion: _propTypes.default.bool,
  onChange: _propTypes.default.func,
  expandIcon: _propTypes.default.node,
  collapseIcon: _propTypes.default.node,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  keepDOM: _propTypes.default.bool,
  motion: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.func, _propTypes.default.object]),
  expandIconPosition: _propTypes.default.oneOf(_constants.strings.iconPosition)
};
Collapse.defaultProps = {
  defaultActiveKey: '',
  onChange: _function.noop,
  expandIconPosition: 'right'
};
var _default = Collapse;
exports.default = _default;
