// @douyinfe/semi-ui/lib/cjs/collapse/item.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/collapse/constants");

var _collapsible = _interopRequireDefault(require("../collapsible"));

var _collapseContext = _interopRequireDefault(require("./collapse-context"));

var _semiIcons = require("@douyinfe/semi-icons");

var _uuid = require("@douyinfe/semi-foundation/lib/cjs/utils/uuid");

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

class CollapsePanel extends _react.PureComponent {
  constructor() {
    super(...arguments);
    this.ariaID = (0, _uuid.getUuidShort)({});
  }

  renderHeader(active) {
    let expandIconEnable = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
    const {
      showArrow,
      header,
      extra
    } = this.props;
    let {
      expandIcon,
      collapseIcon
    } = this.context;
    const {
      expandIconPosition
    } = this.context;

    if (typeof expandIcon === 'undefined') {
      expandIcon = /*#__PURE__*/_react.default.createElement(_semiIcons.IconChevronDown, null);
    }

    if (typeof collapseIcon === 'undefined') {
      collapseIcon = /*#__PURE__*/_react.default.createElement(_semiIcons.IconChevronUp, null);
    }

    const icon = /*#__PURE__*/_react.default.createElement("span", {
      "aria-hidden": 'true',
      className: (0, _classnames.default)(["".concat(_constants.cssClasses.PREFIX, "-header-icon"), {
        ["".concat(_constants.cssClasses.PREFIX, "-header-iconDisabled")]: !expandIconEnable
      }])
    }, expandIconEnable ? active ? collapseIcon : expandIcon : expandIcon);

    const iconPosLeft = expandIconPosition === 'left';

    if (typeof header === 'string') {
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, showArrow && (iconPosLeft ? icon : null), /*#__PURE__*/_react.default.createElement("span", null, header), /*#__PURE__*/_react.default.createElement("span", {
        className: "".concat(_constants.cssClasses.PREFIX, "-header-right")
      }, /*#__PURE__*/_react.default.createElement("span", null, extra), showArrow && (iconPosLeft ? null : icon)));
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, showArrow && (iconPosLeft ? icon : null), header, showArrow && (iconPosLeft ? null : icon));
  }

  render() {
    const _a = this.props,
          {
      className,
      children,
      itemKey,
      reCalcKey,
      header,
      extra,
      showArrow,
      disabled
    } = _a,
          restProps = __rest(_a, ["className", "children", "itemKey", "reCalcKey", "header", "extra", "showArrow", "disabled"]);

    const {
      keepDOM,
      expandIconPosition,
      activeSet,
      onClick,
      motion
    } = this.context;
    const active = activeSet.has(itemKey);
    const itemCls = (0, _classnames.default)(className, {
      ["".concat(_constants.cssClasses.PREFIX, "-item")]: true
    });
    const headerCls = (0, _classnames.default)({
      ["".concat(_constants.cssClasses.PREFIX, "-header")]: true,
      ["".concat(_constants.cssClasses.PREFIX, "-header-disabled")]: disabled,
      ["".concat(_constants.cssClasses.PREFIX, "-header-iconLeft")]: expandIconPosition === 'left'
    });
    const contentCls = (0, _classnames.default)({
      ["".concat(_constants.cssClasses.PREFIX, "-content")]: true
    });
    return /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({
      className: itemCls
    }, restProps), /*#__PURE__*/_react.default.createElement("div", {
      role: "button",
      tabIndex: 0,
      className: headerCls,
      "aria-disabled": disabled,
      "aria-expanded": active ? 'true' : 'false',
      "aria-owns": this.ariaID,
      onClick: e => !disabled && onClick(itemKey, e)
    }, this.renderHeader(active, children !== undefined && !disabled)), children && /*#__PURE__*/_react.default.createElement(_collapsible.default, {
      isOpen: active,
      keepDOM: keepDOM,
      motion: motion,
      reCalcKey: reCalcKey
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: contentCls,
      "aria-hidden": !active,
      id: this.ariaID
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(_constants.cssClasses.PREFIX, "-content-wrapper")
    }, children))));
  }

}

exports.default = CollapsePanel;
CollapsePanel.contextType = _collapseContext.default;
CollapsePanel.propTypes = {
  itemKey: _propTypes.default.string,
  extra: _propTypes.default.node,
  header: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),
  className: _propTypes.default.string,
  reCalcKey: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  showArrow: _propTypes.default.bool,
  disabled: _propTypes.default.bool
};
CollapsePanel.defaultProps = {
  showArrow: true,
  disabled: false
};
