// @douyinfe/semi-ui/lib/es/table/TableHeaderRow.js
import _findIndex from "lodash/findIndex";
import _omit from "lodash/omit";
import _set from "lodash/set";
import _map from "lodash/map";
import _noop from "lodash/noop";
import _get from "lodash/get";
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _Set from "@babel/runtime-corejs3/core-js-stable/set";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && _indexOfInstanceProperty(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _Object$getOwnPropertySymbols === "function") for (var i = 0, p = _Object$getOwnPropertySymbols(s); i < p.length; i++) {
    if (_indexOfInstanceProperty(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/* eslint-disable eqeqeq */


import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/table/constants';
import { arrayAdd, isFirstFixedRight, isLastLeftFixed, isFixedLeft, isFixedRight, sliceColumnsByLevel } from '@douyinfe/semi-foundation/lib/es/table/utils';
import BaseComponent from '../_base/baseComponent';
import TableContext from './table-context';
export default class TableHeaderRow extends BaseComponent {
  constructor(props) {
    super(props);

    this.cacheRef = node => {
      this.headerNode = node;

      if (node && this.context.setHeadWidths) {
        const {
          prefixCls,
          row,
          index
        } = this.props;
        const cellSelector = ".".concat(prefixCls, "-row-head");
        const heads = node && node.querySelectorAll && node.querySelectorAll(cellSelector);
        this.context.setHeadWidths(_map(heads, (head, headIndex) => {
          let configWidth = _get(row, [headIndex, 'column', 'width']);

          const key = _get(row, [headIndex, 'column', 'key']);

          if (typeof configWidth !== 'number') {
            configWidth = head && head.getBoundingClientRect().width || 0;
          }

          return {
            width: configWidth,
            key
          };
        }), index);
      }
    };

    this.headerNode = null;
  }

  get adapter() {
    return _Object$assign({}, super.adapter);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.columns !== this.props.columns && this.headerNode) {
      this.cacheRef(this.headerNode);
    }
  }

  render() {
    const {
      components,
      row,
      prefixCls,
      onHeaderRow,
      index,
      style,
      columns
    } = this.props;
    const {
      getCellWidths
    } = this.context;
    const slicedColumns = sliceColumnsByLevel(columns, index);
    const headWidths = getCellWidths(slicedColumns);

    const HeaderRow = _get(components, 'header.row', 'tr');

    const HeaderCell = _get(components, 'header.cell', 'th');

    const rowProps = onHeaderRow(columns, index) || {};

    _set(rowProps, 'className', classnames(_get(rowProps, 'className'), "".concat(prefixCls, "-row")));

    const cells = _map(row, (cell, cellIndex) => {
      const {
        column
      } = cell,
            cellProps = __rest(cell, ["column"]);

      const customProps = typeof column.onHeaderCell === 'function' ? column.onHeaderCell(column, cellIndex, index) : {};

      let cellStyle = _Object$assign({}, customProps.style);

      if (column.align) {
        var _context;

        cellStyle = _Object$assign(_Object$assign({}, cellStyle), {
          textAlign: column.align
        });
        customProps.className = classnames(customProps.className, column.className, {
          [_concatInstanceProperty(_context = "".concat(prefixCls, "-align-")).call(_context, column.align)]: Boolean(column.align)
        });
      }

      customProps.className = classnames("".concat(prefixCls, "-row-head"), column.className, customProps.className, // `${prefixCls}-fixed-columns`,
      {
        ["".concat(prefixCls, "-cell-fixed-left")]: isFixedLeft(column),
        ["".concat(prefixCls, "-cell-fixed-left-last")]: isLastLeftFixed(slicedColumns, column),
        ["".concat(prefixCls, "-cell-fixed-right")]: isFixedRight(column),
        ["".concat(prefixCls, "-cell-fixed-right-first")]: isFirstFixedRight(slicedColumns, column)
      });

      if (headWidths.length && slicedColumns.length) {
        const indexOfSlicedColumns = _findIndex(slicedColumns, item => item && item.key != null && item.key === column.key);

        if (indexOfSlicedColumns > -1) {
          if (isFixedLeft(column)) {
            cellStyle = _Object$assign(_Object$assign({}, cellStyle), {
              position: 'sticky',
              left: arrayAdd(headWidths, 0, indexOfSlicedColumns)
            });
          } else if (isFixedRight(column)) {
            cellStyle = _Object$assign(_Object$assign({}, cellStyle), {
              position: 'sticky',
              right: arrayAdd(headWidths, indexOfSlicedColumns + 1)
            });
          }
        }
      }

      const props = _omit(_Object$assign(_Object$assign({}, cellProps), customProps), ['colStart', 'colEnd', 'hasSubColumns', 'parents', 'level']);

      const {
        rowSpan,
        colSpan
      } = props;

      if (rowSpan === 0 || colSpan === 0) {
        return null;
      }

      return /*#__PURE__*/React.createElement(HeaderCell, _Object$assign({
        role: "columnheader",
        "aria-colindex": cellIndex + 1
      }, props, {
        style: cellStyle,
        key: column.key || column.dataIndex || cellIndex
      }));
    });

    return /*#__PURE__*/React.createElement(HeaderRow, _Object$assign({
      role: "row",
      "aria-rowindex": index + 1
    }, rowProps, {
      style: style,
      ref: this.cacheRef
    }), cells);
  }

}
TableHeaderRow.contextType = TableContext;
TableHeaderRow.propTypes = {
  components: PropTypes.object,
  row: PropTypes.array,
  prefixCls: PropTypes.string,
  onHeaderRow: PropTypes.func,
  index: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  style: PropTypes.object,
  columns: PropTypes.array,
  fixed: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  selectedRowKeysSet: PropTypes.instanceOf(_Set).isRequired
};
TableHeaderRow.defaultProps = {
  onHeaderRow: _noop,
  prefixCls: cssClasses.PREFIX,
  columns: [],
  components: {
    header: {
      wrapper: 'thead',
      row: 'tr',
      cell: 'th'
    }
  }
};
