// @douyinfe/semi-ui/lib/es/skeleton/item.js
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && _indexOfInstanceProperty(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _Object$getOwnPropertySymbols === "function") for (var i = 0, p = _Object$getOwnPropertySymbols(s); i < p.length; i++) {
    if (_indexOfInstanceProperty(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import React, { PureComponent } from 'react';
import cls from 'classnames';
import PropTypes from 'prop-types';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/skeleton/constants';
import { strings } from '@douyinfe/semi-foundation/lib/es/avatar/constants';
import '@douyinfe/semi-foundation/lib/es/skeleton/skeleton.css';
const sizeSet = strings.SIZE;

const generator = type => BasicComponent => props => /*#__PURE__*/React.createElement(BasicComponent, _Object$assign({
  type: type
}, props));

class Generic extends PureComponent {
  render() {
    var _context, _context2, _context3;

    const _a = this.props,
          {
      prefixCls,
      className,
      type,
      size
    } = _a,
          others = __rest(_a, ["prefixCls", "className", "type", "size"]);

    const classString = cls(className, _concatInstanceProperty(_context = "".concat(prefixCls, "-")).call(_context, type), {
      [_concatInstanceProperty(_context2 = _concatInstanceProperty(_context3 = "".concat(prefixCls, "-")).call(_context3, type, "-")).call(_context2, size)]: type.toUpperCase() === 'AVATAR'
    });
    return /*#__PURE__*/React.createElement('div', _Object$assign({
      className: classString
    }, others));
  }

}

Generic.propTypes = {
  type: PropTypes.string,
  prefixCls: PropTypes.string,
  style: PropTypes.object,
  className: PropTypes.string,
  size: PropTypes.oneOf(sizeSet)
};
Generic.defaultProps = {
  prefixCls: cssClasses.PREFIX,
  size: 'medium'
};
export const Avatar = generator('avatar')(Generic);
export const Image = generator('image')(Generic);
export const Title = generator('title')(Generic);
export const Button = generator('button')(Generic);
export class Paragraph extends PureComponent {
  render() {
    var _context4;

    const {
      prefixCls,
      className,
      style,
      rows
    } = this.props;
    const classString = cls(className, "".concat(prefixCls, "-paragraph"));
    return /*#__PURE__*/React.createElement("ul", {
      className: classString,
      style: style
    }, _mapInstanceProperty(_context4 = [...Array(rows)]).call(_context4, (e, i) => /*#__PURE__*/React.createElement("li", {
      key: i
    })));
  }

}
Paragraph.propTypes = {
  rows: PropTypes.number,
  prefixCls: PropTypes.string,
  style: PropTypes.object,
  className: PropTypes.string
};
Paragraph.defaultProps = {
  prefixCls: cssClasses.PREFIX,
  rows: 4
};
