// @douyinfe/semi-ui/lib/cjs/radio/radio.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _radioFoundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/radio/radioFoundation"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/radio/constants");

var _uuid = require("@douyinfe/semi-foundation/lib/cjs/utils/uuid");

require("@douyinfe/semi-foundation/lib/cjs/radio/radio.css");

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _radioInner = _interopRequireDefault(require("./radioInner"));

var _context4 = _interopRequireDefault(require("./context"));

/* eslint-disable prefer-destructuring */
class Radio extends _baseComponent.default {
  constructor(props) {
    super(props);

    this.onChange = e => {
      const {
        onChange
      } = this.props;

      if (this.isInGroup()) {
        const {
          radioGroup
        } = this.context;
        radioGroup.onChange && radioGroup.onChange(e);
      }

      onChange && onChange(e);
    };

    this.handleMouseEnter = e => {
      this.props.onMouseEnter(e);
      this.foundation.setHover(true);
    };

    this.handleMouseLeave = e => {
      this.props.onMouseLeave(e);
      this.foundation.setHover(false);
    };

    this.handleFocusVisible = event => {
      this.foundation.handleFocusVisible(event);
    };

    this.handleBlur = event => {
      this.foundation.handleBlur();
    };

    this.state = {
      hover: false,
      addonId: props.addonId,
      extraId: props.extraId
    };
    this.foundation = new _radioFoundation.default(this.adapter);
    this.radioEntity = null;
  }

  get adapter() {
    return (0, _assign.default)((0, _assign.default)({}, super.adapter), {
      setHover: hover => {
        this.setState({
          hover
        });
      },
      setAddonId: () => {
        this.setState({
          addonId: (0, _uuid.getUuidShort)({
            prefix: 'addon'
          })
        });
      },
      setExtraId: () => {
        this.setState({
          extraId: (0, _uuid.getUuidShort)({
            prefix: 'extra'
          })
        });
      },
      setFocusVisible: focusVisible => {
        this.setState({
          focusVisible
        });
      }
    });
  }

  isInGroup() {
    // eslint-disable-next-line react/destructuring-assignment
    return this.context && this.context.radioGroup;
  }

  focus() {
    this.radioEntity.focus();
  }

  blur() {
    this.radioEntity.blur();
  }

  render() {
    var _context, _context2, _context3;

    const {
      addonClassName,
      addonStyle,
      checked,
      disabled,
      style,
      className,
      prefixCls,
      displayMode,
      children,
      extra,
      mode,
      type,
      value: propValue,
      name
    } = this.props;
    let realChecked, isDisabled, realMode, isButtonRadioGroup, isCardRadioGroup, isPureCardRadioGroup, isButtonRadioComponent, buttonSize, realPrefixCls;
    const {
      hover: isHover,
      addonId,
      extraId,
      focusVisible
    } = this.state;
    let props = {};

    if (this.isInGroup()) {
      realChecked = this.context.radioGroup.value === propValue;
      isDisabled = disabled || this.context.radioGroup.disabled;
      realMode = this.context.mode;
      isButtonRadioGroup = this.context.radioGroup.isButtonRadio;
      isCardRadioGroup = this.context.radioGroup.isCardRadio;
      isPureCardRadioGroup = this.context.radioGroup.isPureCardRadio;
      buttonSize = this.context.radioGroup.buttonSize;
      realPrefixCls = prefixCls || this.context.radioGroup.prefixCls;
      props = {
        checked: realChecked,
        disabled: isDisabled
      };
    } else {
      realChecked = checked;
      isDisabled = disabled;
      realMode = mode;
      isButtonRadioComponent = type === 'button';
      realPrefixCls = prefixCls;
    }

    const isButtonRadio = typeof isButtonRadioGroup === 'undefined' ? isButtonRadioComponent : isButtonRadioGroup;
    const prefix = realPrefixCls || _constants.radioClasses.PREFIX;
    const focusOuter = isCardRadioGroup || isPureCardRadioGroup || isButtonRadio;
    const wrapper = (0, _classnames.default)(prefix, {
      ["".concat(prefix, "-disabled")]: isDisabled,
      ["".concat(prefix, "-checked")]: realChecked,
      [(0, _concat.default)(_context = "".concat(prefix, "-")).call(_context, displayMode)]: Boolean(displayMode),
      ["".concat(prefix, "-buttonRadioComponent")]: isButtonRadioComponent,
      ["".concat(prefix, "-buttonRadioGroup")]: isButtonRadioGroup,
      [(0, _concat.default)(_context2 = "".concat(prefix, "-buttonRadioGroup-")).call(_context2, buttonSize)]: isButtonRadioGroup && buttonSize,
      ["".concat(prefix, "-cardRadioGroup")]: isCardRadioGroup,
      ["".concat(prefix, "-cardRadioGroup_disabled")]: isDisabled && isCardRadioGroup,
      ["".concat(prefix, "-cardRadioGroup_checked")]: isCardRadioGroup && realChecked && !isDisabled,
      ["".concat(prefix, "-cardRadioGroup_checked_disabled")]: isCardRadioGroup && realChecked && isDisabled,
      ["".concat(prefix, "-cardRadioGroup_hover")]: isCardRadioGroup && !realChecked && isHover && !isDisabled,
      [className]: Boolean(className),
      ["".concat(prefix, "-focus")]: focusVisible && (isCardRadioGroup || isPureCardRadioGroup)
    });
    const groupName = this.isInGroup() && this.context.radioGroup.name;
    const addonCls = (0, _classnames.default)({
      ["".concat(prefix, "-addon")]: !isButtonRadio,
      ["".concat(prefix, "-addon-buttonRadio")]: isButtonRadio,
      ["".concat(prefix, "-addon-buttonRadio-checked")]: isButtonRadio && realChecked,
      ["".concat(prefix, "-addon-buttonRadio-disabled")]: isButtonRadio && isDisabled,
      ["".concat(prefix, "-addon-buttonRadio-hover")]: isButtonRadio && !realChecked && !isDisabled && isHover,
      [(0, _concat.default)(_context3 = "".concat(prefix, "-addon-buttonRadio-")).call(_context3, buttonSize)]: isButtonRadio && buttonSize,
      ["".concat(prefix, "-focus")]: focusVisible && isButtonRadio
    }, addonClassName);

    const renderContent = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children ? /*#__PURE__*/_react.default.createElement("span", {
      className: addonCls,
      style: addonStyle,
      id: addonId,
      "x-semi-prop": "children"
    }, children) : null, extra && !isButtonRadio ? /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefix, "-extra"),
      id: extraId,
      "x-semi-prop": "extra"
    }, extra) : null);

    return /*#__PURE__*/_react.default.createElement("label", {
      style: style,
      className: wrapper,
      onMouseEnter: this.handleMouseEnter,
      onMouseLeave: this.handleMouseLeave
    }, /*#__PURE__*/_react.default.createElement(_radioInner.default, (0, _assign.default)({}, this.props, props, {
      mode: realMode,
      name: name !== null && name !== void 0 ? name : groupName,
      isButtonRadio: isButtonRadio,
      isPureCardRadioGroup: isPureCardRadioGroup,
      onChange: this.onChange,
      ref: ref => {
        this.radioEntity = ref;
      },
      addonId: children && addonId,
      extraId: extra && extraId,
      focusInner: focusVisible && !focusOuter,
      onInputFocus: this.handleFocusVisible,
      onInputBlur: this.handleBlur
    })), isCardRadioGroup ? /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefix, "-isCardRadioGroup_content")
    }, renderContent()) : renderContent());
  }

}

Radio.contextType = _context4.default;
Radio.propTypes = {
  autoFocus: _propTypes.default.bool,
  checked: _propTypes.default.bool,
  defaultChecked: _propTypes.default.bool,
  value: _propTypes.default.any,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  prefixCls: _propTypes.default.string,
  displayMode: _propTypes.default.oneOf(['vertical', '']),
  onChange: _propTypes.default.func,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func,
  mode: _propTypes.default.oneOf(_constants.strings.MODE),
  extra: _propTypes.default.node,
  addonStyle: _propTypes.default.object,
  addonClassName: _propTypes.default.string,
  type: _propTypes.default.oneOf([_constants.strings.TYPE_DEFAULT, _constants.strings.TYPE_BUTTON, _constants.strings.TYPE_CARD, _constants.strings.TYPE_PURECARD]),
  'aria-label': _propTypes.default.string,
  preventScroll: _propTypes.default.bool
};
Radio.defaultProps = {
  autoFocus: false,
  defaultChecked: false,
  value: undefined,
  style: undefined,
  onMouseEnter: _noop2.default,
  onMouseLeave: _noop2.default,
  mode: '',
  type: 'default'
};
var _default = Radio;
exports.default = _default;
