// @douyinfe/semi-ui/lib/cjs/modal/useModal/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = useModal;

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _react = _interopRequireDefault(require("react"));

var _HookModal = _interopRequireDefault(require("./HookModal"));

var _confirm = require("../confirm");

let uuid = 0;

function usePatchElement() {
  const [elements, setElements] = _react.default.useState([]);

  function patchElement(element) {
    setElements(originElements => [...originElements, element]);
    return () => {
      setElements(originElements => (0, _filter.default)(originElements).call(originElements, ele => ele !== element));
    };
  }

  return [elements, patchElement];
}

function useModal() {
  const [elements, patchElement] = usePatchElement(); // eslint-disable-next-line max-len

  function getConfirmFunc(withFunc) {
    return function hookConfirm(config) {
      uuid += 1;

      const modalRef = /*#__PURE__*/_react.default.createRef(); // eslint-disable-next-line prefer-const


      let closeFunc;

      const modal = /*#__PURE__*/_react.default.createElement(_HookModal.default, {
        key: "semi-modal-".concat(uuid),
        config: withFunc(config),
        ref: modalRef,
        afterClose: () => {
          closeFunc();
        }
      });

      closeFunc = patchElement(modal);
      return {
        destroy: () => {
          if (modalRef.current) {
            modalRef.current.destroy();
          }
        },
        update: newConfig => {
          if (modalRef.current) {
            modalRef.current.update(newConfig);
          }
        }
      };
    };
  }

  return [{
    info: getConfirmFunc(_confirm.withInfo),
    success: getConfirmFunc(_confirm.withSuccess),
    error: getConfirmFunc(_confirm.withError),
    warning: getConfirmFunc(_confirm.withWarning),
    confirm: getConfirmFunc(_confirm.withConfirm)
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, elements)];
}
