// @douyinfe/semi-animation-styled/lib/cjs/src/constants/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.zoomingExits = exports.zoomingEntrances = exports.specials = exports.slidingExits = exports.slidingEntrances = exports.rotatingExits = exports.rotatingEntrances = exports.lightspeed = exports.flippers = exports.fadingExits = exports.fadingEntrances = exports.bouncingExits = exports.bouncingEntrances = exports.attentionSeekers = void 0;
const attentionSeekers = ['bounce', 'flash', 'headShake', 'heartBeat', 'jello', 'pulse', 'rubberBand', 'shake', 'swing', 'tada', 'wobble'];
exports.attentionSeekers = attentionSeekers;
const bouncingEntrances = ['bounceIn', 'bounceInDown', 'bounceInLeft', 'bounceInRight', 'bounceInUp'];
exports.bouncingEntrances = bouncingEntrances;
const bouncingExits = ['bounceOut', 'bounceOutDown', 'bounceOutLeft', 'bounceOutRight', 'bounceOutUp'];
exports.bouncingExits = bouncingExits;
const fadingEntrances = ['fadeIn', 'fadeInDown', 'fadeInDownBig', 'fadeInLeft', 'fadeInLeftBig', 'fadeInRight', 'fadeInRightBig', 'fadeInUp', 'fadeInUpBig'];
exports.fadingEntrances = fadingEntrances;
const fadingExits = ['fadeOut', 'fadeOutDown', 'fadeOutDownBig', 'fadeOutLeft', 'fadeOutLeftBig', 'fadeOutRight', 'fadeOutRightBig', 'fadeOutUp', 'fadeOutUpBig'];
exports.fadingExits = fadingExits;
const flippers = ['flip', 'flipInX', 'flipInY', 'flipOutX', 'flipOutY'];
exports.flippers = flippers;
const lightspeed = ['lightSpeedIn', 'lightSpeedOut'];
exports.lightspeed = lightspeed;
const rotatingEntrances = ['rotateIn', 'rotateInDownLeft', 'rotateInDownRight', 'rotateInUpLeft', 'rotateInUpRight'];
exports.rotatingEntrances = rotatingEntrances;
const rotatingExits = ['rotateOut', 'rotateOutDownLeft', 'rotateOutDownRight', 'rotateOutUpLeft', 'rotateOutUpRight'];
exports.rotatingExits = rotatingExits;
const slidingEntrances = ['slideInDown', 'slideInLeft', 'slideInRight', 'slideInUp'];
exports.slidingEntrances = slidingEntrances;
const slidingExits = ['slideOutDown', 'slideOutLeft', 'slideOutRight', 'slideOutUp'];
exports.slidingExits = slidingExits;
const specials = ['hinge', 'jackInTheBox', 'rollIn', 'rollOut'];
exports.specials = specials;
const zoomingEntrances = ['zoomIn', 'zoomInDown', 'zoomInLeft', 'zoomInRight', 'zoomInUp'];
exports.zoomingEntrances = zoomingEntrances;
const zoomingExits = ['zoomOut', 'zoomOutDown', 'zoomOutLeft', 'zoomOutRight', 'zoomOutUp'];
exports.zoomingExits = zoomingExits;
