// @douyinfe/semi-ui/lib/cjs/tag/group.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _slice = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/slice"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/tag/constants");

var _index = _interopRequireDefault(require("./index"));

var _index2 = _interopRequireDefault(require("../popover/index"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const prefixCls = _constants.cssClasses.PREFIX;
const tagSize = _constants.strings.TAG_SIZE;
const avatarShapeSet = _constants.strings.AVATAR_SHAPE;

class TagGroup extends _react.PureComponent {
  renderNTag(n, restTags) {
    const {
      size,
      showPopover,
      popoverProps
    } = this.props;

    let nTag = /*#__PURE__*/_react.default.createElement(_index.default, {
      closable: false,
      size: size,
      color: "grey",
      style: {
        backgroundColor: 'transparent'
      },
      key: "_+n"
    }, "+", n);

    if (showPopover) {
      nTag = /*#__PURE__*/_react.default.createElement(_index2.default, (0, _assign.default)({
        showArrow: true,
        content: restTags,
        trigger: "hover",
        position: "top",
        autoAdjustOverflow: true,
        className: "".concat(prefixCls, "-rest-group-popover")
      }, popoverProps, {
        key: "_+n_Popover"
      }), nTag);
    }

    return nTag;
  }

  renderMergeTags(tags) {
    const {
      maxTagCount,
      tagList,
      restCount
    } = this.props;
    const n = restCount ? restCount : tagList.length - maxTagCount;
    let renderTags = tags;
    const normalTags = (0, _slice.default)(tags).call(tags, 0, maxTagCount);
    const restTags = (0, _slice.default)(tags).call(tags, maxTagCount);
    let nTag = null;

    if (n > 0) {
      nTag = this.renderNTag(n, restTags);
      normalTags.push(nTag);
      renderTags = normalTags;
    }

    return renderTags;
  }

  renderAllTags() {
    const {
      tagList,
      size,
      mode,
      avatarShape
    } = this.props;
    const renderTags = (0, _map.default)(tagList).call(tagList, (tag, index) => {
      if (mode === 'custom') {
        return tag;
      }

      if (!tag.size) {
        tag.size = size;
      }

      if (!tag.avatarShape) {
        tag.avatarShape = avatarShape;
      }

      return /*#__PURE__*/_react.default.createElement(_index.default, (0, _assign.default)({
        key: "".concat(index, "-tag")
      }, tag));
    });
    return renderTags;
  }

  render() {
    const {
      style,
      className,
      maxTagCount,
      size
    } = this.props;
    const groupCls = (0, _classnames.default)({
      ["".concat(prefixCls, "-group")]: true,
      ["".concat(prefixCls, "-group-max")]: maxTagCount,
      ["".concat(prefixCls, "-group-small")]: size === 'small',
      ["".concat(prefixCls, "-group-large")]: size === 'large'
    }, className);
    const tags = this.renderAllTags();
    const tagContents = typeof maxTagCount === 'undefined' ? tags : this.renderMergeTags(tags);
    return /*#__PURE__*/_react.default.createElement("div", {
      style: style,
      className: groupCls
    }, tagContents);
  }

}

exports.default = TagGroup;
TagGroup.defaultProps = {
  style: {},
  className: '',
  size: tagSize[0],
  avatarShape: 'square'
};
TagGroup.propTypes = {
  children: _propTypes.default.node,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  maxTagCount: _propTypes.default.number,
  restCount: _propTypes.default.number,
  tagList: _propTypes.default.array,
  size: _propTypes.default.oneOf(tagSize),
  mode: _propTypes.default.string,
  showPopover: _propTypes.default.bool,
  popoverProps: _propTypes.default.object,
  avatarShape: _propTypes.default.oneOf(avatarShapeSet)
};
