// @douyinfe/semi-ui/lib/cjs/autoComplete/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/map"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _from = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/array/from"));

var _values = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/values"));

var _map2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/autoComplete/constants");

var _foundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/autoComplete/foundation"));

var _constants2 = require("@douyinfe/semi-foundation/lib/cjs/popover/constants");

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _spin = _interopRequireDefault(require("../spin"));

var _popover = _interopRequireDefault(require("../popover"));

var _input = _interopRequireDefault(require("../input"));

var _trigger = _interopRequireDefault(require("../trigger"));

var _option = _interopRequireDefault(require("../select/option"));

var _warning = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/utils/warning"));

require("@douyinfe/semi-foundation/lib/cjs/autoComplete/autoComplete.css");

/* eslint-disable @typescript-eslint/ban-types, max-len */
const prefixCls = _constants.cssClasses.PREFIX;
const sizeSet = _constants.strings.SIZE;
const positionSet = _constants.strings.POSITION;
const statusSet = _constants.strings.STATUS;

class AutoComplete extends _baseComponent.default {
  constructor(props) {
    super(props);

    this.onSelect = (option, optionIndex, e) => {
      this.foundation.handleSelect(option, optionIndex);
    };

    this.onSearch = value => {
      this.foundation.handleSearch(value);
    };

    this.onBlur = e => this.foundation.handleBlur(e);

    this.onFocus = e => this.foundation.handleFocus(e);

    this.onInputClear = () => this.foundation.handleClear();

    this.handleInputClick = e => this.foundation.handleInputClick(e);

    this.foundation = new _foundation.default(this.adapter);
    const initRePosKey = 1;
    this.state = {
      dropdownMinWidth: null,
      inputValue: '',
      // option list
      options: [],
      // popover visible
      visible: false,
      // current focus option index
      focusIndex: props.defaultActiveFirstOption ? 0 : -1,
      // current selected options
      selection: new _map.default(),
      rePosKey: initRePosKey
    };
    this.triggerRef = /*#__PURE__*/_react.default.createRef();
    this.optionsRef = /*#__PURE__*/_react.default.createRef();
    this.clickOutsideHandler = null;
    (0, _warning.default)('triggerRender' in this.props && typeof this.props.triggerRender === 'function', "[Semi AutoComplete]\n            - If you are using the following props: 'suffix', 'prefix', 'showClear', 'validateStatus', and 'size',\n            please notice that they will be removed in the next major version.\n            Please use 'componentProps' to retrieve these props instead.\n            - If you are using 'onBlur', 'onFocus', please try to avoid using them and look for changes in the future.");
  }

  get adapter() {
    const keyboardAdapter = {
      registerKeyDown: cb => {
        const keyboardEventSet = {
          onKeyDown: cb
        };
        this.setState({
          keyboardEventSet
        });
      },
      unregisterKeyDown: cb => {
        this.setState({
          keyboardEventSet: {}
        });
      },
      updateFocusIndex: focusIndex => {
        this.setState({
          focusIndex
        });
      }
    };
    return (0, _assign.default)((0, _assign.default)((0, _assign.default)({}, super.adapter), keyboardAdapter), {
      getTriggerWidth: () => {
        const el = this.triggerRef.current;
        return el && el.getBoundingClientRect().width;
      },
      setOptionWrapperWidth: width => {
        this.setState({
          dropdownMinWidth: width
        });
      },
      updateInputValue: inputValue => {
        this.setState({
          inputValue
        });
      },
      toggleListVisible: isShow => {
        this.setState({
          visible: isShow
        });
      },
      updateOptionList: optionList => {
        this.setState({
          options: optionList
        });
      },
      updateSelection: selection => {
        this.setState({
          selection
        });
      },
      notifySearch: inputValue => {
        this.props.onSearch(inputValue);
      },
      notifyChange: value => {
        this.props.onChange(value);
      },
      notifySelect: option => {
        this.props.onSelect(option);
      },
      notifyDropdownVisibleChange: isVisible => {
        this.props.onDropdownVisibleChange(isVisible);
      },
      notifyClear: () => {
        this.props.onClear();
      },
      notifyFocus: event => {
        this.props.onFocus(event);
      },
      notifyBlur: event => {
        this.props.onBlur(event);
      },
      rePositionDropdown: () => {
        let {
          rePosKey
        } = this.state;
        rePosKey = rePosKey + 1;
        this.setState({
          rePosKey
        });
      }
    });
  }

  componentDidMount() {
    this.foundation.init();
  }

  componentWillUnmount() {
    this.foundation.destroy();
  }

  componentDidUpdate(prevProps, prevState) {
    if (!(0, _isEqual2.default)(this.props.data, prevProps.data)) {
      this.foundation.handleDataChange(this.props.data);
    }

    if (this.props.value !== prevProps.value) {
      this.foundation.handleValueChange(this.props.value);
    }
  }

  renderInput() {
    const {
      size,
      prefix,
      insetLabel,
      insetLabelId,
      suffix,
      placeholder,
      style,
      className,
      showClear,
      disabled,
      triggerRender,
      validateStatus,
      autoFocus,
      value,
      id
    } = this.props;
    const {
      inputValue,
      keyboardEventSet,
      selection
    } = this.state;
    const useCustomTrigger = typeof triggerRender === 'function';
    const outerProps = (0, _assign.default)({
      style,
      className: useCustomTrigger ? (0, _classnames.default)(className) : (0, _classnames.default)({
        [prefixCls]: true,
        ["".concat(prefixCls, "-disabled")]: disabled
      }, className),
      onClick: this.handleInputClick,
      ref: this.triggerRef,
      id
    }, keyboardEventSet);
    const innerProps = {
      disabled,
      placeholder,
      autofocus: autoFocus,
      onChange: this.onSearch,
      onClear: this.onInputClear,
      'aria-label': this.props['aria-label'],
      'aria-labelledby': this.props['aria-labelledby'],
      'aria-invalid': this.props['aria-invalid'],
      'aria-errormessage': this.props['aria-errormessage'],
      'aria-describedby': this.props['aria-describedby'],
      'aria-required': this.props['aria-required'],
      // TODO: remove in next major version
      suffix,
      prefix: prefix || insetLabel,
      insetLabelId,
      showClear,
      validateStatus,
      size,
      onBlur: this.onBlur,
      onFocus: this.onFocus
    };
    return /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({}, outerProps), typeof triggerRender === 'function' ? /*#__PURE__*/_react.default.createElement(_trigger.default, (0, _assign.default)({}, innerProps, {
      inputValue: typeof value !== 'undefined' ? value : inputValue,
      value: (0, _from.default)((0, _values.default)(selection).call(selection)),
      triggerRender: triggerRender,
      componentName: "AutoComplete",
      componentProps: (0, _assign.default)({}, this.props)
    })) : /*#__PURE__*/_react.default.createElement(_input.default, (0, _assign.default)({}, innerProps, {
      value: typeof value !== 'undefined' ? value : inputValue
    })));
  }

  renderLoading() {
    const loadingWrapperCls = "".concat(prefixCls, "-loading-wrapper");
    return /*#__PURE__*/_react.default.createElement("div", {
      className: loadingWrapperCls
    }, /*#__PURE__*/_react.default.createElement(_spin.default, null));
  }

  renderOption(option, optionIndex) {
    const {
      focusIndex
    } = this.state;
    const isFocused = optionIndex === focusIndex;
    return /*#__PURE__*/_react.default.createElement(_option.default, (0, _assign.default)({
      showTick: false,
      onSelect: (v, e) => this.onSelect(v, optionIndex, e),
      // selected={selection.has(option.label)}
      focused: isFocused,
      onMouseEnter: () => this.foundation.handleOptionMouseEnter(optionIndex),
      key: option.key || option.label + option.value + optionIndex
    }, option), option.label);
  }

  renderOptionList() {
    const {
      maxHeight,
      dropdownStyle,
      dropdownClassName,
      loading,
      emptyContent
    } = this.props;
    const {
      options,
      dropdownMinWidth
    } = this.state;
    const listCls = (0, _classnames.default)({
      ["".concat(prefixCls, "-option-list")]: true
    }, dropdownClassName);
    let optionsNode;

    if (options.length === 0) {
      optionsNode = emptyContent;
    } else {
      var _context;

      optionsNode = (0, _map2.default)(_context = (0, _filter.default)(options).call(options, option => option.show)).call(_context, (option, i) => this.renderOption(option, i));
    }

    const style = (0, _assign.default)({
      maxHeight: maxHeight,
      minWidth: dropdownMinWidth
    }, dropdownStyle);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: listCls,
      role: "listbox",
      style: style
    }, !loading ? optionsNode : this.renderLoading());
  }

  render() {
    const {
      position,
      motion,
      zIndex,
      mouseEnterDelay,
      mouseLeaveDelay,
      autoAdjustOverflow,
      stopPropagation,
      getPopupContainer
    } = this.props;
    const {
      visible,
      rePosKey
    } = this.state;
    const input = this.renderInput();
    const optionList = this.renderOptionList();
    return /*#__PURE__*/_react.default.createElement(_popover.default, {
      mouseEnterDelay: mouseEnterDelay,
      mouseLeaveDelay: mouseLeaveDelay,
      autoAdjustOverflow: autoAdjustOverflow,
      trigger: "custom",
      motion: motion,
      visible: visible,
      content: optionList,
      position: position,
      ref: this.optionsRef,
      // TransformFromCenter TODO: need to confirm
      zIndex: zIndex,
      stopPropagation: stopPropagation,
      getPopupContainer: getPopupContainer,
      rePosKey: rePosKey
    }, input);
  }

}

AutoComplete.propTypes = {
  'aria-label': _propTypes.default.string,
  'aria-labelledby': _propTypes.default.string,
  'aria-invalid': _propTypes.default.bool,
  'aria-errormessage': _propTypes.default.string,
  'aria-describedby': _propTypes.default.string,
  'aria-required': _propTypes.default.bool,
  autoFocus: _propTypes.default.bool,
  autoAdjustOverflow: _propTypes.default.bool,
  className: _propTypes.default.string,
  children: _propTypes.default.node,
  data: _propTypes.default.array,
  defaultOpen: _propTypes.default.bool,
  defaultValue: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  defaultActiveFirstOption: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  dropdownMatchSelectWidth: _propTypes.default.bool,
  dropdownClassName: _propTypes.default.string,
  dropdownStyle: _propTypes.default.object,
  emptyContent: _propTypes.default.node,
  id: _propTypes.default.string,
  insetLabel: _propTypes.default.node,
  insetLabelId: _propTypes.default.string,
  onSearch: _propTypes.default.func,
  onSelect: _propTypes.default.func,
  onClear: _propTypes.default.func,
  onBlur: _propTypes.default.func,
  onFocus: _propTypes.default.func,
  onChange: _propTypes.default.func,
  position: _propTypes.default.oneOf(positionSet),
  placeholder: _propTypes.default.string,
  prefix: _propTypes.default.node,
  onChangeWithObject: _propTypes.default.bool,
  onSelectWithObject: _propTypes.default.bool,
  renderItem: _propTypes.default.func,
  renderSelectedItem: _propTypes.default.func,
  suffix: _propTypes.default.node,
  showClear: _propTypes.default.bool,
  size: _propTypes.default.oneOf(sizeSet),
  style: _propTypes.default.object,
  stopPropagation: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
  maxHeight: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  mouseEnterDelay: _propTypes.default.number,
  mouseLeaveDelay: _propTypes.default.number,
  motion: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.func, _propTypes.default.object]),
  getPopupContainer: _propTypes.default.func,
  triggerRender: _propTypes.default.func,
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  validateStatus: _propTypes.default.oneOf(statusSet),
  zIndex: _propTypes.default.number
};
AutoComplete.Option = _option.default;
AutoComplete.defaultProps = {
  stopPropagation: true,
  motion: true,
  zIndex: _constants2.numbers.DEFAULT_Z_INDEX,
  position: 'bottomLeft',
  data: [],
  showClear: false,
  size: 'default',
  onFocus: _noop2.default,
  onSearch: _noop2.default,
  onClear: _noop2.default,
  onBlur: _noop2.default,
  onSelect: _noop2.default,
  onChange: _noop2.default,
  onSelectWithObject: false,
  onDropdownVisibleChange: _noop2.default,
  defaultActiveFirstOption: false,
  dropdownMatchSelectWidth: true,
  loading: false,
  maxHeight: 300,
  validateStatus: 'default',
  autoFocus: false,
  emptyContent: null // onPressEnter: () => undefined,
  // defaultOpen: false,

};
var _default = AutoComplete;
exports.default = _default;
