// @douyinfe/semi-foundation/lib/es/carousel/constants.js
import { BASE_CLASS_PREFIX } from '../base/constants';
const cssClasses = {
  CAROUSEL: "".concat(BASE_CLASS_PREFIX, "-carousel"),
  CAROUSEL_INDICATOR: "".concat(BASE_CLASS_PREFIX, "-carousel-indicator"),
  CAROUSEL_INDICATOR_LINE: "".concat(BASE_CLASS_PREFIX, "-carousel-indicator-line"),
  CAROUSEL_INDICATOR_DOT: "".concat(BASE_CLASS_PREFIX, "-carousel-indicator-dot"),
  CAROUSEL_INDICATOR_COLUMNAR: "".concat(BASE_CLASS_PREFIX, "-carousel-indicator-columnar"),
  CAROUSEL_INDICATOR_INACTIVE: "".concat(BASE_CLASS_PREFIX, "-carousel-indicator-inactive"),
  CAROUSEL_INDICATOR_ACTIVE: "".concat(BASE_CLASS_PREFIX, "-carousel-indicator-active"),
  CAROUSEL_CONTENT: "".concat(BASE_CLASS_PREFIX, "-carousel-content"),
  CAROUSEL_ARROW: "".concat(BASE_CLASS_PREFIX, "-carousel-arrow")
};
const numbers = {
  DEFAULT_ACTIVE_INDEX: 0,
  DEFAULT_INTERVAL: 2000,
  DEFAULT_SPEED: 300
};
const strings = {
  ANIMATION_MAP: ['slide', 'fade'],
  DIRECTION: ['left', 'right'],
  TYPE_MAP: ['columnar', 'line', 'dot'],
  THEME_MAP: ['dark', 'primary', 'light'],
  POSITION_MAP: ['left', 'center', 'right'],
  ARROW_MAP: ['always', 'hover'],
  SIZE: ['small', 'medium'],
  TRIGGER: ['click', 'hover']
};
export { cssClasses, numbers, strings };
