// @douyinfe/semi-ui/lib/es/configProvider/index.js
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && _indexOfInstanceProperty(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _Object$getOwnPropertySymbols === "function") for (var i = 0, p = _Object$getOwnPropertySymbols(s); i < p.length; i++) {
    if (_indexOfInstanceProperty(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import React from 'react';
import PropTypes from 'prop-types';
import { BASE_CLASS_PREFIX } from '@douyinfe/semi-foundation/lib/es/base/constants';
import DefaultLocale from '../locale/source/zh_CN';
import Context from './context';
export default class ConfigProvider extends React.Component {
  renderChilren() {
    const {
      direction,
      children
    } = this.props;

    if (direction === 'rtl') {
      return /*#__PURE__*/React.createElement("div", {
        className: "".concat(BASE_CLASS_PREFIX, "-rtl")
      }, children);
    }

    return children;
  }

  render() {
    const _a = this.props,
          {
      children,
      direction
    } = _a,
          rest = __rest(_a, ["children", "direction"]);

    return /*#__PURE__*/React.createElement(Context.Provider, {
      value: _Object$assign({
        direction
      }, rest)
    }, this.renderChilren());
  }

}
ConfigProvider.propTypes = {
  locale: PropTypes.object,
  timeZone: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  getPopupContainer: PropTypes.func,
  direction: PropTypes.oneOf(['ltr', 'rtl'])
};
ConfigProvider.defaultProps = {
  locale: DefaultLocale,
  direction: 'ltr'
};
