// @douyinfe/semi-ui/lib/cjs/table/Body/SectionRow.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _isArray = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/array/is-array"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/includes"));

var _isSet2 = _interopRequireDefault(require("lodash/isSet"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/table/constants");

var _utils = require("@douyinfe/semi-foundation/lib/cjs/table/utils");

var _BaseRow = _interopRequireDefault(require("./BaseRow"));

var _tableContext = _interopRequireDefault(require("../table-context"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/* eslint-disable max-len */

/* eslint-disable @typescript-eslint/ban-types */

/* eslint-disable eqeqeq */


/**
 * Grouping component title row
 */
class SectionRow extends _react.PureComponent {
  constructor() {
    var _this;

    super(...arguments);
    _this = this;

    this.onRow = function () {
      const {
        onGroupedRow,
        clickGroupedRowToExpand,
        onExpand,
        groupKey,
        expanded
      } = _this.props;
      const rowProps = {};

      if (typeof onGroupedRow === 'function') {
        (0, _assign.default)(rowProps, onGroupedRow(...arguments));
      }

      return (0, _assign.default)((0, _assign.default)({}, rowProps), {
        onClick: e => {
          if (typeof onExpand === 'function' && clickGroupedRowToExpand) {
            onExpand(!expanded, groupKey, e);
          }

          if (typeof rowProps.onClick === 'function') {
            rowProps.onClick(e);
          }
        }
      });
    };

    this.collectGroupedData = () => {
      const {
        data,
        group,
        rowKey
      } = this.props;

      if ((0, _isArray.default)(data) && data.length && (0, _isSet2.default)(group)) {
        return (0, _filter.default)(data).call(data, record => {
          const realRowKey = typeof rowKey === 'function' ? rowKey(record) : (0, _get2.default)(record, rowKey);
          return realRowKey != null && realRowKey !== '' && group.has(realRowKey);
        });
      }

      return [];
    };

    this.renderExpandIcon = record => {
      const {
        renderExpandIcon,
        groupKey
      } = this.props;

      if (typeof renderExpandIcon === 'function') {
        return renderExpandIcon(record, false, groupKey);
      }

      return null;
    };
  }

  isInnerColumnKey(key) {
    if (key != null) {
      var _context;

      return (0, _includes.default)(_context = [_constants.strings.DEFAULT_KEY_COLUMN_EXPAND, _constants.strings.DEFAULT_KEY_COLUMN_SELECTION]).call(_context, key);
    }

    return false;
  }

  render() {
    const {
      record,
      columns: propColumns = [],
      prefixCls,
      className,
      expanded,
      renderGroupSection,
      components,
      index,
      store,
      group,
      groupKey,
      virtualized,
      style
    } = this.props;
    const props = {};
    let column = {};
    let children = null; // render title

    const cell = typeof renderGroupSection === 'function' ? renderGroupSection(groupKey, [...group]) : null;

    if ( /*#__PURE__*/(0, _react.isValidElement)(cell)) {
      children = cell;
    } else if (cell && Object.prototype.toString.call(cell) === '[object Object]') {
      const _a = cell,
            {
        children: cellChildren
      } = _a,
            restProps = __rest(_a, ["children"]);

      children = cellChildren;
      column = (0, _assign.default)({}, restProps);
    } // Filter out scroll-bar column


    props.colSpan = (0, _utils.filterColumns)(propColumns).length;
    const columns = [(0, _assign.default)({
      render: () => ({
        props,
        children
      })
    }, column)];
    const rowCls = (0, _classnames.default)(className, "".concat(prefixCls, "-row-section"), {
      on: expanded
    });
    const {
      getCellWidths
    } = this.context;
    const baseRowCellWidths = getCellWidths(columns, null, true);
    return /*#__PURE__*/_react.default.createElement(_BaseRow.default, {
      components: components,
      virtualized: virtualized,
      index: index,
      onRow: this.onRow,
      expanded: expanded,
      expandIcon: true,
      isSection: true,
      record: record,
      replaceClassName: rowCls,
      expandableRow: true,
      renderExpandIcon: this.renderExpandIcon,
      rowKey: groupKey,
      columns: columns,
      store: store,
      style: style,
      cellWidths: baseRowCellWidths
    });
  }

}

SectionRow.contextType = _tableContext.default;
SectionRow.propTypes = {
  record: _propTypes.default.object,
  index: _propTypes.default.number,
  columns: _propTypes.default.array,
  group: _propTypes.default.object.isRequired,
  groupKey: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]).isRequired,
  data: _propTypes.default.array,
  renderGroupSection: _propTypes.default.func,
  onGroupedRow: _propTypes.default.func,
  clickGroupedRowToExpand: _propTypes.default.bool,
  components: _propTypes.default.object,
  expanded: _propTypes.default.bool,
  prefixCls: _propTypes.default.string,
  onExpand: _propTypes.default.func,
  virtualized: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.object]),
  style: _propTypes.default.object,
  renderExpandIcon: _propTypes.default.func,
  className: _propTypes.default.string,
  store: _propTypes.default.object,
  rowKey: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.func])
};
SectionRow.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX,
  components: {
    body: {
      row: 'tr',
      cell: 'td'
    }
  }
};
var _default = SectionRow;
exports.default = _default;
