// @douyinfe/semi-ui/lib/cjs/calendar/index.js
"use strict";

var _context;

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _forEachInstanceProperty = require("@babel/runtime-corejs3/core-js-stable/instance/for-each");

var _Object$keys = require("@babel/runtime-corejs3/core-js-stable/object/keys");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

var _exportNames = {};
exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _dayCalendar = _interopRequireDefault(require("./dayCalendar"));

var _weekCalendar = _interopRequireDefault(require("./weekCalendar"));

var _monthCalendar = _interopRequireDefault(require("./monthCalendar"));

var _rangeCalendar = _interopRequireDefault(require("./rangeCalendar"));

require("@douyinfe/semi-foundation/lib/cjs/calendar/calendar.css");

var _interface = require("./interface");

_forEachInstanceProperty(_context = _Object$keys(_interface)).call(_context, function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;

  _Object$defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

// eslint-disable-next-line @typescript-eslint/ban-types
class Calendar extends _baseComponent.default {
  render() {
    const _a = this.props,
          {
      mode
    } = _a,
          rest = __rest(_a, ["mode"]);

    const component = {
      month: /*#__PURE__*/_react.default.createElement(_monthCalendar.default, null),
      week: /*#__PURE__*/_react.default.createElement(_weekCalendar.default, null),
      day: /*#__PURE__*/_react.default.createElement(_dayCalendar.default, null),
      range: /*#__PURE__*/_react.default.createElement(_rangeCalendar.default, null)
    };
    return /*#__PURE__*/_react.default.cloneElement(component[mode], (0, _assign.default)({}, rest));
  }

}

Calendar.propTypes = {
  displayValue: _propTypes.default.instanceOf(Date),
  header: _propTypes.default.node,
  events: _propTypes.default.arrayOf(_propTypes.default.shape({
    allDay: _propTypes.default.bool,
    start: _propTypes.default.instanceOf(Date),
    end: _propTypes.default.instanceOf(Date),
    key: _propTypes.default.string.isRequired,
    children: _propTypes.default.node
  })),
  mode: _propTypes.default.string,
  showCurrTime: _propTypes.default.bool,
  scrollTop: _propTypes.default.number,
  onClick: _propTypes.default.func,
  renderTimeDisplay: _propTypes.default.func,
  markWeekend: _propTypes.default.bool,
  width: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  height: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  style: _propTypes.default.object,
  className: _propTypes.default.string
};
Calendar.defaultProps = {
  events: [],
  displayValue: new Date(),
  showCurrTime: true,
  mode: 'week',
  markWeekend: false,
  height: 600,
  scrollTop: 400
};
var _default = Calendar;
exports.default = _default;
