// @douyinfe/semi-ui/lib/cjs/motions/Rotate.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireWildcard(require("react"));

var _semiAnimationReact = require("@douyinfe/semi-animation-react");

var _propTypes = _interopRequireDefault(require("prop-types"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const formatStyle = function formatStyle(_ref) {
  let {
    rotate = 0
  } = _ref;
  return {
    transform: "rotate(".concat(rotate, "deg)")
  };
};

function OpenIconTransition() {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const {
    children,
    isOpen,
    enterDeg = 180,
    fromDeg = 0,
    duration = 150
  } = props;
  const [immediate, setImmediate] = (0, _react.useState)(true);
  (0, _react.useEffect)(() => {
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    setImmediate(false);
  }, [isOpen]);
  return /*#__PURE__*/_react.default.createElement(_semiAnimationReact.Transition, {
    immediate: immediate,
    state: isOpen ? 'enter' : 'leave',
    from: {
      rotate: fromDeg
    },
    enter: {
      rotate: {
        val: enterDeg,
        duration,
        easing: 'cubic-bezier(.62, .05, .36, .95)'
      }
    },
    leave: {
      rotate: {
        val: fromDeg,
        duration,
        easing: 'cubic-bezier(.62, .05, .36, .95)'
      }
    }
  }, transitionStyle => (0, _map.default)(_react.Children).call(_react.Children, children, child => /*#__PURE__*/(0, _react.isValidElement)(child) ? /*#__PURE__*/(0, _react.cloneElement)(child, (0, _assign.default)((0, _assign.default)({}, child.props), {
    style: (0, _assign.default)((0, _assign.default)({}, child.props.style), formatStyle(transitionStyle))
  })) : child));
}

OpenIconTransition.propTypes = {
  isOpen: _propTypes.default.bool.isRequired,
  children: _propTypes.default.any.isRequired,
  enterDeg: _propTypes.default.number,
  fromDeg: _propTypes.default.number,
  duration: _propTypes.default.number
};
var _default = OpenIconTransition;
exports.default = _default;
