// @douyinfe/semi-ui/lib/es/modal/confirm.js
import _get from "lodash/get";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import _spliceInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/splice";
import React from 'react';
import ReactDOM from 'react-dom';
import { destroyFns } from './Modal';
import ConfirmModal from './ConfirmModal';
import '@douyinfe/semi-foundation/lib/es/modal/modal.css';
import { IconAlertCircle, IconAlertTriangle, IconHelpCircle, IconInfoCircle, IconTickCircle } from '@douyinfe/semi-icons';
export default function confirm(props) {
  // create a dom in adapter?
  const div = document.createElement('div');
  document.body.appendChild(div);

  let currentConfig = _Object$assign({}, props);

  const destroy = () => {
    const unmountResult = ReactDOM.unmountComponentAtNode(div);

    if (unmountResult && div.parentNode) {
      div.parentNode.removeChild(div);
    }

    for (let i = 0; i < destroyFns.length; i++) {
      const fn = destroyFns[i]; // eslint-disable-next-line @typescript-eslint/no-use-before-define

      if (fn === close) {
        _spliceInstanceProperty(destroyFns).call(destroyFns, i, 1);

        break;
      }
    }
  };

  const mergedMotion = typeof props.motion === 'undefined' || props.motion ? _Object$assign(_Object$assign({}, props.motion), {
    didLeave: function () {
      const didLeave = _get(props.motion, 'didLeave');

      if (typeof didLeave === 'function') {
        didLeave(...arguments);
      }

      destroy();
    }
  }) : false;

  function render(renderProps) {
    ReactDOM.render( /*#__PURE__*/React.createElement(ConfirmModal, _Object$assign({}, renderProps, {
      motion: mergedMotion
    })), div);
  }

  function close() {
    currentConfig = _Object$assign(_Object$assign({}, currentConfig), {
      visible: false
    });
    render(currentConfig);
  }

  function update(newConfig) {
    currentConfig = _Object$assign(_Object$assign({}, currentConfig), newConfig);
    render(currentConfig);
  }

  render(currentConfig);
  destroyFns.push(close);
  return {
    destroy: close,
    update
  };
}
export function withInfo(props) {
  return _Object$assign({
    type: 'info',
    icon: /*#__PURE__*/React.createElement(IconInfoCircle, null)
  }, props);
}
export function withSuccess(props) {
  return _Object$assign({
    type: 'success',
    icon: /*#__PURE__*/React.createElement(IconTickCircle, null)
  }, props);
}
export function withWarning(props) {
  return _Object$assign({
    type: 'warning',
    icon: /*#__PURE__*/React.createElement(IconAlertTriangle, null)
  }, props);
}
export function withError(props) {
  return _Object$assign({
    type: 'error',
    icon: /*#__PURE__*/React.createElement(IconAlertCircle, null)
  }, props);
}
export function withConfirm(props) {
  return _Object$assign({
    type: 'confirm',
    icon: /*#__PURE__*/React.createElement(IconHelpCircle, null)
  }, props);
}
