// @douyinfe/semi-ui/lib/es/tree/nodeList.js
import _isEqual from "lodash/isEqual";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _spliceInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/splice";
import _Array$isArray from "@babel/runtime-corejs3/core-js-stable/array/is-array";
import React, { PureComponent } from 'react';
import TreeContext from './treeContext';
import Collapse from './collapse';

const getTreeNodeKey = treeNode => {
  const {
    data
  } = treeNode;
  const {
    key
  } = data;
  return key;
};

export default class NodeList extends PureComponent {
  constructor(props) {
    super(props);

    this.onMotionEnd = () => {
      typeof this.props.onMotionEnd === 'function' && this.props.onMotionEnd();
      this.setState({
        transitionNodes: []
      });
    };

    this.state = {
      transitionNodes: []
    };
  }

  static getDerivedStateFromProps(props, prevState) {
    var _context;

    const {
      flattenNodes = [],
      motionKeys,
      motionType,
      flattenList = []
    } = props;
    const hasChanged = !_isEqual(prevState.cachedMotionKeys, motionKeys) || !_isEqual(_mapInstanceProperty(_context = prevState.cachedData).call(_context, i => i.key), _mapInstanceProperty(flattenNodes).call(flattenNodes, i => i.key));
    const motionArr = [...motionKeys];

    if (!hasChanged || !motionArr.length) {
      return null;
    }

    const transitionNodes = [];
    const transitionRange = [];
    let rangeStart = 0;
    let newState = {};
    const lookUpTarget = motionType === 'hide' && flattenList ? flattenList : flattenNodes;

    _forEachInstanceProperty(lookUpTarget).call(lookUpTarget, (treeNode, ind) => {
      const nodeKey = getTreeNodeKey(treeNode);

      if (motionKeys.has(nodeKey)) {
        transitionRange.push(treeNode);

        if (nodeKey === motionArr[0]) {
          rangeStart = ind;
        }
      } else {
        transitionNodes.push(treeNode);
      }
    });

    _spliceInstanceProperty(transitionNodes).call(transitionNodes, rangeStart, 0, transitionRange);

    newState = {
      transitionNodes,
      cachedData: flattenNodes,
      cachedMotionKeys: motionKeys,
      cachedMotionType: motionType
    };
    return newState;
  }

  render() {
    const {
      flattenNodes,
      motionType,
      searchTargetIsDeep,
      renderTreeNode
    } = this.props;
    const {
      transitionNodes
    } = this.state;
    const mapData = transitionNodes.length && !searchTargetIsDeep ? transitionNodes : flattenNodes;

    const options = _mapInstanceProperty(mapData).call(mapData, treeNode => {
      const isMotionNode = _Array$isArray(treeNode);

      if (isMotionNode && !treeNode.length) {
        return null;
      }

      if (isMotionNode && treeNode.length) {
        const nodeKey = getTreeNodeKey(treeNode[0]);
        return /*#__PURE__*/React.createElement(Collapse, {
          motionType: motionType === 'show' ? 'enter' : 'leave',
          key: "motion-".concat(nodeKey),
          onMotionEnd: this.onMotionEnd,
          motion: Boolean(motionType)
        }, _mapInstanceProperty(treeNode).call(treeNode, node => renderTreeNode(node)));
      }

      return renderTreeNode(treeNode);
    });

    return options;
  }

}
NodeList.contextType = TreeContext;
