// @douyinfe/semi-foundation/lib/cjs/popover/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.numbers = exports.cssClasses = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _constants = require("../tooltip/constants");

var _constants2 = require("../base/constants");

const cssClasses = {
  PREFIX: "".concat(_constants2.BASE_CLASS_PREFIX, "-popover"),
  ARROW: "".concat(_constants2.BASE_CLASS_PREFIX, "-popover-icon-arrow")
};
exports.cssClasses = cssClasses;
const strings = {
  POSITION_SET: ['top', 'topLeft', 'topRight', 'left', 'leftTop', 'leftBottom', 'right', 'rightTop', 'rightBottom', 'bottom', 'bottomLeft', 'bottomRight', 'leftTopOver', 'rightTopOver'],
  TRIGGER_SET: ['hover', 'focus', 'click', 'custom'],
  DEFAULT_ARROW_STYLE: {
    borderOpacity: '1',
    backgroundColor: 'var(--semi-color-bg-3)',
    // borderColor: 'var(--semi-color-shadow)',
    borderColor: 'var(--semi-color-border)'
  }
};
exports.strings = strings;
const numbers = {
  ARROW_BOUNDING: (0, _assign.default)((0, _assign.default)({}, _constants.numbers.ARROW_BOUNDING), {
    offsetY: 6,
    offsetX: 0,
    height: 8
  }),
  SPACING: 4,
  SPACING_WITH_ARROW: 10,
  DEFAULT_Z_INDEX: 1030
};
exports.numbers = numbers;
