// @douyinfe/semi-ui/lib/cjs/modal/ConfirmModal.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/modal/constants");

var _Modal = _interopRequireDefault(require("./Modal"));

var _utils = require("../_utils");

require("@douyinfe/semi-foundation/lib/cjs/modal/modal.css");

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/* eslint-disable react/destructuring-assignment */

/* eslint-disable eqeqeq */


const ConfirmModal = props => {
  var _context, _context2;

  const [visible, setVisible] = (0, _react.useState)(true);
  const [confirmLoading, setConfirmLoading] = (0, _react.useState)();
  const [cancelLoading, setCancelLoading] = (0, _react.useState)();
  const {
    direction
  } = props;

  const {
    title,
    content,
    icon,
    type,
    onCancel,
    onOk,
    className
  } = props,
        rest = __rest(props, ["title", "content", "icon", "type", "onCancel", "onOk", "className"]);

  const handleOk = (0, _react.useCallback)(e => {
    const res = onOk && onOk(e);

    if (res && res.then) {
      setConfirmLoading(true);
      res.then(function () {
        setVisible(false);
        setConfirmLoading(false);
      }, err => {
        setConfirmLoading(false);
      });
    } else {
      setVisible(false);
    }
  }, [onOk]);
  const handleCancel = (0, _react.useCallback)(e => {
    const res = onCancel && onCancel(e);

    if (res && res.then) {
      setCancelLoading(true);
      res.then(function () {
        setVisible(false);
        setCancelLoading(false);
      }, err => {
        setCancelLoading(false);
      });
    } else {
      setVisible(false);
    }
  }, [onCancel]);
  const confirmCls = "".concat(_constants.cssClasses.DIALOG, "-confirm");
  const wrapperCls = (0, _classnames.default)(className, confirmCls, {
    ["".concat(confirmCls, "-rtl")]: direction === 'rtl'
  });
  const typeCls = (0, _classnames.default)((0, _concat.default)(_context = "".concat(_constants.cssClasses.DIALOG, "-")).call(_context, type));
  const iconNode = (0, _utils.isSemiIcon)(icon) ? /*#__PURE__*/_react.default.cloneElement(icon, {
    className: (0, _concat.default)(_context2 = "".concat(confirmCls, "-icon ")).call(_context2, typeCls, "-icon"),
    size: 'extra-large'
  }) : icon;
  const titleNode = title == null ? null : /*#__PURE__*/_react.default.createElement("span", {
    className: "".concat(confirmCls, "-title-text")
  }, title);
  const contentCls = (0, _classnames.default)("".concat(confirmCls, "-content"), {
    ["".concat(confirmCls, "-content-withIcon")]: props.icon
  });
  return /*#__PURE__*/_react.default.createElement(_Modal.default, (0, _assign.default)({
    className: wrapperCls,
    title: titleNode,
    confirmLoading: confirmLoading,
    cancelLoading: cancelLoading,
    onOk: handleOk,
    onCancel: handleCancel,
    icon: iconNode,
    visible: visible
  }, rest), /*#__PURE__*/_react.default.createElement("div", {
    className: contentCls,
    "x-semi-prop": "content"
  }, content));
};

var _default = ConfirmModal;
exports.default = _default;
