// @douyinfe/semi-ui/lib/cjs/modal/ModalContent.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _setTimeout2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set-timeout"));

var _bind = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/bind"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/modal/constants");

var _context4 = _interopRequireDefault(require("../configProvider/context"));

var _iconButton = _interopRequireDefault(require("../iconButton"));

var _typography = _interopRequireDefault(require("../typography"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _modalContentFoundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/modal/modalContentFoundation"));

var _semiIcons = require("@douyinfe/semi-icons");

var _FocusHandle = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/utils/FocusHandle"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/* eslint-disable eqeqeq */


let uuid = 0;

class ModalContent extends _baseComponent.default {
  constructor(props) {
    super(props);

    this.onKeyDown = e => {
      this.foundation.handleKeyDown(e);
    }; // Record when clicking the modal box


    this.onDialogMouseDown = () => {
      this.foundation.handleDialogMouseDown();
    }; // Cancel recording when clicking the modal box at the end


    this.onMaskMouseUp = () => {
      this.foundation.handleMaskMouseUp();
    }; // onMaskClick will judge dialogMouseDown before onMaskMouseUp updates dialogMouseDown


    this.onMaskClick = e => {
      this.foundation.handleMaskClick(e);
    };

    this.close = e => {
      this.foundation.close(e);
    };

    this.getMaskElement = () => {
      const props = __rest(this.props, []);

      const {
        mask,
        maskClassName
      } = props;

      if (mask) {
        const className = (0, _classnames.default)("".concat(_constants.cssClasses.DIALOG, "-mask"), {// [`${cssClasses.DIALOG}-mask-hidden`]: !props.visible,
        });
        return /*#__PURE__*/_react.default.createElement("div", {
          key: "mask",
          className: (0, _classnames.default)(className, maskClassName),
          style: props.maskStyle
        });
      }

      return null;
    };

    this.renderCloseBtn = () => {
      const {
        closable,
        closeIcon
      } = this.props;
      let closer;

      if (closable) {
        const iconType = closeIcon || /*#__PURE__*/_react.default.createElement(_semiIcons.IconClose, {
          "x-semi-prop": "closeIcon"
        });

        closer = /*#__PURE__*/_react.default.createElement(_iconButton.default, {
          "aria-label": "close",
          className: "".concat(_constants.cssClasses.DIALOG, "-close"),
          key: "close-btn",
          onClick: this.close,
          type: "tertiary",
          icon: iconType,
          theme: "borderless",
          size: "small"
        });
      }

      return closer;
    };

    this.renderIcon = () => {
      const {
        icon
      } = this.props;
      return icon ? /*#__PURE__*/_react.default.createElement("span", {
        className: "".concat(_constants.cssClasses.DIALOG, "-icon-wrapper"),
        "x-semi-prop": "icon"
      }, icon) : null;
    };

    this.renderHeader = () => {
      if ('header' in this.props) {
        return this.props.header;
      }

      const {
        title
      } = this.props;
      const closer = this.renderCloseBtn();
      const icon = this.renderIcon();
      return title === null || title === undefined ? null : /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(_constants.cssClasses.DIALOG, "-header")
      }, icon, /*#__PURE__*/_react.default.createElement(_typography.default.Title, {
        heading: 5,
        className: "".concat(_constants.cssClasses.DIALOG, "-title"),
        id: "".concat(_constants.cssClasses.DIALOG, "-title"),
        "x-semi-prop": "title"
      }, title), closer);
    };

    this.renderBody = () => {
      const {
        bodyStyle,
        children,
        title
      } = this.props;
      const bodyCls = (0, _classnames.default)("".concat(_constants.cssClasses.DIALOG, "-body"), {
        ["".concat(_constants.cssClasses.DIALOG, "-withIcon")]: this.props.icon
      });
      const closer = this.renderCloseBtn();
      const icon = this.renderIcon();
      const hasHeader = title !== null && title !== undefined || 'header' in this.props;
      return hasHeader ? /*#__PURE__*/_react.default.createElement("div", {
        className: bodyCls,
        id: "".concat(_constants.cssClasses.DIALOG, "-body"),
        style: bodyStyle,
        "x-semi-prop": "children"
      }, children) : /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(_constants.cssClasses.DIALOG, "-body-wrapper")
      }, icon, /*#__PURE__*/_react.default.createElement("div", {
        className: bodyCls,
        style: bodyStyle,
        "x-semi-prop": "children"
      }, children), closer);
    };

    this.getDialogElement = () => {
      var _context;

      const props = __rest(this.props, []);

      const style = {};
      const digCls = (0, _classnames.default)("".concat(_constants.cssClasses.DIALOG), {
        ["".concat(_constants.cssClasses.DIALOG, "-centered")]: props.centered,
        [(0, _concat.default)(_context = "".concat(_constants.cssClasses.DIALOG, "-")).call(_context, props.size)]: props.size
      });

      if (props.width) {
        style.width = props.width;
      }

      if (props.height) {
        style.height = props.height;
      }

      if (props.isFullScreen) {
        style.width = '100%';
        style.height = '100%';
        style.margin = 'unset';
      }

      const body = this.renderBody();
      const header = this.renderHeader();
      const footer = props.footer ? /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(_constants.cssClasses.DIALOG, "-footer"),
        "x-semi-prop": "footer"
      }, props.footer) : null;

      const dialogElement =
      /*#__PURE__*/
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      _react.default.createElement("div", {
        key: "dialog-element",
        className: digCls,
        onMouseDown: this.onDialogMouseDown,
        style: (0, _assign.default)((0, _assign.default)({}, props.style), style),
        id: this.dialogId
      }, /*#__PURE__*/_react.default.createElement("div", {
        role: "dialog",
        ref: this.modalDialogRef,
        tabIndex: -1,
        "aria-modal": "true",
        "aria-labelledby": "".concat(_constants.cssClasses.DIALOG, "-title"),
        "aria-describedby": "".concat(_constants.cssClasses.DIALOG, "-body"),
        onAnimationEnd: props.onAnimationEnd,
        className: (0, _classnames.default)(["".concat(_constants.cssClasses.DIALOG, "-content"), props.contentClassName, {
          ["".concat(_constants.cssClasses.DIALOG, "-content-fullScreen")]: props.isFullScreen
        }])
      }, header, body, footer)); // return props.visible ? dialogElement : null;


      return dialogElement;
    };

    this.state = {
      dialogMouseDown: false,
      prevFocusElement: _FocusHandle.default.getActiveElement()
    };
    this.foundation = new _modalContentFoundation.default(this.adapter);
    this.dialogId = "dialog-".concat(uuid++);
    this.modalDialogRef = /*#__PURE__*/_react.default.createRef();
  }

  get adapter() {
    return (0, _assign.default)((0, _assign.default)({}, super.adapter), {
      notifyClose: e => {
        this.props.onClose(e);
      },
      notifyDialogMouseDown: () => {
        this.setState({
          dialogMouseDown: true
        });
      },
      notifyDialogMouseUp: () => {
        if (this.state.dialogMouseDown) {
          // Not setting setTimeout triggers close when modal external mouseUp
          this.timeoutId = (0, _setTimeout2.default)(() => {
            this.setState({
              dialogMouseDown: false
            });
          }, 0);
        }
      },
      addKeyDownEventListener: () => {
        if (this.props.closeOnEsc) {
          var _context2;

          document.addEventListener('keydown', (0, _bind.default)(_context2 = this.foundation.handleKeyDown).call(_context2, this.foundation));
        }
      },
      removeKeyDownEventListener: () => {
        if (this.props.closeOnEsc) {
          var _context3;

          document.removeEventListener('keydown', (0, _bind.default)(_context3 = this.foundation.handleKeyDown).call(_context3, this.foundation));
        }
      },
      getMouseState: () => this.state.dialogMouseDown,
      modalDialogFocus: () => {
        var _a, _b, _c;

        const {
          preventScroll
        } = this.props;
        let activeElementInDialog;

        if (this.modalDialogRef) {
          const activeElement = _FocusHandle.default.getActiveElement();

          activeElementInDialog = this.modalDialogRef.current.contains(activeElement);
          (_a = this.focusTrapHandle) === null || _a === void 0 ? void 0 : _a.destroy();
          this.focusTrapHandle = new _FocusHandle.default(this.modalDialogRef.current, {
            preventScroll
          });
        }

        if (!activeElementInDialog) {
          (_c = (_b = this.modalDialogRef) === null || _b === void 0 ? void 0 : _b.current) === null || _c === void 0 ? void 0 : _c.focus({
            preventScroll
          });
        }
      },
      modalDialogBlur: () => {
        var _a, _b;

        (_a = this.modalDialogRef) === null || _a === void 0 ? void 0 : _a.current.blur();
        (_b = this.focusTrapHandle) === null || _b === void 0 ? void 0 : _b.destroy();
      },
      prevFocusElementReFocus: () => {
        const {
          prevFocusElement
        } = this.state;
        const {
          preventScroll
        } = this.props;
        const focus = (0, _get2.default)(prevFocusElement, 'focus');
        (0, _isFunction2.default)(focus) && prevFocusElement.focus({
          preventScroll
        });
      }
    });
  }

  componentDidMount() {
    this.foundation.handleKeyDownEventListenerMount();
    this.foundation.modalDialogFocus();
  }

  componentWillUnmount() {
    clearTimeout(this.timeoutId);
    this.foundation.destroy();
  }

  render() {
    const {
      maskClosable,
      className,
      getPopupContainer,
      maskFixed,
      getContainerContext
    } = this.props;
    const {
      direction
    } = this.context;
    const classList = (0, _classnames.default)(className, {
      ["".concat(_constants.cssClasses.DIALOG, "-popup")]: getPopupContainer && !maskFixed,
      ["".concat(_constants.cssClasses.DIALOG, "-fixed")]: maskFixed,
      ["".concat(_constants.cssClasses.DIALOG, "-rtl")]: direction === 'rtl'
    });
    const containerContext = getContainerContext();

    const elem = /*#__PURE__*/_react.default.createElement("div", {
      className: classList
    }, this.getMaskElement(), /*#__PURE__*/_react.default.createElement("div", {
      role: "none",
      tabIndex: -1,
      className: "".concat(_constants.cssClasses.DIALOG, "-wrap"),
      onClick: maskClosable ? this.onMaskClick : null,
      onMouseUp: maskClosable ? this.onMaskMouseUp : null
    }, this.getDialogElement())); // eslint-disable-next-line max-len


    return containerContext && containerContext.Provider ? /*#__PURE__*/_react.default.createElement(containerContext.Provider, {
      value: containerContext.value
    }, elem) : elem;
  }

}

exports.default = ModalContent;
ModalContent.contextType = _context4.default;
ModalContent.propTypes = {
  close: _propTypes.default.func,
  getContainerContext: _propTypes.default.func,
  contentClassName: _propTypes.default.string,
  maskClassName: _propTypes.default.string,
  onAnimationEnd: _propTypes.default.func,
  preventScroll: _propTypes.default.bool
};
ModalContent.defaultProps = {
  close: _noop2.default,
  getContainerContext: _noop2.default,
  contentClassName: '',
  maskClassName: ''
};
