// @douyinfe/semi-ui/lib/es/steps/fillStep.js
import _isFunction from "lodash/isFunction";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { stepsClasses as css } from '@douyinfe/semi-foundation/lib/es/steps/constants';
import { IconTickCircle, IconAlertCircle, IconAlertTriangle } from '@douyinfe/semi-icons';

const FillStep = props => {
  var _context;

  const {
    prefixCls,
    className,
    title,
    description,
    status,
    style,
    onClick,
    icon,
    onChange,
    stepNumber,
    onKeyDown
  } = props;

  const renderIcon = () => {
    let inner, progress;

    if ('icon' in props) {
      inner = icon;
    } else if ('status' in props) {
      switch (status) {
        case 'error':
          inner = /*#__PURE__*/React.createElement(IconAlertCircle, {
            size: "extra-large"
          });
          break;

        case 'wait':
          inner = stepNumber;
          break;

        case 'process':
          inner = stepNumber;
          progress = true;
          break;

        case 'finish':
          inner = /*#__PURE__*/React.createElement(IconTickCircle, {
            size: "extra-large"
          });
          break;

        case 'warning':
          inner = /*#__PURE__*/React.createElement(IconAlertTriangle, {
            size: "extra-large"
          });
          break;

        default:
          inner = null;
          break;
      }
    }

    const cls = classnames({
      ["".concat(prefixCls, "-left")]: true,
      ["".concat(prefixCls, "-icon")]: 'icon' in props,
      ["".concat(prefixCls, "-plain")]: !('icon' in props),
      ["".concat(prefixCls, "-icon-process")]: progress
    });
    return inner ? /*#__PURE__*/React.createElement("div", {
      className: cls
    }, inner) : null;
  };

  const handleClick = e => {
    if (_isFunction(onClick)) {
      onClick(e);
    }

    onChange();
  };

  const handleKeyDown = e => {
    if (e.key === 'Enter') {
      if (_isFunction(onKeyDown)) {
        onKeyDown(e);
      }

      onChange();
    }
  };

  return /*#__PURE__*/React.createElement("div", {
    role: props["role"],
    "aria-label": props["aria-label"],
    "aria-current": "step",
    tabIndex: 0,
    className: classnames({
      [prefixCls]: true,
      [_concatInstanceProperty(_context = "".concat(prefixCls, "-")).call(_context, status)]: Boolean(status),
      ["".concat(prefixCls, "-clickable")]: onClick
    }, className),
    style: style,
    onClick: e => {
      handleClick(e);
    },
    onKeyDown: handleKeyDown
  }, renderIcon(), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-content")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-title"),
    title: typeof title === 'string' ? title : null
  }, /*#__PURE__*/React.createElement("span", {
    className: "".concat(prefixCls, "-title-text")
  }, title)), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-description"),
    title: typeof description === 'string' ? description : null
  }, description)));
};

FillStep.propTypes = {
  prefixCls: PropTypes.string,
  description: PropTypes.node,
  icon: PropTypes.node,
  status: PropTypes.oneOf(['wait', 'process', 'finish', 'error', 'warning']),
  title: PropTypes.node,
  className: PropTypes.string,
  style: PropTypes.object,
  onClick: PropTypes.func
};
FillStep.defaultProps = {
  prefixCls: css.ITEM,
  status: 'wait',
  className: ''
};
export default FillStep;
