// @douyinfe/semi-ui/lib/es/steps/index.js
import _noop from "lodash/noop";
import _indexOfInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/index-of";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && _indexOfInstanceProperty(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _Object$getOwnPropertySymbols === "function") for (var i = 0, p = _Object$getOwnPropertySymbols(s); i < p.length; i++) {
    if (_indexOfInstanceProperty(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import '@douyinfe/semi-foundation/lib/es/steps/steps.css';
import Step from './step';
import FillSteps from './fillSteps';
import BasicSteps from './basicSteps';
import NavSteps from './navSteps';
import Context from './context';

class Steps extends Component {
  renderComponent() {
    const _a = this.props,
          {
      type
    } = _a,
          restProps = __rest(_a, ["type"]);

    switch (type) {
      case 'fill':
        return /*#__PURE__*/React.createElement(FillSteps, _Object$assign({}, restProps));

      case 'basic':
        return /*#__PURE__*/React.createElement(BasicSteps, _Object$assign({}, restProps));

      case 'nav':
        return /*#__PURE__*/React.createElement(NavSteps, _Object$assign({}, restProps));

      default:
        return null;
    }
  }

  render() {
    const {
      type
    } = this.props;
    return /*#__PURE__*/React.createElement(Context.Provider, {
      value: {
        type
      }
    }, this.renderComponent());
  }

}

Steps.Step = Step;
Steps.propTypes = {
  onChange: PropTypes.func,
  type: PropTypes.oneOf(['fill', 'basic', 'nav']),
  size: PropTypes.oneOf(['small', 'default'])
};
Steps.defaultProps = {
  onChange: _noop,
  type: 'fill',
  size: 'default'
};
export default Steps;
