// @douyinfe/semi-ui/lib/cjs/grid/col.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _row = require("./row");

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/grid/constants");

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/**
 * Implementation reference from: https://github.com/ant-design/ant-design/blob/master/components/grid/col.tsx
 */


const objectOrNumber = _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.number]);

class Col extends _react.default.Component {
  render() {
    var _context, _context12, _context13, _context14, _context15, _context16;

    const {
      props
    } = this;

    const {
      prefixCls,
      span,
      order,
      offset,
      push,
      pull,
      className,
      children
    } = props,
          others = __rest(props, ["prefixCls", "span", "order", "offset", "push", "pull", "className", "children"]);

    let sizeClassObj = {};
    const prefix = "".concat(prefixCls, "-col");
    (0, _forEach.default)(_context = ['xs', 'sm', 'md', 'lg', 'xl', 'xxl']).call(_context, size => {
      var _context2, _context3, _context4, _context5, _context6, _context7, _context8, _context9, _context10, _context11;

      let sizeProps = {};

      if (typeof props[size] === 'number') {
        sizeProps.span = props[size];
      } else if (typeof props[size] === 'object') {
        sizeProps = props[size] || {};
      }

      delete others[size];
      sizeClassObj = (0, _assign.default)((0, _assign.default)({}, sizeClassObj), {
        [(0, _concat.default)(_context2 = (0, _concat.default)(_context3 = "".concat(prefix, "-")).call(_context3, size, "-")).call(_context2, sizeProps.span)]: sizeProps.span !== undefined,
        [(0, _concat.default)(_context4 = (0, _concat.default)(_context5 = "".concat(prefix, "-")).call(_context5, size, "-order-")).call(_context4, sizeProps.order)]: sizeProps.order || sizeProps.order === 0,
        [(0, _concat.default)(_context6 = (0, _concat.default)(_context7 = "".concat(prefix, "-")).call(_context7, size, "-offset-")).call(_context6, sizeProps.offset)]: sizeProps.offset || sizeProps.offset === 0,
        [(0, _concat.default)(_context8 = (0, _concat.default)(_context9 = "".concat(prefix, "-")).call(_context9, size, "-push-")).call(_context8, sizeProps.push)]: sizeProps.push || sizeProps.push === 0,
        [(0, _concat.default)(_context10 = (0, _concat.default)(_context11 = "".concat(prefix, "-")).call(_context11, size, "-pull-")).call(_context10, sizeProps.pull)]: sizeProps.pull || sizeProps.pull === 0
      });
    });
    const classes = (0, _classnames.default)(prefix, {
      [(0, _concat.default)(_context12 = "".concat(prefix, "-")).call(_context12, span)]: span !== undefined,
      [(0, _concat.default)(_context13 = "".concat(prefix, "-order-")).call(_context13, order)]: order,
      [(0, _concat.default)(_context14 = "".concat(prefix, "-offset-")).call(_context14, offset)]: offset,
      [(0, _concat.default)(_context15 = "".concat(prefix, "-push-")).call(_context15, push)]: push,
      [(0, _concat.default)(_context16 = "".concat(prefix, "-pull-")).call(_context16, pull)]: pull
    }, className, sizeClassObj);
    let {
      style
    } = others;
    let gutters;

    try {
      gutters = this.context.gutters;
    } catch (error) {
      throw new Error('please make sure <Col> inside <Row>');
    }

    style = (0, _assign.default)((0, _assign.default)((0, _assign.default)({}, gutters[0] > 0 ? {
      paddingLeft: gutters[0] / 2,
      paddingRight: gutters[0] / 2
    } : {}), gutters[1] > 0 ? {
      paddingTop: gutters[1] / 2,
      paddingBottom: gutters[1] / 2
    } : {}), style);
    return /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({}, others, {
      style: style,
      className: classes,
      "x-semi-prop": "children"
    }), children);
  }

}

Col.contextType = _row.RowContext;
Col.propTypes = {
  span: _propTypes.default.number,
  order: _propTypes.default.number,
  offset: _propTypes.default.number,
  push: _propTypes.default.number,
  pull: _propTypes.default.number,
  className: _propTypes.default.string,
  children: _propTypes.default.node,
  xs: objectOrNumber,
  sm: objectOrNumber,
  md: objectOrNumber,
  lg: objectOrNumber,
  xl: objectOrNumber,
  xxl: objectOrNumber,
  prefixCls: _propTypes.default.string
};
Col.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX
};
var _default = Col;
exports.default = _default;
