// @douyinfe/semi-ui/lib/cjs/card/index.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _isArray = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/array/is-array"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _isString2 = _interopRequireDefault(require("lodash/isString"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/card/constants");

require("@douyinfe/semi-foundation/lib/cjs/card/card.css");

var _meta = _interopRequireDefault(require("./meta"));

var _classnames = _interopRequireDefault(require("classnames"));

var _skeleton = _interopRequireDefault(require("../skeleton"));

var _typography = _interopRequireDefault(require("../typography"));

var _space = _interopRequireDefault(require("../space"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const prefixcls = _constants.cssClasses.PREFIX;

class Card extends _react.PureComponent {
  constructor() {
    super(...arguments);

    this.renderHeader = () => {
      const {
        title,
        headerExtraContent,
        header,
        headerLine,
        headerStyle
      } = this.props;
      const headerCls = (0, _classnames.default)("".concat(prefixcls, "-header"), {
        ["".concat(prefixcls, "-header-bordered")]: Boolean(headerLine)
      });
      const headerWrapperCls = (0, _classnames.default)("".concat(prefixcls, "-header-wrapper"));
      const titleCls = (0, _classnames.default)("".concat(prefixcls, "-header-wrapper-title"), {
        ["".concat(prefixcls, "-header-wrapper-spacing")]: Boolean(headerExtraContent)
      });

      if (header || headerExtraContent || title) {
        return /*#__PURE__*/_react.default.createElement("div", {
          style: headerStyle,
          className: headerCls
        }, header ||
        /*#__PURE__*/
        // Priority of header over title and headerExtraContent
        _react.default.createElement("div", {
          className: headerWrapperCls
        }, headerExtraContent && /*#__PURE__*/_react.default.createElement("div", {
          className: "".concat(prefixcls, "-header-wrapper-extra"),
          "x-semi-prop": "headerExtraContent"
        }, headerExtraContent), title && /*#__PURE__*/_react.default.createElement("div", {
          className: titleCls
        }, (0, _isString2.default)(title) ? /*#__PURE__*/_react.default.createElement(_typography.default.Title, {
          heading: 6,
          ellipsis: {
            showTooltip: true,
            rows: 1
          },
          "x-semi-prop": "title"
        }, title) : title)));
      }

      return null;
    };

    this.renderCover = () => {
      const {
        cover
      } = this.props;
      const coverCls = (0, _classnames.default)("".concat(prefixcls, "-cover"));
      return cover && /*#__PURE__*/_react.default.createElement("div", {
        className: coverCls,
        "x-semi-prop": "cover"
      }, cover);
    };

    this.renderBody = () => {
      const {
        bodyStyle,
        children,
        actions,
        loading
      } = this.props;
      const bodyCls = (0, _classnames.default)("".concat(prefixcls, "-body"));
      const actionsCls = (0, _classnames.default)("".concat(prefixcls, "-body-actions"));
      const actionsItemCls = (0, _classnames.default)("".concat(prefixcls, "-body-actions-item"));

      const placeholder = /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_skeleton.default.Title, null), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement(_skeleton.default.Paragraph, {
        rows: 3
      }));

      return /*#__PURE__*/_react.default.createElement("div", {
        style: bodyStyle,
        className: bodyCls
      }, children && /*#__PURE__*/_react.default.createElement(_skeleton.default, {
        placeholder: placeholder,
        loading: loading,
        active: true
      }, children), (0, _isArray.default)(actions) && /*#__PURE__*/_react.default.createElement("div", {
        className: actionsCls
      }, /*#__PURE__*/_react.default.createElement(_space.default, {
        spacing: 12
      }, (0, _map.default)(actions).call(actions, (item, idx) => /*#__PURE__*/_react.default.createElement("div", {
        key: idx,
        className: actionsItemCls,
        "x-semi-prop": "actions.".concat(idx)
      }, item)))));
    };

    this.renderFooter = () => {
      const {
        footer,
        footerLine,
        footerStyle
      } = this.props;
      const footerCls = (0, _classnames.default)("".concat(prefixcls, "-footer"), {
        ["".concat(prefixcls, "-footer-bordered")]: footerLine
      });
      return footer && /*#__PURE__*/_react.default.createElement("div", {
        style: footerStyle,
        className: footerCls,
        "x-semi-prop": "footer"
      }, footer);
    };
  }

  render() {
    var _context;

    const _a = this.props,
          {
      bordered,
      shadows,
      style,
      className
    } = _a,
          otherProps = __rest(_a, ["bordered", "shadows", "style", "className"]);

    const others = (0, _omit2.default)(otherProps, ['actions', 'bodyStyle', 'cover', 'headerExtraContent', 'footer', 'footerLine', 'footerStyle', 'header', 'headerLine', 'headerStyle', 'loading', 'title']);
    const cardCls = (0, _classnames.default)(prefixcls, className, {
      ["".concat(prefixcls, "-bordered")]: bordered,
      ["".concat(prefixcls, "-shadows")]: shadows,
      [(0, _concat.default)(_context = "".concat(prefixcls, "-shadows-")).call(_context, shadows)]: shadows
    });
    return /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({}, others, {
      "aria-busy": this.props.loading,
      className: cardCls,
      style: style
    }), this.renderHeader(), this.renderCover(), this.renderBody(), this.renderFooter());
  }

}

Card.Meta = _meta.default;
Card.propTypes = {
  actions: _propTypes.default.array,
  bodyStyle: _propTypes.default.object,
  bordered: _propTypes.default.bool,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  cover: _propTypes.default.node,
  footer: _propTypes.default.node,
  footerLine: _propTypes.default.bool,
  footerStyle: _propTypes.default.object,
  header: _propTypes.default.node,
  headerExtraContent: _propTypes.default.node,
  headerLine: _propTypes.default.bool,
  headerStyle: _propTypes.default.object,
  loading: _propTypes.default.bool,
  shadows: _propTypes.default.oneOf(_constants.strings.SHADOWS),
  style: _propTypes.default.object,
  title: _propTypes.default.node,
  'aria-label': _propTypes.default.string
};
Card.defaultProps = {
  bordered: true,
  footerLine: false,
  headerLine: true,
  loading: false
};
var _default = Card;
exports.default = _default;
