// @douyinfe/semi-ui/lib/cjs/list/item.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/list/constants");

var _grid = require("../grid");

var _listContext = _interopRequireDefault(require("./list-context"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const prefixCls = _constants.cssClasses.PREFIX;

class ListItem extends _react.PureComponent {
  wrapWithGrid(content) {
    const {
      grid
    } = this.context;

    const {
      gutter,
      justify,
      type,
      align
    } = grid,
          rest = __rest(grid, ["gutter", "justify", "type", "align"]);

    return /*#__PURE__*/_react.default.createElement(_grid.Col, (0, _assign.default)({}, rest), content);
  }

  render() {
    var _context;

    const {
      header,
      main,
      className,
      style,
      extra,
      children,
      align,
      onClick,
      onRightClick,
      onMouseEnter,
      onMouseLeave
    } = this.props;
    const {
      onRightClick: contextOnRightClick,
      onClick: contextOnClick,
      grid: contextGrid
    } = this.context;
    const handleContextMenu = onRightClick ? onRightClick : contextOnRightClick;
    const handleClick = onClick ? onClick : contextOnClick;
    const itemCls = (0, _classnames.default)("".concat(prefixCls, "-item"), className);
    const bodyCls = (0, _classnames.default)("".concat(prefixCls, "-item-body"), {
      [(0, _concat.default)(_context = "".concat(prefixCls, "-item-body-")).call(_context, align)]: align
    });
    let body;

    if (header || main) {
      body = /*#__PURE__*/_react.default.createElement("div", {
        className: bodyCls
      }, header ? /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-body-header")
      }, header) : null, main ? /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-body-main")
      }, main) : null);
    }

    let content =
    /*#__PURE__*/
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-noninteractive-element-interactions
    _react.default.createElement("li", {
      className: itemCls,
      style: style,
      onClick: handleClick,
      onContextMenu: handleContextMenu,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave
    }, body ? body : null, children, extra ? /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-item-extra")
    }, extra) : null);

    if (this.context && contextGrid) {
      content = this.wrapWithGrid(content);
    }

    return content;
  }

}

exports.default = ListItem;
ListItem.contextType = _listContext.default;
ListItem.propTypes = {
  extra: _propTypes.default.node,
  header: _propTypes.default.node,
  main: _propTypes.default.node,
  align: _propTypes.default.oneOf(_constants.strings.ALIGN),
  className: _propTypes.default.string,
  children: _propTypes.default.node,
  style: _propTypes.default.object,
  onClick: _propTypes.default.func,
  onRightClick: _propTypes.default.func,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func
};
ListItem.defaultProps = {
  align: 'flex-start',
  onMouseEnter: _noop2.default,
  onMouseLeave: _noop2.default
};
