// @douyinfe/semi-foundation/lib/cjs/breadcrumb/foundation.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _foundation = _interopRequireDefault(require("../base/foundation"));

var _isEnterPress = _interopRequireDefault(require("../utils/isEnterPress"));

/* eslint-disable prefer-const */
class BreadcrumbFoundation extends _foundation.default {
  constructor(adapter) {
    super((0, _assign.default)({}, adapter));
  }

  handleClick(info, event) {
    this._adapter.notifyClick(info, event);
  }

  handleExpand(clickEvent) {
    this._adapter.expandCollapsed(clickEvent);
  }
  /**
   * A11y: simulate clear button click
   */


  handleExpandEnterPress(keyboardEvent) {
    if ((0, _isEnterPress.default)(keyboardEvent)) {
      this.handleExpand(keyboardEvent);
    }
  }

  genRoutes(routes) {
    return (0, _map.default)(routes).call(routes, route => {
      if (typeof route !== 'object') {
        return {
          name: route,
          _origin: {
            name: route
          }
        };
      }

      let config = {};
      config._origin = route;
      return (0, _assign.default)((0, _assign.default)({}, config), route);
    });
  }

}

exports.default = BreadcrumbFoundation;
