// @douyinfe/semi-foundation/lib/cjs/radio/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.radioGroupClasses = exports.radioClasses = exports.numbers = void 0;

var _constants = require("../base/constants");

const radioClasses = {
  PREFIX: "".concat(_constants.BASE_CLASS_PREFIX, "-radio"),
  INNER: "".concat(_constants.BASE_CLASS_PREFIX, "-radio-inner"),
  TEXT: "".concat(_constants.BASE_CLASS_PREFIX, "-radio-text"),
  INPUT: "".concat(_constants.BASE_CLASS_PREFIX, "-radio-input"),
  CHECKED: "".concat(_constants.BASE_CLASS_PREFIX, "-radio-checked"),
  DISABLED: "".concat(_constants.BASE_CLASS_PREFIX, "-radio-disabled"),
  BUTTON: "".concat(_constants.BASE_CLASS_PREFIX, "-radio-button")
};
exports.radioClasses = radioClasses;
const radioGroupClasses = {
  PREFIX: "".concat(_constants.BASE_CLASS_PREFIX, "-radioGroup"),
  INNER: "".concat(_constants.BASE_CLASS_PREFIX, "-radioGroup-inner"),
  TEXT: "".concat(_constants.BASE_CLASS_PREFIX, "-radioGroup-text"),
  INPUT: "".concat(_constants.BASE_CLASS_PREFIX, "-radioGroup-input"),
  CHECKED: "".concat(_constants.BASE_CLASS_PREFIX, "-radioGroup-checked"),
  DISABLED: "".concat(_constants.BASE_CLASS_PREFIX, "-radioGroup-disabled")
};
exports.radioGroupClasses = radioGroupClasses;
const strings = {
  DIRECTION_SET: ['horizontal', 'vertical'],
  DEFAULT_DIRECTION: 'horizontal',
  MODE: ['advanced', ''],
  TYPE_DEFAULT: 'default',
  TYPE_BUTTON: 'button',
  TYPE_CARD: 'card',
  TYPE_PURECARD: 'pureCard',
  BUTTON_SIZE: ['middle', 'small', 'large']
};
exports.strings = strings;
const numbers = {};
exports.numbers = numbers;
