// @douyinfe/semi-ui/lib/cjs/backtop/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireDefault(require("react"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/backtop/constants");

var _foundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/backtop/foundation"));

require("@douyinfe/semi-foundation/lib/cjs/backtop/backtop.css");

var _iconButton = _interopRequireDefault(require("../iconButton"));

var _semiIcons = require("@douyinfe/semi-icons");

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const prefixCls = _constants.cssClasses.PREFIX;

const getDefaultTarget = () => window;

class BackTop extends _baseComponent.default {
  constructor(props) {
    super(props);
    this.state = {
      visible: false
    };
    this.foundation = new _foundation.default(this.adapter);
  }

  componentDidMount() {
    this.foundation.init();
  }

  componentWillUnmount() {
    this.foundation.destroy();
  }

  get adapter() {
    return (0, _assign.default)((0, _assign.default)({}, super.adapter), {
      updateVisible: visible => {
        this.setState({
          visible
        });
      },
      notifyClick: e => {
        this.props.onClick && this.props.onClick(e);
      },
      targetIsWindow: target => target === window,
      isWindowUndefined: () => window === undefined,
      targetScrollToTop: (targetNode, scrollTop) => {
        if (targetNode === window) {
          document.body.scrollTop = scrollTop;
          document.documentElement.scrollTop = scrollTop;
        } else {
          targetNode.scrollTop = scrollTop;
        }
      }
    });
  }

  handleClick(e) {
    this.foundation.onClick(e);
  }

  renderDefault() {
    return /*#__PURE__*/_react.default.createElement(_iconButton.default, {
      theme: "light",
      icon: /*#__PURE__*/_react.default.createElement(_semiIcons.IconChevronUp, null)
    });
  }

  render() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _a = this.props,
          {
      children,
      className,
      style,
      onClick,
      visibilityHeight,
      target
    } = _a,
          others = __rest(_a, ["children", "className", "style", "onClick", "visibilityHeight", "target"]);

    const {
      visible
    } = this.state;
    const preCls = (0, _classnames.default)(prefixCls, className);
    const backtopBtn = children ? children : this.renderDefault();
    const content = visible ? /*#__PURE__*/_react.default.createElement("div", (0, _assign.default)({}, others, {
      className: preCls,
      style: style,
      onClick: e => this.handleClick(e),
      "x-semi-prop": "children"
    }), backtopBtn) : null;
    return content;
  }

}

exports.default = BackTop;
BackTop.defaultProps = {
  visibilityHeight: 400,
  target: getDefaultTarget,
  duration: 450
};
BackTop.propTypes = {
  target: _propTypes.default.func,
  visibilityHeight: _propTypes.default.number,
  duration: _propTypes.default.number,
  onClick: _propTypes.default.func,
  style: _propTypes.default.object,
  className: _propTypes.default.string
};
