// @douyinfe/semi-ui/lib/cjs/_portal/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/keys"));

var _react = require("react");

var _reactDom = require("react-dom");

var _constants = require("@douyinfe/semi-foundation/lib/cjs/base/constants");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _context = _interopRequireDefault(require("../configProvider/context"));

require("@douyinfe/semi-foundation/lib/cjs/_portal/portal.css");

const defaultGetContainer = () => document.body;

class Portal extends _react.PureComponent {
  constructor(props) {
    var _this;

    super(props);
    _this = this;

    this.addStyle = function () {
      let style = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      if (_this.el) {
        for (const key of (0, _keys.default)(style)) {
          _this.el.style[key] = style[key];
        }
      }
    };

    this.addClass = function (prefixCls) {
      const {
        direction
      } = _this.context;

      for (var _len = arguments.length, classNames = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        classNames[_key - 1] = arguments[_key];
      }

      const cls = (0, _classnames.default)(prefixCls, ...classNames, {
        ["".concat(prefixCls, "-rtl")]: direction === 'rtl'
      });

      if (_this.el) {
        _this.el.className = cls;
      }
    };

    try {
      this.el = document.createElement('div');
    } catch (e) {}

    this.state = {
      container: undefined
    };
  }

  componentDidMount() {
    if (!this.el) {
      this.el = document.createElement('div');
    }

    const {
      state,
      props,
      context
    } = this;
    const getContainer = props.getPopupContainer || context.getPopupContainer || defaultGetContainer;
    const container = getContainer();

    if (container !== state.container) {
      // const computedStyle = window.getComputedStyle(container);
      // if (computedStyle.position !== 'relative') {
      //    container.style.position = 'relative';
      // }
      container.appendChild(this.el);
      this.addStyle(props.style);
      this.addClass(props.prefixCls, props.className);
      this.setState({
        container
      });
    }
  }

  componentDidUpdate(prevProps) {
    // visible callback
    const {
      didUpdate
    } = this.props;

    if (didUpdate) {
      didUpdate(prevProps);
    }
  }

  componentWillUnmount() {
    const {
      container
    } = this.state;

    if (container) {
      container.removeChild(this.el);
    }
  }

  render() {
    const {
      state,
      props
    } = this;

    if (state.container) {
      return /*#__PURE__*/(0, _reactDom.createPortal)(props.children, this.el);
    }

    return null;
  }

}

Portal.contextType = _context.default;
Portal.defaultProps = {
  // getPopupContainer: () => document.body,
  prefixCls: "".concat(_constants.BASE_CLASS_PREFIX, "-portal")
};
Portal.propTypes = {
  children: _propTypes.default.node,
  prefixCls: _propTypes.default.string,
  getPopupContainer: _propTypes.default.func,
  className: _propTypes.default.string,
  didUpdate: _propTypes.default.func
};
var _default = Portal;
exports.default = _default;
