// @douyinfe/semi-ui/lib/cjs/skeleton/item.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.Title = exports.Paragraph = exports.Image = exports.Button = exports.Avatar = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/skeleton/constants");

var _constants2 = require("@douyinfe/semi-foundation/lib/cjs/avatar/constants");

require("@douyinfe/semi-foundation/lib/cjs/skeleton/skeleton.css");

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const sizeSet = _constants2.strings.SIZE;

const generator = type => BasicComponent => props => /*#__PURE__*/_react.default.createElement(BasicComponent, (0, _assign.default)({
  type: type
}, props));

class Generic extends _react.PureComponent {
  render() {
    var _context, _context2, _context3;

    const _a = this.props,
          {
      prefixCls,
      className,
      type,
      size
    } = _a,
          others = __rest(_a, ["prefixCls", "className", "type", "size"]);

    const classString = (0, _classnames.default)(className, (0, _concat.default)(_context = "".concat(prefixCls, "-")).call(_context, type), {
      [(0, _concat.default)(_context2 = (0, _concat.default)(_context3 = "".concat(prefixCls, "-")).call(_context3, type, "-")).call(_context2, size)]: type.toUpperCase() === 'AVATAR'
    });
    return /*#__PURE__*/_react.default.createElement('div', (0, _assign.default)({
      className: classString
    }, others));
  }

}

Generic.propTypes = {
  type: _propTypes.default.string,
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  size: _propTypes.default.oneOf(sizeSet)
};
Generic.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX,
  size: 'medium'
};
const Avatar = generator('avatar')(Generic);
exports.Avatar = Avatar;
const Image = generator('image')(Generic);
exports.Image = Image;
const Title = generator('title')(Generic);
exports.Title = Title;
const Button = generator('button')(Generic);
exports.Button = Button;

class Paragraph extends _react.PureComponent {
  render() {
    var _context4;

    const {
      prefixCls,
      className,
      style,
      rows
    } = this.props;
    const classString = (0, _classnames.default)(className, "".concat(prefixCls, "-paragraph"));
    return /*#__PURE__*/_react.default.createElement("ul", {
      className: classString,
      style: style
    }, (0, _map.default)(_context4 = [...Array(rows)]).call(_context4, (e, i) => /*#__PURE__*/_react.default.createElement("li", {
      key: i
    })));
  }

}

exports.Paragraph = Paragraph;
Paragraph.propTypes = {
  rows: _propTypes.default.number,
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string
};
Paragraph.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX,
  rows: 4
};
