// @douyinfe/semi-ui/lib/cjs/scrollList/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/scrollList/constants");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _scrollItem = _interopRequireDefault(require("./scrollItem"));

var _foundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/scrollList/foundation"));

require("@douyinfe/semi-foundation/lib/cjs/scrollList/scrollList.css");

// eslint-disable-next-line @typescript-eslint/ban-types
class ScrollList extends _baseComponent.default {
  constructor(props) {
    super(props);
    this.foundation = new _foundation.default(this.adapter);
  }

  render() {
    const {
      children,
      header,
      footer,
      prefixCls,
      bodyHeight,
      className,
      style
    } = this.props;
    const clsWrapper = (0, _classnames.default)(className, {
      [prefixCls || _constants.cssClasses.PREFIX]: true
    });
    const clsHeader = (0, _classnames.default)({
      ["".concat(prefixCls || _constants.cssClasses.PREFIX, "-header")]: true
    });
    return /*#__PURE__*/_react.default.createElement("div", {
      className: clsWrapper,
      style: style
    }, header ? /*#__PURE__*/_react.default.createElement("div", {
      className: clsHeader
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(clsHeader, "-title"),
      "x-semi-prop": this.props['x-semi-header-alias'] || "header"
    }, header), /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(clsWrapper, "-line")
    })) : null, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(clsWrapper, "-body"),
      style: {
        height: bodyHeight ? bodyHeight : ''
      },
      "x-semi-prop": "children"
    }, children), footer ? /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(clsWrapper, "-footer"),
      "x-semi-prop": this.props['x-semi-footer-alias'] || "footer"
    }, footer) : null);
  }

}

ScrollList.Item = _scrollItem.default;
ScrollList.propTypes = {
  className: _propTypes.default.string,
  header: _propTypes.default.node,
  footer: _propTypes.default.node,
  children: _propTypes.default.node,
  bodyHeight: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string])
};
var _default = ScrollList;
exports.default = _default;
