// @douyinfe/semi-ui/lib/cjs/sideSheet/SideSheetTransition.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _semiAnimationReact = require("@douyinfe/semi-animation-react");

var _react = _interopRequireDefault(require("react"));

// @ts-ignore  Currently there is no types definition for semi-animation-react;
// eslint-disable-next-line max-len
const formatStyles = function formatStyles() {
  let styles = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const {
    placement
  } = props;
  const {
    translate
  } = styles;
  const {
    opacity
  } = styles;
  let transform = '';

  switch (placement) {
    case 'left':
      transform = "translateX(-".concat(translate, "%)");
      break;

    case 'top':
      transform = "translateY(-".concat(translate, "%)");
      break;

    case 'right':
      transform = "translateX(".concat(translate, "%)");
      break;

    case 'bottom':
      transform = "translateY(".concat(translate, "%)");
      break;

    default:
      break;
  }

  return {
    transform,
    opacity
  };
};

class SideSheetTransition extends _react.default.PureComponent {
  render() {
    let {
      motion = {}
    } = this.props;
    const {
      children,
      controlled = false,
      visible
    } = this.props;

    if (typeof motion === 'function') {
      motion = motion(this.props);
    } else if (!motion || typeof motion !== 'object') {
      motion = {};
    }

    let extra = {};

    if (controlled) {
      extra = {
        state: visible ? 'enter' : 'leave'
      };
    }

    return /*#__PURE__*/_react.default.createElement(_semiAnimationReact.Transition, (0, _assign.default)({
      config: {
        tension: 170,
        friction: 14,
        easing: 'linear',
        duration: 200
      },
      from: {
        translate: 100,
        opacity: {
          val: 0,
          duration: 180
        }
      },
      enter: {
        translate: 0,
        opacity: {
          val: 1,
          duration: 180
        }
      },
      leave: {
        translate: 100,
        opacity: {
          val: 0,
          duration: 180
        }
      }
    }, extra, motion), typeof children === 'function' ? styles => children(formatStyles(styles, this.props)) : children);
  }

}

exports.default = SideSheetTransition;
