// @douyinfe/semi-ui/lib/cjs/sideSheet/SideSheetContent.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/sideSheet/constants");

var _iconButton = _interopRequireDefault(require("../iconButton"));

var _semiIcons = require("@douyinfe/semi-icons");

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

let uuid = 0;
const prefixCls = _constants.cssClasses.PREFIX;

class SideSheetContent extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    this.onMaskClick = e => {
      if (e.target === e.currentTarget) {
        this.close(e);
      }
    };

    this.close = e => {
      const {
        onClose
      } = this.props;
      onClose && onClose(e);
    };
  }

  componentDidMount() {
    this.sideSheetId = "sidesheet-".concat(uuid++);
  }

  componentWillUnmount() {
    clearTimeout(this.timeoutId);
  }

  getMaskElement() {
    const {
      mask,
      maskStyle,
      maskClosable
    } = this.props;

    if (mask) {
      return /*#__PURE__*/_react.default.createElement("div", {
        "aria-hidden": true,
        key: "mask",
        className: "".concat(prefixCls, "-mask"),
        style: maskStyle,
        onClick: maskClosable ? this.onMaskClick : null
      });
    }

    return null;
  }

  renderHeader() {
    const {
      title,
      closable,
      headerStyle
    } = this.props;
    let header, closer;

    if (title) {
      header = /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "-title"),
        "x-semi-prop": "title"
      }, this.props.title);
    }

    if (closable) {
      closer = /*#__PURE__*/_react.default.createElement(_iconButton.default, {
        className: "".concat(prefixCls, "-close"),
        key: "close-btn",
        onClick: this.close,
        type: "tertiary",
        icon: /*#__PURE__*/_react.default.createElement(_semiIcons.IconClose, null),
        theme: "borderless",
        size: "small"
      });
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-header"),
      role: 'heading',
      "aria-level": 1,
      style: (0, _assign.default)({}, headerStyle)
    }, header, closer);
  }

  getDialogElement() {
    var _context;

    const props = __rest(this.props, []);

    const style = {};

    if (props.width) {
      style.width = props.width; // When the mask is false, the width is set on the wrapper. At this time, sidesheet-inner does not need to set the width again, otherwise, the percentage will be accumulated repeatedly when the width is a percentage

      if (!props.mask) {
        style.width = '100%';
      }
    }

    if (props.height) {
      style.height = props.height;
    }

    const header = this.renderHeader();

    const dialogElement = /*#__PURE__*/_react.default.createElement("div", {
      key: "dialog-element",
      role: "dialog",
      tabIndex: -1,
      className: (0, _concat.default)(_context = "".concat(prefixCls, "-inner ")).call(_context, prefixCls, "-inner-wrap"),
      // onMouseDown={this.onDialogMouseDown}
      style: (0, _assign.default)((0, _assign.default)({}, props.style), style)
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-content")
    }, header, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-body"),
      style: props.bodyStyle,
      "x-semi-prop": "children"
    }, props.children), props.footer ? /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-footer"),
      "x-semi-prop": "footer"
    }, props.footer) : null));

    return dialogElement;
  }

  render() {
    const {
      mask,
      className,
      width
    } = this.props;
    const wrapperCls = (0, _classnames.default)(className, {
      ["".concat(prefixCls, "-fixed")]: !mask
    });
    const wrapperStyle = {};

    if (!mask && width) {
      wrapperStyle.width = width;
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: wrapperCls,
      style: wrapperStyle
    }, this.getMaskElement(), this.getDialogElement());
  }

}

exports.default = SideSheetContent;
SideSheetContent.propTypes = {
  onClose: _propTypes.default.func
};
SideSheetContent.defaultProps = {
  onClose: _noop2.default
};
