// @douyinfe/semi-ui/lib/cjs/timeline/index.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

require("@douyinfe/semi-foundation/lib/cjs/timeline/timeline.css");

var _constants = require("@douyinfe/semi-foundation/lib/cjs/timeline/constants");

var _context7 = _interopRequireDefault(require("../configProvider/context"));

var _item = _interopRequireDefault(require("./item"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const prefixCls = _constants.cssClasses.PREFIX;

class Timeline extends _react.PureComponent {
  constructor() {
    super(...arguments);

    this.getPosCls = (ele, idx) => {
      const {
        mode
      } = this.props;

      if (mode === 'alternate') {
        if (ele.props.position) {
          var _context;

          return (0, _concat.default)(_context = "".concat(prefixCls, "-item-")).call(_context, ele.props.position);
        }

        return idx % 2 === 0 ? "".concat(prefixCls, "-item-left") : "".concat(prefixCls, "-item-right");
      }

      if (mode === 'center') {
        if (ele.props.position) {
          var _context2;

          return (0, _concat.default)(_context2 = "".concat(prefixCls, "-item-")).call(_context2, ele.props.position);
        }

        return "".concat(prefixCls, "-item-left");
      }

      if (mode === 'left' || mode === 'right') {
        var _context3;

        return (0, _concat.default)(_context3 = "".concat(prefixCls, "-item-")).call(_context3, mode);
      }

      if (ele.props.position) {
        var _context4;

        return (0, _concat.default)(_context4 = "".concat(prefixCls, "-item-")).call(_context4, ele.props.position);
      }

      return '';
    };

    this.addClassName = items => {
      var _context5;

      return (0, _map.default)(_context5 = _react.default.Children).call(_context5, items, (ele, idx) => {
        if ( /*#__PURE__*/_react.default.isValidElement(ele)) {
          return /*#__PURE__*/_react.default.cloneElement(ele, {
            className: (0, _classnames.default)(ele.props.className, this.getPosCls(ele, idx))
          });
        }

        return ele;
      });
    };
  }

  render() {
    var _context6;

    const {
      children,
      className,
      style,
      mode,
      dataSource
    } = this.props;
    const classString = (0, _classnames.default)(prefixCls, className, {
      [(0, _concat.default)(_context6 = "".concat(prefixCls, "-")).call(_context6, mode)]: mode
    });
    let childrenList;

    if (dataSource && dataSource.length) {
      const items = (0, _map.default)(dataSource).call(dataSource, (item, index) => /*#__PURE__*/_react.default.createElement(_item.default, (0, _assign.default)({
        key: "timeline-item-".concat(index)
      }, item), item.content));
      childrenList = this.addClassName(items);
    }

    const items = childrenList || this.addClassName(children);
    return /*#__PURE__*/_react.default.createElement("ul", {
      "aria-label": this.props['aria-label'],
      style: style,
      className: classString
    }, items);
  }

}

Timeline.contextType = _context7.default;
Timeline.Item = _item.default;
Timeline.propTypes = {
  mode: _propTypes.default.oneOf(_constants.strings.MODE),
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  dataSource: _propTypes.default.array
};
Timeline.defaultProps = {
  mode: 'left'
};
var _default = Timeline;
exports.default = _default;
