// @douyinfe/semi-ui/lib/cjs/button/Button.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/includes"));

var _omit2 = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/button/constants");

require("@douyinfe/semi-foundation/lib/cjs/button/button.css");

var _function = require("@douyinfe/semi-foundation/lib/cjs/utils/function");

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};
/* eslint-disable react/destructuring-assignment */


const btnSizes = _constants.strings.sizes;
const {
  htmlTypes,
  btnTypes
} = _constants.strings; // TODO: icon configuration

class Button extends _react.PureComponent {
  render() {
    var _context;

    const _a = this.props,
          {
      children,
      block,
      htmlType,
      loading,
      circle,
      className,
      style,
      disabled,
      size,
      theme,
      type,
      prefixCls,
      iconPosition
    } = _a,
          attr = __rest(_a, ["children", "block", "htmlType", "loading", "circle", "className", "style", "disabled", "size", "theme", "type", "prefixCls", "iconPosition"]);

    const baseProps = (0, _assign.default)((0, _assign.default)({
      disabled
    }, (0, _omit2.default)(attr, ['x-semi-children-alias'])), {
      className: (0, _classnames.default)(prefixCls, {
        [(0, _concat.default)(_context = "".concat(prefixCls, "-")).call(_context, type)]: !disabled && type,
        ["".concat(prefixCls, "-disabled")]: disabled,
        ["".concat(prefixCls, "-size-large")]: size === 'large',
        ["".concat(prefixCls, "-size-small")]: size === 'small',
        // [`${prefixCls}-loading`]: loading,
        ["".concat(prefixCls, "-light")]: theme === 'light',
        ["".concat(prefixCls, "-block")]: block,
        ["".concat(prefixCls, "-circle")]: circle,
        ["".concat(prefixCls, "-borderless")]: theme === 'borderless'
      }, className),
      type: htmlType,
      'aria-disabled': disabled
    });
    const xSemiProps = {};

    if (!(className && (0, _includes.default)(className).call(className, '-with-icon'))) {
      xSemiProps['x-semi-prop'] = this.props['x-semi-children-alias'] || 'children';
    }

    return (
      /*#__PURE__*/
      // eslint-disable-next-line react/button-has-type
      _react.default.createElement("button", (0, _assign.default)({}, baseProps, {
        onClick: this.props.onClick,
        onMouseDown: this.props.onMouseDown,
        style: style
      }), /*#__PURE__*/_react.default.createElement("span", (0, _assign.default)({
        className: "".concat(prefixCls, "-content"),
        onClick: e => disabled && e.stopPropagation()
      }, xSemiProps), children))
    );
  }

}

exports.default = Button;
Button.defaultProps = {
  disabled: false,
  size: 'default',
  type: 'primary',
  theme: 'light',
  block: false,
  htmlType: 'button',
  onMouseDown: _function.noop,
  onClick: _function.noop,
  onMouseEnter: _function.noop,
  onMouseLeave: _function.noop,
  prefixCls: _constants.cssClasses.PREFIX
};
Button.propTypes = {
  children: _propTypes.default.node,
  disabled: _propTypes.default.bool,
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  size: _propTypes.default.oneOf(btnSizes),
  type: _propTypes.default.oneOf(btnTypes),
  block: _propTypes.default.bool,
  onClick: _propTypes.default.func,
  onMouseDown: _propTypes.default.func,
  circle: _propTypes.default.bool,
  loading: _propTypes.default.bool,
  htmlType: _propTypes.default.oneOf(htmlTypes),
  theme: _propTypes.default.oneOf(_constants.strings.themes),
  className: _propTypes.default.string,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func,
  'aria-label': _propTypes.default.string
};
