// @douyinfe/semi-ui/lib/cjs/toast/ToastTransition.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = ToastTransition;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _semiAnimationReact = require("@douyinfe/semi-animation-react");

var _react = _interopRequireDefault(require("react"));

// @ts-ignore  Currently there is no types for semi-animation-react
function ToastTransition() {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let {
    motion = {}
  } = props;

  if (typeof motion === 'function') {
    motion = motion(props);
  } else if (!motion || typeof motion !== 'object') {
    motion = {};
  }

  return /*#__PURE__*/_react.default.createElement(_semiAnimationReact.Transition // onFrame={style => console.log(style)}
  , (0, _assign.default)({
    // onFrame={style => console.log(style)}
    from: {
      translateY: -100,
      opacity: 0
    },
    enter: {
      translateY: {
        val: 0,
        tension: 560,
        friction: 32
      },
      opacity: {
        val: 1,
        duration: 300
      }
    },
    leave: {
      translateY: {
        val: -100,
        easing: 'easeOutCubic',
        duration: 300
      },
      opacity: {
        val: 0,
        duration: 200
      }
    }
  }, motion), typeof props.children === 'function' ? _ref => {
    let {
      translateY,
      opacity
    } = _ref;
    return props.children({
      transform: "translateY(".concat(translateY, "%)"),
      opacity
    });
  } : props.children);
}
