// @douyinfe/semi-foundation/lib/cjs/avatar/foundation.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _foundation = _interopRequireDefault(require("../base/foundation"));

var _warning = _interopRequireDefault(require("../utils/warning"));

class AvatarFoundation extends _foundation.default {
  constructor(adapter) {
    super((0, _assign.default)({}, adapter));

    this.handleFocusVisible = event => {
      const {
        target
      } = event;

      try {
        if (target.matches(':focus-visible')) {
          this._adapter.setFocusVisible(true);
        }
      } catch (error) {
        (0, _warning.default)(true, 'Warning: [Semi Avatar] The current browser does not support the focus-visible');
      }
    };

    this.handleBlur = () => {
      this._adapter.setFocusVisible(false);
    };
  }

  init() {} // eslint-disable-line


  destroy() {} // eslint-disable-line


  handleImgLoadError() {
    const {
      onError
    } = this.getProps();
    const errorFlag = onError ? onError() : undefined;

    if (errorFlag !== false) {
      this._adapter.notifyImgState(false);
    }
  }

  handleEnter(e) {
    this._adapter.notifyEnter(e);
  }

  handleLeave(e) {
    this._adapter.notifyLeave(e);
  }

}

exports.default = AvatarFoundation;
