// @douyinfe/semi-ui/lib/cjs/datePicker/insetInput.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.InsetDateInput = InsetDateInput;
exports.InsetTimeInput = InsetTimeInput;

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/includes"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = _interopRequireDefault(require("react"));

var _input = _interopRequireDefault(require("../input"));

function InsetDateInput(props) {
  const {
    insetInputValue,
    valuePath,
    onFocus,
    onChange,
    placeholder,
    forwardRef
  } = props;
  const value = (0, _get2.default)(insetInputValue, valuePath);
  return /*#__PURE__*/_react.default.createElement(_input.default, {
    value: value,
    onChange: (value, event) => {
      onChange({
        value,
        event,
        insetInputValue,
        valuePath
      });
    },
    onFocus: onFocus,
    placeholder: placeholder,
    ref: forwardRef
  });
}

function InsetTimeInput(props) {
  const {
    insetInputValue,
    valuePath,
    type,
    onFocus,
    onChange,
    placeholder,
    disabled
  } = props;

  const _isTimeType = (0, _includes.default)(type).call(type, 'Time');

  if (!_isTimeType) {
    return null;
  }

  const value = (0, _get2.default)(insetInputValue, valuePath);
  return /*#__PURE__*/_react.default.createElement(_input.default, {
    value: value,
    onChange: (value, event) => {
      onChange({
        value,
        event,
        insetInputValue,
        valuePath
      });
    },
    onFocus: onFocus,
    placeholder: placeholder,
    disabled: disabled
  });
}
