// @douyinfe/semi-ui/lib/cjs/steps/navSteps.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/steps/constants");

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const Steps = props => {
  var _context2;

  const {
    size,
    current,
    initial,
    children,
    prefixCls,
    className,
    style,
    onChange
  } = props;
  const inner = (0, _react.useMemo)(() => {
    var _context;

    const filteredChildren = (0, _filter.default)(_context = _react.Children.toArray(children)).call(_context, c => /*#__PURE__*/(0, _react.isValidElement)(c));
    const total = filteredChildren.length;
    const content = (0, _map.default)(_react.Children).call(_react.Children, filteredChildren, (child, index) => {
      if (!child) {
        return null;
      }

      const childProps = (0, _assign.default)({
        index,
        total
      }, child.props);
      childProps.active = index === current;

      childProps.onChange = () => {
        if (index !== current) {
          onChange(index + initial);
        }
      };

      return /*#__PURE__*/(0, _react.cloneElement)(child, (0, _assign.default)({}, childProps));
    });
    return content;
  }, [children, prefixCls, current, size, initial, onChange]);
  const wrapperCls = (0, _classnames.default)(className, {
    ["".concat(prefixCls, "-nav")]: true,
    [(0, _concat.default)(_context2 = "".concat(prefixCls, "-")).call(_context2, size)]: size !== 'default'
  });
  return /*#__PURE__*/_react.default.createElement("div", {
    "aria-label": props["aria-label"],
    className: wrapperCls,
    style: style
  }, inner);
};

Steps.propTypes = {
  prefixCls: _propTypes.default.string,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  current: _propTypes.default.number,
  initial: _propTypes.default.number,
  size: _propTypes.default.oneOf(['small', 'default'])
};
Steps.defaultProps = {
  prefixCls: _constants.stepsClasses.PREFIX,
  current: 0,
  direction: 'horizontal',
  size: 'default',
  initial: 0,
  status: 'process'
};
var _default = Steps;
exports.default = _default;
