// @douyinfe/semi-ui/lib/cjs/steps/navStep.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/steps/constants");

var _semiIcons = require("@douyinfe/semi-icons");

const NavStep = props => {
  const {
    prefixCls,
    className,
    title,
    style,
    active,
    index,
    total,
    onClick,
    onKeyDown,
    onChange
  } = props;
  const classString = (0, _classnames.default)(prefixCls, {
    ["".concat(prefixCls, "-active")]: active
  }, className);

  const handleClick = e => {
    if ((0, _isFunction2.default)(onClick)) {
      onClick(e);
    }

    onChange();
  };

  const handleKeyDown = e => {
    if (e.key === 'Enter') {
      if ((0, _isFunction2.default)(onKeyDown)) {
        onKeyDown(e);
      }

      onChange();
    }
  };

  return /*#__PURE__*/_react.default.createElement("div", {
    role: props["role"],
    "aria-label": props["aria-label"],
    "aria-current": "step",
    tabIndex: 0,
    className: classString,
    style: style,
    onClick: e => handleClick(e),
    onKeyDown: handleKeyDown
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefixCls, "-container")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefixCls, "-content")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefixCls, "-title")
  }, title)), index !== total - 1 && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefixCls, "-icon")
  }, /*#__PURE__*/_react.default.createElement(_semiIcons.IconChevronRight, {
    size: "small"
  }))));
};

NavStep.propTypes = {
  prefixCls: _propTypes.default.string,
  title: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  onClick: _propTypes.default.func,
  active: _propTypes.default.bool
};
NavStep.defaultProps = {
  prefixCls: _constants.stepsClasses.ITEM,
  active: false,
  className: ''
};
var _default = NavStep;
exports.default = _default;
