// @douyinfe/semi-ui/lib/cjs/form/field.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.FormUpload = exports.FormTreeSelect = exports.FormTimePicker = exports.FormTextArea = exports.FormTagInput = exports.FormSwitch = exports.FormSlider = exports.FormSelect = exports.FormRating = exports.FormRadioGroup = exports.FormRadio = exports.FormInputNumber = exports.FormInput = exports.FormDatePicker = exports.FormCheckboxGroup = exports.FormCheckbox = exports.FormCascader = exports.FormAutoComplete = void 0;

var _withField = _interopRequireDefault(require("./hoc/withField"));

var _index = _interopRequireDefault(require("../input/index"));

var _textarea = _interopRequireDefault(require("../input/textarea"));

var _index2 = _interopRequireDefault(require("../inputNumber/index"));

var _index3 = _interopRequireDefault(require("../select/index"));

var _index4 = require("../checkbox/index");

var _checkboxGroup = _interopRequireDefault(require("../checkbox/checkboxGroup"));

var _index5 = require("../radio/index");

var _radioGroup = _interopRequireDefault(require("../radio/radioGroup"));

var _index6 = _interopRequireDefault(require("../datePicker/index"));

var _index7 = _interopRequireDefault(require("../switch/index"));

var _index8 = _interopRequireDefault(require("../slider/index"));

var _index9 = _interopRequireDefault(require("../timePicker/index"));

var _index10 = _interopRequireDefault(require("../treeSelect/index"));

var _index11 = _interopRequireDefault(require("../cascader/index"));

var _index12 = _interopRequireDefault(require("../rating/index"));

var _index13 = _interopRequireDefault(require("../autoComplete/index"));

var _index14 = _interopRequireDefault(require("../upload/index"));

var _index15 = _interopRequireDefault(require("../tagInput/index"));

/* eslint-disable max-len */
// Basic component
const FormInput = (0, _withField.default)(_index.default, {
  maintainCursor: true
});
exports.FormInput = FormInput;
const FormInputNumber = (0, _withField.default)(_index2.default, {
  maintainCursor: true
});
exports.FormInputNumber = FormInputNumber;
const FormTextArea = (0, _withField.default)(_textarea.default, {
  maintainCursor: true
});
exports.FormTextArea = FormTextArea;
const FormSelect = (0, _withField.default)(_index3.default); // Select after withField is a new Component, without the Option attribute, it needs to be manually assigned once

exports.FormSelect = FormSelect;
FormSelect.Option = _index3.default.Option;
FormSelect.OptGroup = _index3.default.OptGroup;
const FormCheckboxGroup = (0, _withField.default)(_checkboxGroup.default);
exports.FormCheckboxGroup = FormCheckboxGroup;
const FormCheckbox = (0, _withField.default)(_index4.Checkbox, {
  valueKey: 'checked',
  valuePath: 'target.checked',
  shouldInject: false
});
exports.FormCheckbox = FormCheckbox;
const FormRadioGroup = (0, _withField.default)(_radioGroup.default, {
  valuePath: 'target.value'
});
exports.FormRadioGroup = FormRadioGroup;
const FormRadio = (0, _withField.default)(_index5.Radio, {
  valueKey: 'checked',
  valuePath: 'target.checked',
  shouldInject: false
});
exports.FormRadio = FormRadio;
const FormDatePicker = (0, _withField.default)(_index6.default);
exports.FormDatePicker = FormDatePicker;
const FormSwitch = (0, _withField.default)(_index7.default, {
  valueKey: 'checked'
});
exports.FormSwitch = FormSwitch;
const FormSlider = (0, _withField.default)(_index8.default);
exports.FormSlider = FormSlider;
const FormTimePicker = (0, _withField.default)(_index9.default);
exports.FormTimePicker = FormTimePicker;
const FormTreeSelect = (0, _withField.default)(_index10.default);
exports.FormTreeSelect = FormTreeSelect;
const FormCascader = (0, _withField.default)(_index11.default);
exports.FormCascader = FormCascader;
const FormRating = (0, _withField.default)(_index12.default);
exports.FormRating = FormRating;
const FormAutoComplete = (0, _withField.default)(_index13.default, {
  valueKey: 'value',
  onKeyChangeFnName: 'onChange'
});
exports.FormAutoComplete = FormAutoComplete;
const FormUpload = (0, _withField.default)(_index14.default, {
  valueKey: 'fileList',
  valuePath: 'fileList',
  onKeyChangeFnName: 'onChange'
});
exports.FormUpload = FormUpload;
const FormTagInput = (0, _withField.default)(_index15.default);
exports.FormTagInput = FormTagInput;
