// @douyinfe/semi-ui/lib/cjs/form/index.js
"use strict";

var _context;

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _forEachInstanceProperty = require("@babel/runtime-corejs3/core-js-stable/instance/for-each");

var _Object$keys = require("@babel/runtime-corejs3/core-js-stable/object/keys");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

var _exportNames = {
  Form: true,
  ArrayField: true,
  useFormApi: true,
  useFormState: true,
  useFieldState: true,
  useFieldApi: true,
  withField: true,
  withFormState: true,
  withFormApi: true
};

_Object$defineProperty(exports, "ArrayField", {
  enumerable: true,
  get: function () {
    return _arrayField.default;
  }
});

_Object$defineProperty(exports, "Form", {
  enumerable: true,
  get: function () {
    return _baseForm.default;
  }
});

_Object$defineProperty(exports, "useFieldApi", {
  enumerable: true,
  get: function () {
    return _index.useFieldApi;
  }
});

_Object$defineProperty(exports, "useFieldState", {
  enumerable: true,
  get: function () {
    return _index.useFieldState;
  }
});

_Object$defineProperty(exports, "useFormApi", {
  enumerable: true,
  get: function () {
    return _index.useFormApi;
  }
});

_Object$defineProperty(exports, "useFormState", {
  enumerable: true,
  get: function () {
    return _index.useFormState;
  }
});

_Object$defineProperty(exports, "withField", {
  enumerable: true,
  get: function () {
    return _withField.default;
  }
});

_Object$defineProperty(exports, "withFormApi", {
  enumerable: true,
  get: function () {
    return _withFormApi.default;
  }
});

_Object$defineProperty(exports, "withFormState", {
  enumerable: true,
  get: function () {
    return _withFormState.default;
  }
});

var _baseForm = _interopRequireDefault(require("./baseForm"));

var _arrayField = _interopRequireDefault(require("./arrayField"));

var _index = require("./hooks/index");

var _withField = _interopRequireDefault(require("./hoc/withField"));

var _withFormState = _interopRequireDefault(require("./hoc/withFormState"));

var _withFormApi = _interopRequireDefault(require("./hoc/withFormApi"));

var _interface = require("./interface");

_forEachInstanceProperty(_context = _Object$keys(_interface)).call(_context, function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;

  _Object$defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});
