// @douyinfe/semi-ui/lib/cjs/progress/index.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _parseInt2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/parse-int"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/progress/constants");

require("@douyinfe/semi-foundation/lib/cjs/progress/progress.css");

var _semiAnimation = require("@douyinfe/semi-animation");

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const prefixCls = _constants.cssClasses.PREFIX;

class Progress extends _react.Component {
  constructor(props) {
    super(props);
    this._mounted = true;
    this._mounted = true;
    this.state = {
      percentNumber: this.props.percent // Specially used for animation of numbers

    };
  }

  componentDidUpdate(prevProps) {
    if (isNaN(this.props.percent) || isNaN(prevProps.percent)) {
      throw new Error('[Semi Progress]:percent can not be NaN');
      return;
    }

    if (prevProps.percent !== this.props.percent) {
      if (!this.props.motion) {
        // eslint-disable-next-line
        this.setState({
          percentNumber: this.props.percent
        });
        return;
      }

      if (this.animation && this.animation.destroy) {
        this.animation.destroy();
      }

      this.animation = new _semiAnimation.Animation({
        from: {
          value: prevProps.percent
        },
        to: {
          value: this.props.percent
        }
      }, {
        // easing: 'cubic-bezier(0, .68, .3, 1)'
        easing: 'linear',
        duration: 300
      });
      this.animation.on('frame', props => {
        // prevent setState while component is unmounted but this timer is called
        if (this._mounted === false) {
          return;
        } // let percentNumber = Number.isInteger(props.value) ? props.value : Math.floor(props.value * 100) / 100;


        const percentNumber = (0, _parseInt2.default)(props.value);
        this.setState({
          percentNumber
        });
      });
      this.animation.on('rest', () => {
        // prevent setState while component is unmounted but this timer is called
        if (this._mounted === false) {
          return;
        }

        this.setState({
          percentNumber: this.props.percent
        });
      });
      this.animation.start();
    }
  }

  componentWillUnmount() {
    this.animation && this.animation.destroy();
    this._mounted = false;
  }

  renderCircleProgress() {
    var _context;

    const {
      strokeLinecap,
      style,
      className,
      strokeWidth,
      format,
      size,
      stroke,
      showInfo,
      percent,
      orbitStroke,
      id
    } = this.props;
    const ariaLabel = this.props['aria-label'];
    const ariaLabelledBy = this.props['aria-labelledby'];
    const ariaValueText = this.props['aria-valuetext'];
    const {
      percentNumber
    } = this.state;
    const classNames = {
      wrapper: (0, _classnames.default)("".concat(prefixCls, "-circle"), className),
      svg: (0, _classnames.default)("".concat(prefixCls, "-circle-ring")),
      circle: (0, _classnames.default)("".concat(prefixCls, "-circle-ring-inner"))
    };
    const perc = this.calcPercent(percent);
    const percNumber = this.calcPercent(percentNumber);
    let width;

    if (this.props.width) {
      width = this.props.width;
    } else {
      size === _constants.strings.DEFAULT_SIZE ? width = 72 : width = 24;
    } // cx, cy is circle center


    const cy = width / 2;
    const cx = width / 2;
    const radius = (width - strokeWidth) / 2; // radius

    const circumference = radius * 2 * Math.PI;
    const strokeDashoffset = (1 - perc / 100) * circumference; // Offset

    const strokeDasharray = (0, _concat.default)(_context = "".concat(circumference, " ")).call(_context, circumference);
    const text = format(percNumber);
    return /*#__PURE__*/_react.default.createElement("div", {
      id: id,
      className: classNames.wrapper,
      style: style,
      role: 'progressbar',
      "aria-valuemin": 0,
      "aria-valuemax": 100,
      "aria-valuenow": percNumber,
      "aria-labelledby": ariaLabelledBy,
      "aria-label": ariaLabel,
      "aria-valuetext": ariaValueText
    }, /*#__PURE__*/_react.default.createElement("svg", {
      key: size,
      className: classNames.svg,
      height: width,
      width: width,
      "aria-hidden": true
    }, /*#__PURE__*/_react.default.createElement("circle", {
      strokeDashoffset: 0,
      strokeWidth: strokeWidth,
      strokeDasharray: strokeDasharray,
      strokeLinecap: strokeLinecap,
      fill: "transparent",
      stroke: orbitStroke,
      r: radius,
      cx: cx,
      cy: cy,
      "aria-hidden": true
    }), /*#__PURE__*/_react.default.createElement("circle", {
      className: classNames.circle,
      strokeDashoffset: strokeDashoffset,
      strokeWidth: strokeWidth,
      strokeDasharray: strokeDasharray,
      strokeLinecap: strokeLinecap,
      fill: "transparent",
      stroke: stroke,
      r: radius,
      cx: cx,
      cy: cy,
      "aria-hidden": true
    })), showInfo && size !== 'small' ? /*#__PURE__*/_react.default.createElement("span", {
      className: "".concat(prefixCls, "-circle-text")
    }, text) : null);
  }

  calcPercent(percent) {
    let perc;

    if (percent > 100) {
      perc = 100;
    } else if (percent < 0) {
      perc = 0;
    } else {
      perc = percent;
    }

    return perc;
  }

  renderLineProgress() {
    const {
      className,
      style,
      stroke,
      direction,
      format,
      showInfo,
      size,
      percent,
      orbitStroke,
      id
    } = this.props;
    const ariaLabel = this.props['aria-label'];
    const ariaLabelledBy = this.props['aria-labelledby'];
    const ariaValueText = this.props['aria-valuetext'];
    const {
      percentNumber
    } = this.state;
    const progressWrapperCls = (0, _classnames.default)(prefixCls, className, {
      ["".concat(prefixCls, "-horizontal")]: direction === _constants.strings.DEFAULT_DIRECTION,
      ["".concat(prefixCls, "-vertical")]: direction !== _constants.strings.DEFAULT_DIRECTION,
      ["".concat(prefixCls, "-large")]: size === 'large'
    });
    const progressTrackCls = (0, _classnames.default)({
      ["".concat(prefixCls, "-track")]: true
    });
    const innerCls = (0, _classnames.default)("".concat(prefixCls, "-track-inner"));
    const perc = this.calcPercent(percent);
    const percNumber = this.calcPercent(percentNumber);
    const innerStyle = {
      background: stroke
    };

    if (direction === _constants.strings.DEFAULT_DIRECTION) {
      innerStyle.width = "".concat(perc, "%");
    } else {
      innerStyle.height = "".concat(perc, "%");
    }

    const text = format(percNumber);
    return /*#__PURE__*/_react.default.createElement("div", {
      id: id,
      className: progressWrapperCls,
      style: style,
      role: 'progressbar',
      "aria-valuemin": 0,
      "aria-valuemax": 100,
      "aria-valuenow": perc,
      "aria-labelledby": ariaLabelledBy,
      "aria-label": ariaLabel,
      "aria-valuetext": ariaValueText
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: progressTrackCls,
      style: orbitStroke ? {
        backgroundColor: orbitStroke
      } : {},
      "aria-hidden": true
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: innerCls,
      style: innerStyle,
      "aria-hidden": true
    })), showInfo ? /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-line-text")
    }, text) : null);
  }

  render() {
    const {
      type
    } = this.props;

    if (type === 'line') {
      return this.renderLineProgress();
    } else {
      return this.renderCircleProgress();
    }
  }

}

Progress.propTypes = {
  'aria-label': _propTypes.default.string,
  'aria-labelledby': _propTypes.default.string,
  'aria-valuetext': _propTypes.default.string,
  className: _propTypes.default.string,
  direction: _propTypes.default.oneOf(_constants.strings.directions),
  format: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.node]),
  id: _propTypes.default.string,
  motion: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.func, _propTypes.default.object]),
  orbitStroke: _propTypes.default.string,
  percent: _propTypes.default.number,
  scale: _propTypes.default.number,
  showInfo: _propTypes.default.bool,
  size: _propTypes.default.oneOf(_constants.strings.sizes),
  stroke: _propTypes.default.string,
  strokeLinecap: _propTypes.default.oneOf(_constants.strings.strokeLineCap),
  strokeWidth: _propTypes.default.number,
  style: _propTypes.default.object,
  type: _propTypes.default.oneOf(_constants.strings.types),
  width: _propTypes.default.number
};
Progress.defaultProps = {
  className: '',
  direction: _constants.strings.DEFAULT_DIRECTION,
  format: text => "".concat(text, "%"),
  motion: true,
  orbitStroke: 'var(--semi-color-fill-0)',
  percent: 0,
  showInfo: false,
  size: _constants.strings.DEFAULT_SIZE,
  stroke: 'var(--semi-color-success)',
  strokeLinecap: _constants.strings.DEFAULT_LINECAP,
  strokeWidth: 4,
  style: {},
  type: _constants.strings.DEFAULT_TYPE
};
var _default = Progress;
exports.default = _default;
