// @douyinfe/semi-ui/lib/cjs/notification/NoticeTransition.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = NoticeTransition;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _semiAnimationReact = require("@douyinfe/semi-animation-react");

var _react = _interopRequireDefault(require("react"));

// @ts-ignore  Currently there is no types for semi-animation-react
function NoticeTransition() {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let {
    motion = {}
  } = props;
  const {
    position = 'topRight'
  } = props;

  const formatStyle = function formatStyle(_ref) {
    let {
      translate,
      opacity
    } = _ref;
    let transform = "translateX(".concat(translate, "%)");

    if (position && typeof position === 'string') {
      if (/left/i.test(position)) {
        transform = "translateX(".concat(-translate, "%)");
      } else if (/right/i.test(position)) {
        transform = "translateX(".concat(translate, "%)");
      } else if (/top/i.test(position)) {
        transform = "translateY(".concat(-translate, "%)");
      } else {
        transform = "translateY(".concat(translate, "%)");
      }
    }

    return {
      transform,
      opacity
    };
  };

  if (typeof motion === 'function') {
    motion = motion(props);
  } else if (!motion || typeof motion !== 'object') {
    motion = {};
  }

  return /*#__PURE__*/_react.default.createElement(_semiAnimationReact.Transition // onFrame={style => console.log(formatStyle(style))}
  , (0, _assign.default)({
    // onFrame={style => console.log(formatStyle(style))}
    from: {
      translate: 100,
      opacity: 0
    },
    enter: {
      translate: {
        val: 0,
        tension: 560,
        friction: 32
      },
      opacity: {
        val: 1,
        duration: 200
      }
    },
    leave: {
      translate: {
        val: 100,
        easing: 'easeOutCubic',
        duration: 300
      },
      opacity: {
        val: 0,
        duration: 200
      }
    }
  }, motion), typeof props.children === 'function' ? transitionStyle => props.children(formatStyle(transitionStyle)) : props.children);
}
