// @douyinfe/semi-foundation/lib/es/sideSheet/sideSheetFoundation.js
import _noop from "lodash/noop";
import _get from "lodash/get";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import _bindInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/bind";
import BaseFoundation from '../base/foundation';
import KeyCode from '../utils/keyCode';
export default class SideSheetFoundation extends BaseFoundation {
  constructor(adapter) {
    var _this;

    super(_Object$assign(_Object$assign({}, SideSheetFoundation.defaultAdapter), adapter));
    _this = this;

    this.mergeMotionProp = (motion, prop, cb) => {
      const mergedMotion = typeof motion === 'undefined' || motion ? _Object$assign(_Object$assign({}, motion), {
        [prop]: function () {
          const curr = _get(motion, prop);

          if (typeof curr === 'function') {
            curr(...arguments);
          }

          cb();
        }
      }) : false;
      return mergedMotion;
    };

    this.getMergedMotion = () => {
      const {
        motion,
        visible,
        keepDOM
      } = this.getProps();
      let mergedMotion = this.mergeMotionProp(motion, 'didEnter', function () {
        const didEnter = _get(motion, 'didEnter');

        if (typeof didEnter === 'function') {
          didEnter(...arguments);
        }

        _this._adapter.notifyVisibleChange(visible);
      });
      mergedMotion = this.mergeMotionProp(mergedMotion, 'didLeave', function () {
        const didLeave = _get(motion, 'didLeave');

        if (typeof didLeave === 'function') {
          didLeave(...arguments);
        }

        _this._adapter.notifyVisibleChange(visible);
      });

      if (keepDOM) {
        var _context;

        mergedMotion = this.mergeMotionProp(mergedMotion, 'didLeave', _bindInstanceProperty(_context = this._adapter.toggleHidden).call(_context, this, true));
      }

      return mergedMotion;
    };
  }

  get defaultAdapter() {
    return {
      handleCancel: _noop,
      beforeShow: _noop,
      afterHide: _noop
    };
  }

  destroy() {
    this.afterHide();
  }

  handleCancel(e) {
    this._adapter.notifyCancel(e);
  }

  beforeShow() {
    const allowDisable = this.getProp('disableScroll');
    allowDisable && this._adapter.disabledBodyScroll();

    this._adapter.setOnKeyDownListener();
  }

  afterHide() {
    const allowDisable = this.getProp('disableScroll');
    allowDisable && this._adapter.enabledBodyScroll();

    this._adapter.removeKeyDownListener();
  }

  handleKeyDown(e) {
    const {
      closeOnEsc
    } = this.getProps();

    if (closeOnEsc && e.keyCode === KeyCode.ESC) {
      e.stopPropagation();
      this.handleCancel(e);
      return;
    }
  }

}
