// @douyinfe/semi-foundation/lib/cjs/carousel/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.numbers = exports.cssClasses = void 0;

var _constants = require("../base/constants");

const cssClasses = {
  CAROUSEL: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel"),
  CAROUSEL_INDICATOR: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-indicator"),
  CAROUSEL_INDICATOR_LINE: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-indicator-line"),
  CAROUSEL_INDICATOR_DOT: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-indicator-dot"),
  CAROUSEL_INDICATOR_COLUMNAR: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-indicator-columnar"),
  CAROUSEL_INDICATOR_INACTIVE: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-indicator-inactive"),
  CAROUSEL_INDICATOR_ACTIVE: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-indicator-active"),
  CAROUSEL_CONTENT: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-content"),
  CAROUSEL_ARROW: "".concat(_constants.BASE_CLASS_PREFIX, "-carousel-arrow")
};
exports.cssClasses = cssClasses;
const numbers = {
  DEFAULT_ACTIVE_INDEX: 0,
  DEFAULT_INTERVAL: 2000,
  DEFAULT_SPEED: 300
};
exports.numbers = numbers;
const strings = {
  ANIMATION_MAP: ['slide', 'fade'],
  DIRECTION: ['left', 'right'],
  TYPE_MAP: ['columnar', 'line', 'dot'],
  THEME_MAP: ['dark', 'primary', 'light'],
  POSITION_MAP: ['left', 'center', 'right'],
  ARROW_MAP: ['always', 'hover'],
  SIZE: ['small', 'medium'],
  TRIGGER: ['click', 'hover']
};
exports.strings = strings;
