// @douyinfe/semi-foundation/lib/cjs/carousel/foundation.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _setInterval2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set-interval"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _isObject2 = _interopRequireDefault(require("lodash/isObject"));

var _foundation = _interopRequireDefault(require("../base/foundation"));

var _constants = require("./constants");

class CarouselFoundation extends _foundation.default {
  constructor(adapter) {
    super((0, _assign.default)({}, adapter));
    this._interval = null;
  }

  play(interval) {
    if (this._interval) {
      clearInterval(this._interval);
    }

    this._interval = (0, _setInterval2.default)(() => {
      this.next();
    }, interval);
  }

  stop() {
    if (this._interval) {
      clearInterval(this._interval);
    }
  }

  goTo(activeIndex) {
    const {
      activeIndex: stateActiveIndex
    } = this.getStates();
    const targetIndex = this.getValidIndex(activeIndex);

    this._adapter.setIsReverse(stateActiveIndex > targetIndex);

    if (this.getIsControledComponent()) {
      this._notifyChange(targetIndex);
    } else {
      this._notifyChange(targetIndex);

      this.handleNewActiveIndex(targetIndex);
    }
  }

  next() {
    const {
      activeIndex: stateActiveIndex
    } = this.getStates();
    const targetIndex = this.getValidIndex(stateActiveIndex + 1);

    this._adapter.setIsReverse(false);

    if (this.getIsControledComponent()) {
      this._notifyChange(targetIndex);
    } else {
      this._notifyChange(targetIndex);

      this.handleNewActiveIndex(targetIndex);
    }
  }

  prev() {
    const {
      activeIndex: stateActiveIndex
    } = this.getStates();
    const targetIndex = this.getValidIndex(stateActiveIndex - 1);

    this._adapter.setIsReverse(true);

    if (this.getIsControledComponent()) {
      this._notifyChange(targetIndex);
    } else {
      this._notifyChange(targetIndex);

      this.handleNewActiveIndex(targetIndex);
    }
  }

  destroy() {
    this._unregisterInterval();
  }

  _unregisterInterval() {
    if (this._interval) {
      clearInterval(this._interval);
      this._interval = null;
    }
  }

  _notifyChange(activeIndex) {
    const {
      activeIndex: stateActiveIndex,
      isInit
    } = this.getStates();

    if (isInit) {
      this._adapter.setIsInit(false);
    }

    if (stateActiveIndex !== activeIndex) {
      this._adapter.setPreActiveIndex(stateActiveIndex);

      this._adapter.notifyChange(activeIndex, stateActiveIndex);
    }
  }

  getValidIndex(index) {
    const {
      children
    } = this.getStates();
    return (index + children.length) % children.length;
  }

  getSwitchingTime() {
    const {
      autoPlay,
      speed
    } = this.getProps();
    const autoPlayType = typeof autoPlay;

    if (autoPlayType === 'boolean' && autoPlay) {
      return _constants.numbers.DEFAULT_INTERVAL + speed;
    }

    if ((0, _isObject2.default)(autoPlay)) {
      return (0, _get2.default)(autoPlay, 'interval', _constants.numbers.DEFAULT_INTERVAL) + speed;
    }

    return speed;
  }

  getIsControledComponent() {
    return this._isInProps('activeIndex');
  }

  handleAutoPlay() {
    const {
      autoPlay
    } = this.getProps();
    const autoPlayType = typeof autoPlay;

    if (autoPlayType === 'boolean' && autoPlay || (0, _isObject2.default)(autoPlay)) {
      this.play(this.getSwitchingTime());
    }
  }

  handleKeyDown(event) {
    if (event.key === 'ArrowLeft') {
      this.prev();
    }

    if (event.key === 'ArrowRight') {
      this.next();
    }
  }

  onIndicatorChange(activeIndex) {
    const {
      activeIndex: stateActiveIndex
    } = this.getStates();

    this._adapter.setIsReverse(stateActiveIndex > activeIndex);

    this._notifyChange(activeIndex);

    if (!this.getIsControledComponent()) {
      this.handleNewActiveIndex(activeIndex);
    }
  }

  handleNewActiveIndex(activeIndex) {
    const {
      activeIndex: stateActiveIndex
    } = this.getStates();

    if (stateActiveIndex !== activeIndex) {
      this._adapter.setNewActiveIndex(activeIndex);
    }
  }

  getDefaultActiveIndex() {
    let activeIndex;
    const props = this.getProps();

    if ('activeIndex' in props) {
      activeIndex = props.activeIndex;
    } else if ('defaultActiveIndex' in props) {
      activeIndex = props.defaultActiveIndex;
    }

    return activeIndex;
  }

}

var _default = CarouselFoundation;
exports.default = _default;
