// @douyinfe/semi-ui/lib/cjs/notification/useNotification/index.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = useNotification;

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _entries = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/entries"));

var _isArray = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/array/is-array"));

var _map2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/map"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/notification/constants");

var _HookNotice = _interopRequireDefault(require("./HookNotice"));

require("@douyinfe/semi-foundation/lib/cjs/notification/notification.css");

var _uuid = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/utils/uuid"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// TODO: Automatic folding + unfolding function when there are more than N
const defaultConfig = {
  duration: 3,
  position: 'topRight',
  motion: true,
  content: '',
  title: '',
  zIndex: 1010
};

function usePatchElement() {
  const [elements, setElements] = (0, _react.useState)([]);

  function patchElement(element, config) {
    setElements(originElements => [{
      element,
      config
    }, ...originElements]);
    return id => {
      setElements(originElements => (0, _filter.default)(originElements).call(originElements, _ref => {
        let {
          config: configOfCurrentElement
        } = _ref;
        return configOfCurrentElement.id !== id;
      }));
    };
  }

  function renderList() {
    var _context;

    const noticesInPosition = {
      top: [],
      topLeft: [],
      topRight: [],
      bottom: [],
      bottomLeft: [],
      bottomRight: []
    };
    (0, _forEach.default)(elements).call(elements, _ref2 => {
      let {
        element,
        config
      } = _ref2;
      const {
        position
      } = config;
      noticesInPosition[position].push(element);
    });
    return (0, _map.default)(_context = (0, _entries.default)(noticesInPosition)).call(_context, obj => {
      const pos = obj[0];
      const notices = obj[1]; // @ts-ignore

      return (0, _isArray.default)(notices) && notices.length ? /*#__PURE__*/_react.default.createElement("div", {
        key: pos,
        className: (0, _classnames.default)(_constants.cssClasses.LIST),
        placement: pos
      }, notices) : null;
    });
  }

  return [renderList(), patchElement];
}

function useNotification() {
  const [elements, patchElement] = usePatchElement();
  const noticeRef = new _map2.default();

  const addNotice = config => {
    const id = (0, _uuid.default)('semi_notice_');
    const mergeConfig = (0, _assign.default)((0, _assign.default)({}, config), {
      id
    }); // eslint-disable-next-line prefer-const

    let closeFunc;

    const ref = ele => {
      noticeRef.set(id, ele);
    };

    const notice = /*#__PURE__*/_react.default.createElement(_HookNotice.default, (0, _assign.default)({
      key: id
    }, mergeConfig, {
      afterClose: instanceID => closeFunc(instanceID),
      ref: ref
    }));

    closeFunc = patchElement(notice, (0, _assign.default)({}, mergeConfig));
    return id;
  };

  const removeElement = instanceID => {
    const ele = noticeRef.get(instanceID);
    ele && ele.close();
  };

  return [{
    success: config => addNotice((0, _assign.default)((0, _assign.default)((0, _assign.default)({}, defaultConfig), config), {
      type: 'success'
    })),
    info: config => addNotice((0, _assign.default)((0, _assign.default)((0, _assign.default)({}, defaultConfig), config), {
      type: 'info'
    })),
    error: config => addNotice((0, _assign.default)((0, _assign.default)((0, _assign.default)({}, defaultConfig), config), {
      type: 'error'
    })),
    warning: config => addNotice((0, _assign.default)((0, _assign.default)((0, _assign.default)({}, defaultConfig), config), {
      type: 'warning'
    })),
    open: config => addNotice((0, _assign.default)((0, _assign.default)((0, _assign.default)({}, defaultConfig), config), {
      type: 'default'
    })),
    close: removeElement
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, elements)];
}
