// @douyinfe/semi-ui/lib/es/calendar/monthCalendar.js
import _isEqual from "lodash/isEqual";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _filterInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/filter";
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _bindInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/bind";
import _Map from "@babel/runtime-corejs3/core-js-stable/map";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";

/* eslint-disable jsx-a11y/no-noninteractive-element-to-interactive-role */
import React from 'react';
import ReactDOM from 'react-dom';
import cls from 'classnames';
import PropTypes from 'prop-types';
import { IconClose } from '@douyinfe/semi-icons'; // eslint-disable-next-line max-len

import CalendarFoundation from '@douyinfe/semi-foundation/lib/es/calendar/foundation';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/calendar/constants';
import LocaleConsumer from '../locale/localeConsumer';
import localeContext from '../locale/context';
import BaseComponent from '../_base/baseComponent';
import Popover from '../popover';
import Button from '../iconButton';
import '@douyinfe/semi-foundation/lib/es/calendar/calendar.css';

const toPercent = num => {
  const res = num < 1 ? num * 100 : 100;
  return "".concat(res, "%");
};

const prefixCls = "".concat(cssClasses.PREFIX, "-month");
const contentPadding = 60;
const contentHeight = 24;
export default class monthCalendar extends BaseComponent {
  constructor(props) {
    var _this, _context4;

    super(props);
    _this = this;

    this.calcItemLimit = () => {
      this.contentCellHeight = this.cellDom.current.getBoundingClientRect().height;
      return Math.max(0, Math.ceil((this.contentCellHeight - contentPadding) / contentHeight));
    };

    this.handleClick = (e, val) => {
      const {
        onClick
      } = this.props;
      const value = this.foundation.formatCbValue(val);
      onClick && onClick(e, value);
    };

    this.showCard = (e, key) => {
      this.foundation.showCard(e, key);
    };

    this.renderHeader = dateFnsLocale => {
      var _context;

      const {
        markWeekend,
        displayValue
      } = this.props;
      this.monthlyData = this.foundation.getMonthlyData(displayValue, dateFnsLocale);
      return /*#__PURE__*/React.createElement("div", {
        className: "".concat(prefixCls, "-header"),
        role: "presentation"
      }, /*#__PURE__*/React.createElement("div", {
        role: "presentation",
        className: "".concat(prefixCls, "-grid")
      }, /*#__PURE__*/React.createElement("ul", {
        role: "row",
        className: "".concat(prefixCls, "-grid-row")
      }, _mapInstanceProperty(_context = this.monthlyData[0]).call(_context, day => {
        const {
          weekday
        } = day;
        const listCls = cls({
          ["".concat(cssClasses.PREFIX, "-weekend")]: markWeekend && day.isWeekend
        });
        return /*#__PURE__*/React.createElement("li", {
          role: "columnheader",
          "aria-label": weekday,
          key: "".concat(weekday, "-monthheader"),
          className: listCls
        }, /*#__PURE__*/React.createElement("span", null, weekday));
      }))));
    };

    this.renderEvents = events => {
      if (!events) {
        return undefined;
      }

      const list = _mapInstanceProperty(events).call(events, (event, ind) => {
        var _context2;

        const {
          leftPos,
          width,
          topInd,
          key,
          children
        } = event;
        const style = {
          left: toPercent(leftPos),
          width: toPercent(width),
          top: "".concat(topInd, "em")
        };
        return /*#__PURE__*/React.createElement("li", {
          className: _concatInstanceProperty(_context2 = "".concat(cssClasses.PREFIX, "-event-item ")).call(_context2, cssClasses.PREFIX, "-event-month"),
          key: key || "".concat(ind, "-monthevent"),
          style: style
        }, children);
      });

      return list;
    };

    this.renderCollapsed = (events, itemInfo, listCls, month) => {
      const {
        itemLimit,
        showCard
      } = this.state;
      const {
        weekday,
        dayString,
        date
      } = itemInfo;
      const key = date.toString();
      const remained = _filterInstanceProperty(events).call(events, i => Boolean(i)).length - itemLimit;
      const cardCls = "".concat(prefixCls, "-event-card"); // const top = contentPadding / 2 + this.state.itemLimit * contentHeight;

      const shouldRenderCard = remained > 0;
      const closer = /*#__PURE__*/React.createElement(Button, {
        className: "".concat(cardCls, "-close"),
        onClick: e => this.closeCard(e, key),
        type: "tertiary",
        icon: /*#__PURE__*/React.createElement(IconClose, null),
        theme: "borderless",
        size: "small"
      });
      const header = /*#__PURE__*/React.createElement("div", {
        className: "".concat(cardCls, "-header-info")
      }, /*#__PURE__*/React.createElement("div", {
        className: "".concat(cardCls, "-header-info-weekday")
      }, weekday), /*#__PURE__*/React.createElement("div", {
        className: "".concat(cardCls, "-header-info-date")
      }, dayString));
      const content = /*#__PURE__*/React.createElement("div", {
        className: cardCls
      }, /*#__PURE__*/React.createElement("div", {
        className: "".concat(cardCls, "-content")
      }, /*#__PURE__*/React.createElement("div", {
        className: "".concat(cardCls, "-header")
      }, header, closer), /*#__PURE__*/React.createElement("div", {
        className: "".concat(cardCls, "-body")
      }, /*#__PURE__*/React.createElement("ul", {
        className: "".concat(cardCls, "-list")
      }, _mapInstanceProperty(events).call(events, item => /*#__PURE__*/React.createElement("li", {
        key: item.key || "".concat(item.start.toString(), "-event")
      }, item.children))))));
      const pos = showCard && showCard[key] ? showCard[key][1] : 'leftTopOver';
      const text = /*#__PURE__*/React.createElement(LocaleConsumer, {
        componentName: "Calendar"
      }, locale =>
      /*#__PURE__*/
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      React.createElement("div", {
        className: "".concat(cardCls, "-wrapper"),
        style: {
          bottom: 0
        },
        onClick: e => this.showCard(e, key)
      }, locale.remaining.replace('${remained}', String(remained))));
      return /*#__PURE__*/React.createElement(Popover, {
        key: "".concat(date.valueOf()),
        content: content,
        position: pos,
        trigger: "custom",
        visible: showCard && showCard[key] && showCard[key][0],
        ref: ref => this.cardRef.set(key, ref)
      }, /*#__PURE__*/React.createElement("li", {
        key: date,
        className: listCls,
        onClick: e => this.handleClick(e, [date])
      }, this.formatDayString(month, dayString), shouldRenderCard ? text : null, this.renderCusDateGrid(date)));
    };

    this.formatDayString = (month, date) => {
      if (date === '1') {
        return /*#__PURE__*/React.createElement(LocaleConsumer, {
          componentName: "Calendar"
        }, (locale, localeCode) => /*#__PURE__*/React.createElement("span", {
          className: "".concat(prefixCls, "-date")
        }, month, /*#__PURE__*/React.createElement("span", {
          className: "".concat(cssClasses.PREFIX, "-today-date")
        }, "\u00A0", date), locale.datestring));
      }

      return (
        /*#__PURE__*/
        // eslint-disable-next-line max-len
        React.createElement("span", {
          className: "".concat(prefixCls, "-date")
        }, /*#__PURE__*/React.createElement("span", {
          className: "".concat(cssClasses.PREFIX, "-today-date")
        }, date))
      );
    };

    this.renderCusDateGrid = date => {
      const {
        dateGridRender
      } = this.props;

      if (!dateGridRender) {
        return null;
      }

      return dateGridRender(date.toString(), date);
    };

    this.renderWeekRow = function (index, weekDay) {
      let events = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      const {
        markWeekend
      } = _this.props;
      const {
        itemLimit
      } = _this.state;
      const {
        display,
        day
      } = events;
      return /*#__PURE__*/React.createElement("div", {
        role: "presentation",
        className: "".concat(prefixCls, "-weekrow"),
        ref: _this.cellDom,
        key: "".concat(index, "-weekrow")
      }, /*#__PURE__*/React.createElement("ul", {
        role: "row",
        className: "".concat(prefixCls, "-skeleton")
      }, _mapInstanceProperty(weekDay).call(weekDay, each => {
        const {
          date,
          dayString,
          isToday,
          isSameMonth,
          isWeekend,
          month,
          ind
        } = each;
        const listCls = cls({
          ["".concat(cssClasses.PREFIX, "-today")]: isToday,
          ["".concat(cssClasses.PREFIX, "-weekend")]: markWeekend && isWeekend,
          ["".concat(prefixCls, "-same")]: isSameMonth
        });
        const shouldRenderCollapsed = Boolean(day && day[ind] && day[ind].length > itemLimit);
        const inner = /*#__PURE__*/React.createElement("li", {
          role: "gridcell",
          "aria-label": date.toLocaleDateString(),
          "aria-current": isToday ? "date" : false,
          key: "".concat(date, "-weeksk"),
          className: listCls,
          onClick: e => _this.handleClick(e, [date])
        }, _this.formatDayString(month, dayString), _this.renderCusDateGrid(date));

        if (!shouldRenderCollapsed) {
          return inner;
        }

        return _this.renderCollapsed(day[ind], each, listCls, month);
      })), /*#__PURE__*/React.createElement("ul", {
        className: "".concat(cssClasses.PREFIX, "-event-items")
      }, display ? _this.renderEvents(display) : null));
    };

    this.renderMonthGrid = () => {
      var _context3;

      const {
        parsedEvents
      } = this.state;
      return /*#__PURE__*/React.createElement("div", {
        role: "presentation",
        className: "".concat(prefixCls, "-week")
      }, /*#__PURE__*/React.createElement("ul", {
        role: "presentation",
        className: "".concat(prefixCls, "-grid-col")
      }, _mapInstanceProperty(_context3 = _Object$keys(this.monthlyData)).call(_context3, weekInd => this.renderWeekRow(weekInd, this.monthlyData[weekInd], parsedEvents[weekInd]))));
    };

    this.state = {
      itemLimit: 0,
      showCard: {},
      parsedEvents: {},
      cachedKeys: []
    };
    this.cellDom = /*#__PURE__*/React.createRef();
    this.foundation = new CalendarFoundation(this.adapter);
    this.handleClick = _bindInstanceProperty(_context4 = this.handleClick).call(_context4, this);
    this.cardRef = new _Map();
  }

  get adapter() {
    return _Object$assign(_Object$assign({}, super.adapter), {
      registerClickOutsideHandler: (key, cb) => {
        const clickOutsideHandler = e => {
          const cardInstance = this.cardRef && this.cardRef.get(key); // eslint-disable-next-line react/no-find-dom-node

          const cardDom = ReactDOM.findDOMNode(cardInstance);

          if (cardDom && !cardDom.contains(e.target)) {
            cb();
          }
        };

        this.clickOutsideHandler = clickOutsideHandler;
        document.addEventListener('mousedown', clickOutsideHandler, false);
      },
      unregisterClickOutsideHandler: () => {
        document.removeEventListener('mousedown', this.clickOutsideHandler, false);
      },
      setMonthlyData: data => {
        this.monthlyData = data;
      },
      getMonthlyData: () => this.monthlyData,
      notifyClose: (e, key) => {
        const updates = {};
        updates[key] = [false];
        this.setState(prevState => ({
          showCard: _Object$assign(_Object$assign({}, prevState.showCard), updates)
        }));
        this.props.onClose && this.props.onClose(e);
      },
      openCard: (key, spacing) => {
        const updates = {};
        const pos = spacing ? 'leftTopOver' : 'rightTopOver';
        updates[key] = [true, pos];
        this.setState(prevState => ({
          showCard: _Object$assign({}, updates)
        }));
      },
      setParsedEvents: parsedEvents => {
        this.setState({
          parsedEvents: parsedEvents
        });
      },
      setItemLimit: itemLimit => {
        this.setState({
          itemLimit
        });
      },
      cacheEventKeys: cachedKeys => {
        this.setState({
          cachedKeys
        });
      }
    });
  }

  componentDidMount() {
    this.foundation.init();
    const itemLimit = this.calcItemLimit();
    this.foundation.parseMonthlyEvents(itemLimit);
  }

  componentWillUnmount() {
    this.foundation.destroy();
  }

  componentDidUpdate(prevProps, prevState) {
    var _context5;

    const prevEventKeys = prevState.cachedKeys;

    const nowEventKeys = _mapInstanceProperty(_context5 = this.props.events).call(_context5, event => event.key);

    let itemLimitUpdate = false;
    let {
      itemLimit
    } = this.state;

    if (prevProps.height !== this.props.height) {
      itemLimit = this.calcItemLimit();

      if (prevState.itemLimit !== itemLimit) {
        itemLimitUpdate = true;
      }
    }

    if (!_isEqual(prevEventKeys, nowEventKeys) || itemLimitUpdate) {
      this.foundation.parseMonthlyEvents(itemLimit || this.props.events);
    }
  }

  closeCard(e, key) {
    this.foundation.closeCard(e, key);
  }

  render() {
    const {
      className,
      height,
      width,
      style,
      header
    } = this.props;
    const monthCls = cls(prefixCls, className);

    const monthStyle = _Object$assign({
      height,
      width
    }, style);

    return /*#__PURE__*/React.createElement(LocaleConsumer, {
      componentName: "Calendar"
    }, (locale, localeCode, dateFnsLocale) => /*#__PURE__*/React.createElement("div", {
      role: "grid",
      className: monthCls,
      key: this.state.itemLimit,
      style: monthStyle
    }, /*#__PURE__*/React.createElement("div", {
      role: "presentation",
      className: "".concat(prefixCls, "-sticky-top")
    }, header, this.renderHeader(dateFnsLocale)), /*#__PURE__*/React.createElement("div", {
      role: "presentation",
      className: "".concat(prefixCls, "-grid-wrapper")
    }, this.renderMonthGrid())));
  }

}
monthCalendar.propTypes = {
  displayValue: PropTypes.instanceOf(Date),
  header: PropTypes.node,
  events: PropTypes.array,
  mode: PropTypes.string,
  markWeekend: PropTypes.bool,
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  style: PropTypes.object,
  className: PropTypes.string,
  dateGridRender: PropTypes.func,
  onClick: PropTypes.func,
  onClose: PropTypes.func
};
monthCalendar.defaultProps = {
  displayValue: new Date(),
  events: [],
  mode: 'month'
};
monthCalendar.contextType = localeContext;
