// @douyinfe/semi-ui/lib/es/calendar/weekCalendar.js
import _isEqual from "lodash/isEqual";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _concatInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/concat";
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _Map from "@babel/runtime-corejs3/core-js-stable/map";
import _bindInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/bind";
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import React from 'react';
import cls from 'classnames';
import PropTypes from 'prop-types';
import CalendarFoundation from '@douyinfe/semi-foundation/lib/es/calendar/foundation';
import LocaleConsumer from '../locale/localeConsumer';
import localeContext from '../locale/context';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/calendar/constants';
import BaseComponent from '../_base/baseComponent';
import DayCol from './dayCol';
import TimeCol from './timeCol';
import { calcRowHeight } from '@douyinfe/semi-foundation/lib/es/calendar/eventUtil';
import '@douyinfe/semi-foundation/lib/es/calendar/calendar.css';

const toPercent = num => {
  const res = num < 1 ? num * 100 : 100;
  return "".concat(res, "%");
};

const prefixCls = "".concat(cssClasses.PREFIX, "-week");
const allDayCls = "".concat(cssClasses.PREFIX, "-all-day");
export default class WeekCalendar extends BaseComponent {
  constructor(props) {
    var _context8;

    super(props);

    this.checkWeekend = val => this.foundation.checkWeekend(val);

    this.handleClick = (e, val) => {
      const {
        onClick
      } = this.props;
      const value = this.foundation.formatCbValue(val);
      onClick && onClick(e, value);
    };

    this.renderDayGrid = () => {
      const {
        parsedEvents
      } = this.state;
      const events = parsedEvents.day;
      const {
        week
      } = this.weeklyData;
      const {
        markWeekend,
        dateGridRender
      } = this.props;

      const inner = _mapInstanceProperty(week).call(week, day => {
        const dateString = day.date.toString();
        const dayEvents = events.has(dateString) ? events.get(dateString) : [];
        const parsed = this.foundation.getParseDailyEvents(dayEvents, day.date);
        return /*#__PURE__*/React.createElement(DayCol, {
          key: "".concat(dateString, "-weekday"),
          displayValue: day.date,
          scrollHeight: this.state.scrollHeight,
          handleClick: this.handleClick,
          events: parsed.day,
          showCurrTime: this.props.showCurrTime,
          isWeekend: markWeekend && day.isWeekend,
          dateGridRender: dateGridRender
        });
      });

      return inner;
    };

    this.renderHeader = dateFnsLocale => {
      var _context, _context2;

      const {
        markWeekend,
        displayValue
      } = this.props;
      const {
        month,
        week
      } = this.foundation.getWeeklyData(displayValue, dateFnsLocale);
      return /*#__PURE__*/React.createElement("div", {
        className: "".concat(prefixCls, "-header")
      }, /*#__PURE__*/React.createElement("ul", {
        className: _concatInstanceProperty(_context = _concatInstanceProperty(_context2 = "".concat(cssClasses.PREFIX, "-tag ")).call(_context2, prefixCls, "-tag ")).call(_context, prefixCls, "-sticky-left")
      }, /*#__PURE__*/React.createElement("span", null, month)), /*#__PURE__*/React.createElement("div", {
        role: "gridcell",
        className: "".concat(prefixCls, "-grid")
      }, /*#__PURE__*/React.createElement("ul", {
        className: "".concat(prefixCls, "-grid-row")
      }, _mapInstanceProperty(week).call(week, day => {
        const {
          date,
          dayString,
          weekday,
          isToday
        } = day;
        const listCls = cls({
          ["".concat(cssClasses.PREFIX, "-today")]: isToday,
          ["".concat(cssClasses.PREFIX, "-weekend")]: markWeekend && day.isWeekend
        });
        return /*#__PURE__*/React.createElement("li", {
          key: "".concat(date.toString(), "-weekheader"),
          className: listCls
        }, /*#__PURE__*/React.createElement("span", {
          className: "".concat(cssClasses.PREFIX, "-today-date")
        }, dayString), /*#__PURE__*/React.createElement("span", null, weekday));
      }))));
    };

    this.renderAllDayEvents = events => {
      const list = _mapInstanceProperty(events).call(events, (event, ind) => {
        var _context3;

        const {
          leftPos,
          width,
          topInd,
          children,
          key
        } = event;
        const top = "".concat(topInd, "em");
        const style = {
          left: toPercent(leftPos),
          width: toPercent(width),
          top
        };
        return /*#__PURE__*/React.createElement("li", {
          className: _concatInstanceProperty(_context3 = "".concat(cssClasses.PREFIX, "-event-item ")).call(_context3, cssClasses.PREFIX, "-event-allday"),
          key: "allDay-".concat(ind),
          style: style
        }, children);
      });

      return list;
    };

    this.renderAllDay = locale => {
      var _context4, _context5, _context6, _context7;

      const {
        allDay
      } = this.state.parsedEvents;
      const parsed = this.foundation.parseWeeklyAllDayEvents(allDay);
      const maxRowHeight = calcRowHeight(parsed);
      const style = {
        height: "".concat(maxRowHeight, "em")
      };
      const {
        markWeekend
      } = this.props;
      const {
        week
      } = this.weeklyData;
      return /*#__PURE__*/React.createElement("div", {
        className: "".concat(allDayCls),
        style: style
      }, /*#__PURE__*/React.createElement("ul", {
        className: _concatInstanceProperty(_context4 = _concatInstanceProperty(_context5 = "".concat(cssClasses.PREFIX, "-tag ")).call(_context5, allDayCls, "-tag ")).call(_context4, prefixCls, "-sticky-left")
      }, /*#__PURE__*/React.createElement("span", null, locale.allDay)), /*#__PURE__*/React.createElement("div", {
        role: "gridcell",
        className: _concatInstanceProperty(_context6 = "".concat(cssClasses.PREFIX, "-content ")).call(_context6, allDayCls, "-content")
      }, /*#__PURE__*/React.createElement("ul", {
        className: "".concat(allDayCls, "-skeleton")
      }, _mapInstanceProperty(_context7 = _Object$keys(week)).call(_context7, (date, ind) => {
        const listCls = cls({
          ["".concat(cssClasses.PREFIX, "-weekend")]: markWeekend && week[date].isWeekend
        });
        return /*#__PURE__*/React.createElement("li", {
          key: "".concat(date, "-weekgrid"),
          className: listCls
        });
      })), /*#__PURE__*/React.createElement("ul", {
        className: "".concat(cssClasses.PREFIX, "-event-items")
      }, this.renderAllDayEvents(parsed))));
    };

    this.state = {
      scrollHeight: 0,
      parsedEvents: {
        day: new _Map(),
        allDay: new _Map()
      },
      cachedKeys: []
    };
    this.foundation = new CalendarFoundation(this.adapter);
    this.dom = /*#__PURE__*/React.createRef();
    this.scrollDom = /*#__PURE__*/React.createRef();
    this.handleClick = _bindInstanceProperty(_context8 = this.handleClick).call(_context8, this);
    this.allDayRowHeight = 1;
  }

  get adapter() {
    return _Object$assign(_Object$assign({}, super.adapter), {
      setWeeklyData: data => {
        this.weeklyData = data;
      },
      getWeeklyData: () => this.weeklyData,
      updateScrollHeight: scrollHeight => {
        this.setState({
          scrollHeight
        });
      },
      setParsedEvents: parsedEvents => {
        this.setState({
          parsedEvents: parsedEvents
        });
      },
      cacheEventKeys: cachedKeys => {
        this.setState({
          cachedKeys
        });
      }
    });
  }

  componentDidMount() {
    this.foundation.init();
    const {
      scrollHeight
    } = this.scrollDom.current;
    this.dom.current.scrollTop = this.props.scrollTop;
    this.foundation.notifyScrollHeight(scrollHeight);
    this.foundation.parseWeeklyEvents();
  }

  componentDidUpdate(prevProps, prevState) {
    var _context9;

    const prevEventKeys = prevState.cachedKeys;

    const nowEventKeys = _mapInstanceProperty(_context9 = this.props.events).call(_context9, event => event.key);

    if (!_isEqual(prevEventKeys, nowEventKeys)) {
      this.foundation.parseWeeklyEvents();
    }
  }

  componentWillUnmount() {
    this.foundation.destroy();
  }

  render() {
    const {
      renderTimeDisplay,
      className,
      height,
      width,
      style,
      header
    } = this.props;
    const weekCls = cls(prefixCls, className);

    const weekStyle = _Object$assign({
      height,
      width
    }, style);

    return /*#__PURE__*/React.createElement(LocaleConsumer, {
      componentName: "Calendar"
    }, (locale, localeCode, dateFnsLocale) => /*#__PURE__*/React.createElement("div", {
      className: weekCls,
      style: weekStyle,
      ref: this.dom
    }, /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-sticky-top")
    }, header, this.renderHeader(dateFnsLocale), this.renderAllDay(locale)), /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-scroll-wrapper")
    }, /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-scroll"),
      ref: this.scrollDom
    }, /*#__PURE__*/React.createElement(TimeCol, {
      className: "".concat(prefixCls, "-sticky-left"),
      renderTimeDisplay: renderTimeDisplay
    }), this.renderDayGrid()))));
  }

}
WeekCalendar.propTypes = {
  displayValue: PropTypes.instanceOf(Date),
  header: PropTypes.node,
  events: PropTypes.array,
  mode: PropTypes.string,
  showCurrTime: PropTypes.bool,
  markWeekend: PropTypes.bool,
  scrollTop: PropTypes.number,
  renderTimeDisplay: PropTypes.func,
  dateGridRender: PropTypes.func,
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  style: PropTypes.object,
  className: PropTypes.string
};
WeekCalendar.defaultProps = {
  displayValue: new Date(),
  events: [],
  mode: 'week'
};
WeekCalendar.contextType = localeContext;
