// @douyinfe/semi-ui/lib/es/navigation/Footer.js
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cls from 'classnames';
import { strings, cssClasses } from '@douyinfe/semi-foundation/lib/es/navigation/constants';
import CollapseButton from './CollapseButton';
import '@douyinfe/semi-foundation/lib/es/navigation/navigation.css';
import NavContext from './nav-context';
export default class NavFooter extends PureComponent {
  constructor() {
    super(...arguments);

    this.renderCollapseButton = () => {
      const {
        collapseButton,
        collapseText
      } = this.props;

      if ( /*#__PURE__*/React.isValidElement(collapseButton)) {
        return collapseButton;
      }

      const {
        onCollapseChange,
        prefixCls,
        locale,
        isCollapsed
      } = this.context;
      return /*#__PURE__*/React.createElement(CollapseButton, _Object$assign({
        prefixCls: prefixCls,
        isCollapsed: isCollapsed,
        locale: locale,
        onClick: onCollapseChange,
        collapseText: collapseText
      }, collapseButton));
    };
  }

  render() {
    const {
      style,
      className,
      collapseButton
    } = this.props;
    let {
      children
    } = this.props;
    const {
      isCollapsed,
      mode
    } = this.context;

    if (! /*#__PURE__*/React.isValidElement(children) && collapseButton && mode !== strings.MODE_HORIZONTAL) {
      children = this.renderCollapseButton();
    }

    const wrapCls = cls(className, "".concat(cssClasses.PREFIX, "-footer"), {
      ["".concat(cssClasses.PREFIX, "-footer-collapsed")]: isCollapsed
    });
    return /*#__PURE__*/React.createElement("div", {
      className: wrapCls,
      style: style
    }, children);
  }

}
NavFooter.contextType = NavContext;
NavFooter.propTypes = {
  children: PropTypes.node,
  style: PropTypes.object,
  className: PropTypes.string,
  collapseButton: PropTypes.oneOfType([PropTypes.node, PropTypes.bool]),
  collapseText: PropTypes.func
};
NavFooter.defaultProps = {
  collapseButton: false
};
