// @douyinfe/semi-ui/lib/cjs/navigation/SubNavTransition.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _semiAnimationReact = require("@douyinfe/semi-animation-react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const ease = 'cubicBezier(.25,.1,.25,1)';

const formatStyle = function formatStyle(_ref) {
  let {
    maxHeight,
    opacity
  } = _ref;
  return {
    maxHeight,
    opacity
  };
};

function SubNavTransition() {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const {
    children,
    isCollapsed,
    maxHeight = 999
  } = props; // eslint-disable-next-line no-unused-vars

  const [immediate, setImmediate] = (0, _react.useState)(false); // useEffect(() => {
  //     setImmediate(isCollapsed);
  // }, [isCollapsed]);

  return /*#__PURE__*/_react.default.createElement(_semiAnimationReact.Transition, {
    from: {
      maxHeight: 0,
      opacity: 0
    },
    enter: {
      maxHeight: {
        val: maxHeight,
        easing: 'easeInQuad',
        duration: 250
      },
      opacity: {
        val: 1,
        duration: 200,
        easing: 'cubic-bezier(0.5, -0.1, 1, 0.4)'
      }
    },
    leave: {
      maxHeight: {
        val: 0,
        easing: ease,
        duration: 250
      },
      opacity: {
        val: 0,
        duration: isCollapsed ? 1 : 200,
        easing: 'cubic-bezier(0.5, -0.1, 1, 0.4)'
      }
    },
    immediate: immediate
  }, typeof children === 'function' ? transitionStyle => children(formatStyle(transitionStyle)) : children);
}

SubNavTransition.propTypes = {
  children: _propTypes.default.any,
  isCollapsed: _propTypes.default.bool
};
var _default = SubNavTransition;
exports.default = _default;
