// @douyinfe/semi-ui/lib/cjs/dropdown/dropdownItem.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/dropdown/constants");

var _context3 = _interopRequireDefault(require("./context"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var _semiIcons = require("@douyinfe/semi-icons");

const prefixCls = _constants.cssClasses.PREFIX;

class DropdownItem extends _baseComponent.default {
  render() {
    var _context;

    const {
      children,
      disabled,
      className,
      forwardRef,
      style,
      type,
      active,
      icon
    } = this.props;
    const {
      showTick
    } = this.context;
    const itemclass = (0, _classnames.default)(className, {
      ["".concat(prefixCls, "-item")]: true,
      ["".concat(prefixCls, "-item-disabled")]: disabled,
      ["".concat(prefixCls, "-item-withTick")]: showTick,
      [(0, _concat.default)(_context = "".concat(prefixCls, "-item-")).call(_context, type)]: type,
      ["".concat(prefixCls, "-item-active")]: active
    });
    const events = {};

    if (!disabled) {
      var _context2;

      (0, _forEach.default)(_context2 = ['onClick', 'onMouseEnter', 'onMouseLeave', 'onContextMenu']).call(_context2, eventName => {
        events[eventName] = this.props[eventName];
      });
    }

    let tick = null;

    switch (true) {
      case showTick && active:
        tick = /*#__PURE__*/_react.default.createElement(_semiIcons.IconTick, null);
        break;

      case showTick && !active:
        tick = /*#__PURE__*/_react.default.createElement(_semiIcons.IconTick, {
          style: {
            color: 'transparent'
          }
        });
        break;

      default:
        tick = null;
        break;
    }

    let iconContent = null;

    if (icon) {
      iconContent = /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "-item-icon")
      }, icon);
    }

    return /*#__PURE__*/_react.default.createElement("li", (0, _assign.default)({
      role: "menuitem",
      tabIndex: -1,
      "aria-disabled": disabled
    }, events, {
      ref: ref => forwardRef(ref),
      className: itemclass,
      style: style
    }), tick, iconContent, children);
  }

}

DropdownItem.propTypes = {
  children: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),
  name: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  selected: _propTypes.default.bool,
  onClick: _propTypes.default.func,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func,
  onContextMenu: _propTypes.default.func,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  forwardRef: _propTypes.default.func,
  type: _propTypes.default.oneOf(_constants.strings.ITEM_TYPE),
  active: _propTypes.default.bool,
  icon: _propTypes.default.node
};
DropdownItem.contextType = _context3.default;
DropdownItem.defaultProps = {
  disabled: false,
  divided: false,
  selected: false,
  onMouseEnter: _noop2.default,
  onMouseLeave: _noop2.default,
  forwardRef: _noop2.default
};
var _default = DropdownItem;
exports.default = _default;
