// @douyinfe/semi-ui/lib/cjs/dropdown/dropdownMenu.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/dropdown/constants");

var _menuFoundation = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/dropdown/menuFoundation"));

var _context = _interopRequireDefault(require("./context"));

var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const prefixCls = _constants.cssClasses.PREFIX;

class DropdownMenu extends _baseComponent.default {
  constructor(props) {
    super(props);
    this.menuRef = /*#__PURE__*/_react.default.createRef();
    this.foundation = new _menuFoundation.default(this.adapter);
  }

  get adapter() {
    return (0, _assign.default)({}, super.adapter);
  }

  componentDidMount() {
    this.foundation.autoFocus(this.menuRef.current);
  }

  render() {
    const _a = this.props,
          {
      children,
      className,
      style
    } = _a,
          rest = __rest(_a, ["children", "className", "style"]);

    return /*#__PURE__*/_react.default.createElement("ul", (0, _assign.default)({
      role: "menu",
      "aria-orientation": "vertical",
      ref: this.menuRef
    }, rest, {
      className: (0, _classnames.default)("".concat(prefixCls, "-menu"), className),
      style: style,
      onKeyDown: e => this.foundation.onMenuKeydown(e)
    }), children);
  }

}

DropdownMenu.propTypes = {
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object
};
DropdownMenu.contextType = _context.default;
var _default = DropdownMenu;
exports.default = _default;
