// @douyinfe/semi-ui/lib/cjs/layout/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = exports.Layout = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/layout/constants");

require("@douyinfe/semi-foundation/lib/cjs/layout/layout.css");

var _layoutContext = _interopRequireDefault(require("./layout-context"));

var _Sider = _interopRequireDefault(require("./Sider"));

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const htmlTag = {
  Header: 'header',
  Footer: 'footer',
  Content: 'main',
  Layout: 'section'
};

function generator(type) {
  const tagName = htmlTag[type];
  const typeName = type.toLowerCase();
  return BasicComponent => class Adapter extends _react.default.PureComponent {
    render() {
      return /*#__PURE__*/_react.default.createElement(BasicComponent, (0, _assign.default)({
        role: this.props.role,
        "aria-label": this.props['aria-label'],
        type: typeName,
        tagName: tagName
      }, this.props));
    }

  };
}

class Basic extends _react.default.PureComponent {
  render() {
    var _context;

    const _a = this.props,
          {
      prefixCls,
      type,
      className,
      children,
      tagName
    } = _a,
          others = __rest(_a, ["prefixCls", "type", "className", "children", "tagName"]);

    const classString = (0, _classnames.default)(className, (0, _concat.default)(_context = "".concat(prefixCls, "-")).call(_context, type));
    return /*#__PURE__*/_react.default.createElement(tagName, (0, _assign.default)({
      className: classString
    }, others), children);
  }

}

Basic.propTypes = {
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string
};
Basic.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX
};
const Header = generator('Header')(Basic);
const Footer = generator('Footer')(Basic);
const Content = generator('Content')(Basic);

class Layout extends _react.default.Component {
  constructor(props) {
    super(props);
    this.state = {
      siders: []
    };
  }

  getSiderHook() {
    return {
      addSider: id => {
        this.setState(state => ({
          siders: [...state.siders, id]
        }));
      },
      removeSider: id => {
        this.setState(state => {
          var _context2;

          return {
            siders: (0, _filter.default)(_context2 = state.siders).call(_context2, curr => curr !== id)
          };
        });
      }
    };
  }

  render() {
    const _a = this.props,
          {
      prefixCls,
      className,
      children,
      hasSider,
      tagName
    } = _a,
          others = __rest(_a, ["prefixCls", "className", "children", "hasSider", "tagName"]);

    const {
      siders
    } = this.state;
    const classString = (0, _classnames.default)(className, prefixCls, {
      ["".concat(prefixCls, "-has-sider")]: typeof hasSider === 'boolean' ? hasSider : siders.length > 0
    });
    const Tag = tagName;
    return /*#__PURE__*/_react.default.createElement(_layoutContext.default.Provider, {
      value: {
        siderHook: this.getSiderHook()
      }
    }, /*#__PURE__*/_react.default.createElement(Tag, (0, _assign.default)({
      className: classString
    }, others), children));
  }

}

exports.Layout = Layout;
Layout.propTypes = {
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string
};
Layout.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX,
  tagName: 'section'
};
Layout.Header = Header;
Layout.Footer = Footer;
Layout.Content = Content;
Layout.Sider = _Sider.default;
var _default = Layout;
exports.default = _default;
