// @douyinfe/semi-ui/lib/cjs/layout/Sider.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/keys"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/layout/constants");

var _getDataAttr = _interopRequireDefault(require("@douyinfe/semi-foundation/lib/cjs/utils/getDataAttr"));

var _layoutContext = _interopRequireDefault(require("./layout-context"));

var _utils = require("../_utils");

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

const responsiveMap = {
  xs: '(max-width: 575px)',
  sm: '(min-width: 576px)',
  md: '(min-width: 768px)',
  lg: '(min-width: 992px)',
  xl: '(min-width: 1200px)',
  xxl: '(min-width: 1600px)'
};

const generateId = (() => {
  let i = 0;
  return () => {
    var _context;

    i += 1;
    return (0, _concat.default)(_context = "".concat(_constants.cssClasses.PREFIX, "-sider-")).call(_context, i);
  };
})();

const bpt = _constants.strings.BREAKPOINT;

class Sider extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    this.unRegisters = [];
    this.uniqueId = '';
    this.uniqueId = generateId();
  }

  componentDidMount() {
    var _context2;

    const {
      breakpoint
    } = this.props;
    const matchBpt = (0, _filter.default)(_context2 = (0, _keys.default)(responsiveMap)).call(_context2, item => breakpoint && (0, _indexOf.default)(breakpoint).call(breakpoint, item) !== -1);
    const unRegisters = (0, _map.default)(matchBpt).call(matchBpt, screen => (0, _utils.registerMediaQuery)(responsiveMap[screen], {
      match: () => {
        this.responsiveHandler(screen, true);
      },
      unmatch: () => {
        this.responsiveHandler(screen, false);
      }
    }));
    this.unRegisters = unRegisters;

    if (this.context.siderHook) {
      this.context.siderHook.addSider(this.uniqueId);
    }
  }

  componentWillUnmount() {
    var _context3;

    (0, _forEach.default)(_context3 = this.unRegisters).call(_context3, unRegister => unRegister());

    if (this.context.siderHook) {
      this.context.siderHook.removeSider(this.uniqueId);
    }
  }

  responsiveHandler(screen, matches) {
    const {
      onBreakpoint
    } = this.props;

    if (onBreakpoint) {
      onBreakpoint(screen, matches);
    }
  }

  render() {
    const _a = this.props,
          {
      prefixCls,
      className,
      children,
      style
    } = _a,
          others = __rest(_a, ["prefixCls", "className", "children", "style"]);

    const classString = (0, _classnames.default)(className, {
      ["".concat(prefixCls, "-sider")]: true
    });
    return /*#__PURE__*/_react.default.createElement("aside", (0, _assign.default)({
      className: classString,
      "aria-label": this.props['aria-label'],
      style: style
    }, (0, _getDataAttr.default)(others)), /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-sider-children")
    }, children));
  }

}

Sider.propTypes = {
  prefixCls: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  breakpoint: _propTypes.default.arrayOf(_propTypes.default.oneOf(bpt)),
  onBreakpoint: _propTypes.default.func,
  'aria-label': _propTypes.default.string,
  role: _propTypes.default.string
};
Sider.defaultProps = {
  prefixCls: _constants.cssClasses.PREFIX
};
Sider.contextType = _layoutContext.default;
var _default = Sider;
exports.default = _default;
