// @douyinfe/semi-ui/lib/cjs/descriptions/item.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/descriptions/constants");

require("@douyinfe/semi-foundation/lib/cjs/descriptions/descriptions.css");

var _descriptionsContext = _interopRequireDefault(require("./descriptions-context"));

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const prefixCls = _constants.cssClasses.PREFIX;
const keyCls = "".concat(prefixCls, "-key");
const valCls = "".concat(prefixCls, "-value");

class Item extends _react.PureComponent {
  render() {
    var _context, _context2;

    const {
      itemKey,
      hidden,
      className,
      style,
      children
    } = this.props;
    const {
      align
    } = this.context;

    if (hidden) {
      return null;
    }

    const item = align === 'plain' ? /*#__PURE__*/_react.default.createElement("tr", {
      className: className,
      style: style
    }, /*#__PURE__*/_react.default.createElement("td", {
      className: "".concat(prefixCls, "-item")
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: keyCls
    }, itemKey, ":"), /*#__PURE__*/_react.default.createElement("span", {
      className: valCls
    }, typeof children === 'function' ? children() : children))) : /*#__PURE__*/_react.default.createElement("tr", {
      className: className,
      style: style
    }, /*#__PURE__*/_react.default.createElement("th", {
      className: (0, _concat.default)(_context = "".concat(prefixCls, "-item ")).call(_context, prefixCls, "-item-th")
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: keyCls
    }, itemKey)), /*#__PURE__*/_react.default.createElement("td", {
      className: (0, _concat.default)(_context2 = "".concat(prefixCls, "-item ")).call(_context2, prefixCls, "-item-td")
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: valCls
    }, typeof children === 'function' ? children() : children)));
    return item;
  }

}

exports.default = Item;
Item.propTypes = {
  itemKey: _propTypes.default.node,
  hidden: _propTypes.default.bool,
  className: _propTypes.default.string,
  style: _propTypes.default.object
};
Item.contextType = _descriptionsContext.default;
