// @douyinfe/semi-ui/lib/cjs/table/ColumnFilter.js
"use strict";

var _WeakMap = require("@babel/runtime-corejs3/core-js-stable/weak-map");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = ColumnFilter;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _isArray = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/array/is-array"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/includes"));

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _splice = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/splice"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _semiIcons = require("@douyinfe/semi-icons");

var _constants = require("@douyinfe/semi-foundation/lib/cjs/table/constants");

var _dropdown = _interopRequireDefault(require("../dropdown"));

var _radio = require("../radio");

var _checkbox = require("../checkbox");

function _getRequireWildcardCache(nodeInterop) { if (typeof _WeakMap !== "function") return null; var cacheBabelInterop = new _WeakMap(); var cacheNodeInterop = new _WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = _Object$defineProperty && _Object$getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? _Object$getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { _Object$defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* eslint-disable no-nested-ternary */

/* eslint-disable eqeqeq */
function renderDropdown() {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let nestedElem = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  let level = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  const {
    filterMultiple = true,
    filters = [],
    filteredValue = [],
    filterDropdownVisible,
    onSelect = _noop2.default,
    onFilterDropdownVisibleChange = _noop2.default,
    trigger = 'click',
    position = 'bottom',
    renderFilterDropdownItem
  } = props;
  const dropdownProps = (0, _assign.default)((0, _assign.default)({}, props), {
    onVisibleChange: visible => onFilterDropdownVisibleChange(visible),
    trigger,
    position,
    render: /*#__PURE__*/_react.default.createElement(_dropdown.default.Menu, null, (0, _isArray.default)(filters) && (0, _map.default)(filters).call(filters, (filter, index) => {
      var _context;

      const changeFn = e => {
        const domEvent = e && e.nativeEvent;

        if (domEvent) {
          // Block this event to prevent the pop-up layer from closing
          domEvent.stopImmediatePropagation(); // Prevent bubbling and default events to prevent label click events from triggering twice

          domEvent.stopPropagation();
          domEvent.preventDefault();
        }

        let values = [...filteredValue];
        const included = (0, _includes.default)(values).call(values, filter.value);
        const idx = (0, _indexOf.default)(values).call(values, filter.value);

        if (idx > -1) {
          (0, _splice.default)(values).call(values, idx, 1);
        } else if (filterMultiple) {
          values.push(filter.value);
        } else {
          values = [filter.value];
        }

        return onSelect({
          value: filter.value,
          filteredValue: values,
          included: !included,
          domEvent
        });
      };

      const checked = (0, _includes.default)(filteredValue).call(filteredValue, filter.value);
      const {
        text
      } = filter;
      const {
        value
      } = filter;
      const key = (0, _concat.default)(_context = "".concat(level, "_")).call(_context, index);
      const dropdownItem = typeof renderFilterDropdownItem === 'function' ? renderFilterDropdownItem({
        onChange: changeFn,
        filterMultiple,
        value,
        text,
        checked,
        filteredValue,
        level
      }) : null;
      let item = dropdownItem && /*#__PURE__*/_react.default.isValidElement(dropdownItem) ? /*#__PURE__*/_react.default.cloneElement(dropdownItem, {
        key
      }) : /*#__PURE__*/_react.default.createElement(_dropdown.default.Item, {
        key: key,
        onClick: changeFn
      }, filterMultiple ? /*#__PURE__*/_react.default.createElement(_checkbox.Checkbox, {
        checked: checked
      }, text) : /*#__PURE__*/_react.default.createElement(_radio.Radio, {
        checked: checked
      }, text));

      if ((0, _isArray.default)(filter.children) && filter.children.length) {
        const childrenDropdownProps = (0, _assign.default)((0, _assign.default)({}, props), {
          filters: filter.children,
          trigger: 'hover',
          position: 'right'
        });
        delete childrenDropdownProps.filterDropdownVisible;
        item = renderDropdown(childrenDropdownProps, item, level + 1);
      }

      return item;
    }))
  });

  if (filterDropdownVisible != null) {
    dropdownProps.visible = filterDropdownVisible;
  }

  return /*#__PURE__*/_react.default.createElement(_dropdown.default, (0, _assign.default)({}, dropdownProps, {
    key: "Dropdown_level_".concat(level)
  }), nestedElem);
}

function ColumnFilter() {
  let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const {
    prefixCls = _constants.cssClasses.PREFIX,
    filteredValue,
    filterIcon = 'filter',
    renderFilterDropdown,
    filterDropdownProps
  } = props;
  let {
    filterDropdown = null
  } = props;
  const finalCls = (0, _classnames.default)("".concat(prefixCls, "-column-filter"), {
    on: (0, _isArray.default)(filteredValue) && filteredValue.length
  });
  let iconElem;

  if (typeof filterIcon === 'function') {
    iconElem = filterIcon((0, _isArray.default)(filteredValue) && filteredValue.length > 0);
  } else if ( /*#__PURE__*/(0, _react.isValidElement)(filterIcon)) {
    iconElem = filterIcon;
  } else {
    iconElem = /*#__PURE__*/_react.default.createElement("div", {
      className: finalCls
    }, /*#__PURE__*/_react.default.createElement(_semiIcons.IconFilter, {
      role: "button",
      "aria-label": "Filter data with this column",
      "aria-haspopup": "listbox",
      tabIndex: -1,
      size: "small"
    }));
  }

  const renderProps = (0, _assign.default)((0, _assign.default)({}, props), filterDropdownProps);
  filterDropdown = /*#__PURE__*/_react.default.isValidElement(filterDropdown) ? filterDropdown : typeof renderFilterDropdown === 'function' ? renderFilterDropdown(renderProps, {
    iconElem
  }) : renderDropdown(renderProps, iconElem);
  return filterDropdown;
}
