// @douyinfe/semi-ui/lib/cjs/table/index.js
"use strict";

var _context;

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _forEachInstanceProperty = require("@babel/runtime-corejs3/core-js-stable/instance/for-each");

var _Object$keys = require("@babel/runtime-corejs3/core-js-stable/object/keys");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

var _exportNames = {};
exports.default = void 0;

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Table = _interopRequireDefault(require("./Table"));

var _ResizableTable = _interopRequireDefault(require("./ResizableTable"));

var _Column = _interopRequireDefault(require("./Column"));

var _constants = require("@douyinfe/semi-foundation/lib/cjs/table/constants");

var _interface = require("./interface");

_forEachInstanceProperty(_context = _Object$keys(_interface)).call(_context, function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;

  _Object$defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});

class Table extends _react.default.PureComponent {
  constructor(props) {
    super(props);

    this.getCurrentPageData = () => this.tableRef.current && this.tableRef.current.getCurrentPageData();

    this.tableRef = /*#__PURE__*/_react.default.createRef();
  }

  render() {
    // eslint-disable-next-line prefer-destructuring
    const props = this.props;

    if (props.resizable) {
      return /*#__PURE__*/_react.default.createElement(_ResizableTable.default, (0, _assign.default)({}, props, {
        ref: this.tableRef
      }));
    } else {
      return /*#__PURE__*/_react.default.createElement(_Table.default, (0, _assign.default)({}, props, {
        ref: this.tableRef
      }));
    }
  }

}

Table.Column = _Column.default;
Table.DEFAULT_KEY_COLUMN_SELECTION = _constants.strings.DEFAULT_KEY_COLUMN_SELECTION;
Table.DEFAULT_KEY_COLUMN_EXPAND = _constants.strings.DEFAULT_KEY_COLUMN_EXPAND;
Table.propTypes = (0, _assign.default)((0, _assign.default)({}, _Table.default.propTypes), {
  resizable: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.object])
});
Table.defaultProps = {
  hideExpandedColumn: true
};
var _default = Table;
exports.default = _default;
