// @douyinfe/semi-ui/lib/cjs/table/ResizableHeaderCell.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _react = _interopRequireDefault(require("react"));

var _reactResizable = require("react-resizable");

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && (0, _indexOf.default)(e).call(e, p) < 0) t[p] = s[p];

  if (s != null && typeof _getOwnPropertySymbols.default === "function") for (var i = 0, p = (0, _getOwnPropertySymbols.default)(s); i < p.length; i++) {
    if ((0, _indexOf.default)(e).call(e, p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

class ResizableHeaderCell extends _react.default.PureComponent {
  render() {
    var _context;

    const _a = this.props,
          {
      onResize,
      onResizeStart,
      onResizeStop,
      width
    } = _a,
          restProps = __rest(_a, ["onResize", "onResizeStart", "onResizeStop", "width"]);

    if (typeof width !== 'number') {
      return /*#__PURE__*/_react.default.createElement("th", (0, _assign.default)({}, restProps));
    }

    let {
      children
    } = restProps; // Fragment must be used here, otherwise there will be an error (seemingly a react-resizable@1.9.0 problem)

    children = (0, _map.default)(_context = _react.default.Children).call(_context, children, (child, index) => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: index
    }, child));
    return /*#__PURE__*/_react.default.createElement(_reactResizable.Resizable, {
      width: width,
      height: 0,
      onResize: onResize,
      onResizeStart: onResizeStart,
      onResizeStop: onResizeStop,
      draggableOpts: {
        enableUserSelectHack: false
      }
    }, /*#__PURE__*/_react.default.createElement("th", (0, _assign.default)({}, restProps), children));
  }

}

var _default = ResizableHeaderCell;
exports.default = _default;
