// @douyinfe/semi-foundation/lib/cjs/checkbox/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.numbers = exports.checkboxGroupClasses = exports.checkboxClasses = void 0;

var _constants = require("../base/constants");

const checkboxClasses = {
  PREFIX: "".concat(_constants.BASE_CLASS_PREFIX, "-checkbox"),
  INNER: "".concat(_constants.BASE_CLASS_PREFIX, "-checkbox-inner"),
  TEXT: "".concat(_constants.BASE_CLASS_PREFIX, "-checkbox-text"),
  INPUT: "".concat(_constants.BASE_CLASS_PREFIX, "-checkbox-input"),
  CHECKED: "".concat(_constants.BASE_CLASS_PREFIX, "-checkbox-checked"),
  DISABLED: "".concat(_constants.BASE_CLASS_PREFIX, "-checkbox-disabled"),
  BUTTON: "".concat(_constants.BASE_CLASS_PREFIX, "-checkbox-button"),
  WRAPPER: ''
};
exports.checkboxClasses = checkboxClasses;
const checkboxGroupClasses = {
  PREFIX: "".concat(_constants.BASE_CLASS_PREFIX, "-checkboxGroup"),
  INNER: "".concat(_constants.BASE_CLASS_PREFIX, "-checkboxGroup-inner"),
  TEXT: "".concat(_constants.BASE_CLASS_PREFIX, "-checkboxGroup-text"),
  INPUT: "".concat(_constants.BASE_CLASS_PREFIX, "-checkboxGroup-input"),
  CHECKED: "".concat(_constants.BASE_CLASS_PREFIX, "-checkboxGroup-checked"),
  DISABLED: "".concat(_constants.BASE_CLASS_PREFIX, "-checkboxGroup-disabled")
};
exports.checkboxGroupClasses = checkboxGroupClasses;
const strings = {
  DIRECTION_SET: ['horizontal', 'vertical'],
  TYPE_DEFAULT: 'default',
  TYPE_CARD: 'card',
  TYPE_PURECARD: 'pureCard',
  DEFAULT_DIRECTION: 'vertical'
};
exports.strings = strings;
const numbers = {};
exports.numbers = numbers;
