// @douyinfe/semi-foundation/lib/cjs/tabs/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.strings = exports.numbers = exports.cssClasses = void 0;

var _constants = require("../base/constants");

const cssClasses = {
  TABS: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs"),
  TABS_BAR: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-bar"),
  TABS_BAR_LINE: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-bar-line"),
  TABS_BAR_CARD: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-bar-card"),
  TABS_BAR_BUTTON: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-bar-button"),
  TABS_BAR_EXTRA: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-bar-extra"),
  TABS_TAB: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-tab"),
  TABS_TAB_ACTIVE: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-tab-active"),
  TABS_TAB_DISABLED: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-tab-disabled"),
  TABS_CONTENT: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-content"),
  TABS_CONTENT_ANIMATED: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-content-animated"),
  TABS_CONTENT_NO_ANIMATED: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-content-no-animated"),
  TABS_PANE: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-pane"),
  TABS_PANE_INACTIVE: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-pane-inactive"),
  TABS_PANE_ACTIVE: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-pane-active"),
  TABS_PANE_MOTION_OVERLAY: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-pane-motion-overlay"),
  TABS_PANE_ANIMATING: "".concat(_constants.BASE_CLASS_PREFIX, "-tabs-pane-animating")
};
exports.cssClasses = cssClasses;
const numbers = {
  DEFAULT_ACTIVE_KEY: 1
};
exports.numbers = numbers;
const strings = {
  TYPE_MAP: ['line', 'card', 'button'],
  SIZE: ['small', 'medium', 'large'],
  POSITION_MAP: ['top', 'left']
};
exports.strings = strings;
